// metalsmith-bundled-components/lib/index.cjs
'use strict';

var path = require('path');
var fs = require('fs');
var os = require('os');
var esbuild = require('esbuild');
var postcssPlugin = require('esbuild-plugin-postcss');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var path__default = /*#__PURE__*/_interopDefaultLegacy(path);
var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
var os__default = /*#__PURE__*/_interopDefaultLegacy(os);
var postcssPlugin__default = /*#__PURE__*/_interopDefaultLegacy(postcssPlugin);

/**
 * @typedef {Object} PostCSSConfiguration
 * @property {Array} [plugins] - Array of PostCSS plugins
 * @property {Object} [options] - PostCSS processing options
 * @property {boolean} [enabled] - Whether PostCSS processing is enabled
 */

/**
 * @typedef {Object} ValidationOptions
 * @property {boolean} [enabled] - Enable/disable validation
 * @property {boolean} [strict] - Fail build on validation errors (vs warnings only)
 * @property {boolean} [reportAllErrors] - Report all errors vs stop on first error
 */

/**
 * @typedef {Object} BundledComponentsOptions
 * @property {string} [basePath] - Path to base/partial components directory
 * @property {string} [sectionsPath] - Path to section components directory
 * @property {string} [layoutsPath] - Path to layouts directory for scanning template includes
 * @property {string} [cssDest] - Output path for bundled CSS file
 * @property {string} [jsDest] - Output path for bundled JavaScript file
 * @property {string} [mainCSSEntry] - Main CSS entry point (design tokens, base styles)
 * @property {string} [mainJSEntry] - Main JavaScript entry point (app initialization)
 * @property {boolean} [minifyOutput] - Enable esbuild minification for production
 * @property {PostCSSConfiguration} [postcss] - PostCSS configuration via esbuild plugin
 * @property {ValidationOptions} [validation] - Component property validation settings
 */

/** @type {BundledComponentsOptions} */
const defaults = {
  basePath: 'lib/layouts/components/_partials',
  // Base/partial components (buttons, cards, etc.)
  sectionsPath: 'lib/layouts/components/sections',
  // Section components (hero, banner, etc.)
  layoutsPath: 'lib/layouts',
  // Layouts directory for scanning includes
  cssDest: 'assets/main.css',
  // Output path for bundled CSS (main + components)
  jsDest: 'assets/main.js',
  // Output path for bundled JS (main + components)
  mainCSSEntry: 'lib/assets/main.css',
  // Main CSS entry (design tokens, base styles)
  mainJSEntry: 'lib/assets/main.js',
  // Main JS entry (app initialization)
  minifyOutput: false,
  // Enable esbuild minification
  postcss: {
    enabled: false,
    // PostCSS via esbuild plugin
    plugins: [],
    // PostCSS plugins array
    options: {} // Additional PostCSS options
  },
  validation: {
    enabled: true,
    // Component property validation
    strict: false,
    // Warn vs fail on validation errors
    reportAllErrors: true // Report all errors vs stop on first
  }
};

/**
 * Normalize and merge plugin options with defaults
 * 
 * Ensures all configuration objects have required properties and applies
 * sensible defaults for the simplified esbuild-based architecture.
 * 
 * @param {BundledComponentsOptions} [options] - User-provided options
 * @returns {BundledComponentsOptions} Normalized options with all defaults applied
 */
function normalizeOptions(options) {
  const normalized = {
    ...defaults,
    ...(options || {})
  };

  // Ensure postcss configuration has all required properties
  normalized.postcss = {
    ...defaults.postcss,
    ...(normalized.postcss || {})
  };

  // Ensure validation configuration has all required properties
  normalized.validation = {
    ...defaults.validation,
    ...(normalized.validation || {})
  };
  return normalized;
}

/**
 * @typedef {Object} ComponentManifest
 * @property {string} name - Component name
 * @property {string} [type] - Component type (e.g., 'auto', 'manual')
 * @property {string[]} styles - Array of CSS file names
 * @property {string[]} scripts - Array of JS file names
 * @property {string[]} dependencies - Array of component names this component depends on (legacy)
 * @property {string[]} requires - Array of component names this component requires (new format)
 */

/**
 * @typedef {Object} Component
 * @property {string} name - Component name
 * @property {string} [type] - Component type
 * @property {string[]} styles - Array of CSS file names
 * @property {string[]} scripts - Array of JS file names
 * @property {string[]} dependencies - Array of component names this component depends on (legacy)
 * @property {string[]} requires - Array of component names this component requires (new format)
 * @property {string} path - Full path to component directory
 */

/**
 * Collect all components from a directory (base or sections)
 *
 * Scans the provided directory for component subdirectories and loads
 * their manifests or auto-generates them. Supports both explicit manifests
 * and auto-discovery based on file patterns.
 *
 * @param {string} dirPath - Directory to scan for components
 * @returns {Array} Array of component objects with metadata and file paths
 */
function collectComponents(dirPath) {
  // Component directories are optional - return empty array if not found
  // This allows projects to have only partials or only sections
  if (!fs__default["default"].existsSync(dirPath)) {
    return [];
  }

  // Read all items in the component directory (e.g., 'button', 'hero', etc.)
  const items = fs__default["default"].readdirSync(dirPath);
  const components = [];
  items.forEach(item => {
    const itemPath = path__default["default"].join(dirPath, item);
    const stats = fs__default["default"].statSync(itemPath);

    // Skip files - we only process directories (each directory is a component)
    if (!stats.isDirectory()) {
      return;
    }

    // Load the component from its directory (reads manifest or auto-generates)
    const component = loadComponent(itemPath, item);

    // Only add successfully loaded components (null means loading failed)
    if (component) {
      components.push(component);
    }
  });
  return components;
}

/**
 * Load a single component from directory
 *
 * Attempts to read a manifest.json file from the component directory.
 * If no manifest exists, auto-generates one based on file conventions.
 *
 * @param {string} componentPath - Component directory path
 * @param {string} componentName - Component name
 * @returns {Object|null} Component object or null if invalid
 */
function loadComponent(componentPath, componentName) {
  const manifestPath = path__default["default"].join(componentPath, 'manifest.json');
  let manifest;

  // Try to load an explicit manifest file if it exists
  if (fs__default["default"].existsSync(manifestPath)) {
    try {
      manifest = JSON.parse(fs__default["default"].readFileSync(manifestPath, 'utf8'));
    } catch (error) {
      // Invalid JSON - log error and skip this component
      console.error(`Invalid manifest.json in ${componentName}: ${error.message}`);
      return null;
    }
  } else {
    // No manifest found - auto-generate based on component name conventions
    // Looks for componentName.css and componentName.js files
    manifest = autoGenerateManifest(componentPath, componentName);
  }

  // Validate that manifest has the required 'name' field
  if (!manifest.name) {
    console.error(`Missing 'name' in manifest for ${componentName}`);
    return null;
  }

  // Return the complete component object with normalized arrays
  // and the full filesystem path for later asset resolution
  return {
    ...manifest,
    path: componentPath,
    // Ensure arrays exist even if not defined in manifest
    styles: manifest.styles || [],
    scripts: manifest.scripts || [],
    dependencies: manifest.dependencies || []
  };
}

/**
 * Auto-generate manifest for components without one
 *
 * Creates a manifest based on naming conventions:
 * - Looks for {componentName}.css
 * - Looks for {componentName}.js
 * - No dependencies (must be explicit via manifest.json)
 *
 * This enables "convention over configuration" for simple components.
 *
 * @param {string} componentPath - Component directory
 * @param {string} componentName - Component name
 * @returns {Object} Generated manifest
 */
function autoGenerateManifest(componentPath, componentName) {
  // Expected file names based on component name
  // Example: 'button' component looks for 'button.css' and 'button.js'
  const cssFile = `${componentName}.css`;
  const jsFile = `${componentName}.js`;
  return {
    name: componentName,
    type: 'auto',
    // Mark as auto-generated for debugging
    // Only include files that actually exist on disk
    styles: fs__default["default"].existsSync(path__default["default"].join(componentPath, cssFile)) ? [cssFile] : [],
    scripts: fs__default["default"].existsSync(path__default["default"].join(componentPath, jsFile)) ? [jsFile] : [],
    // Auto-generated manifests can't know dependencies - must be explicit
    dependencies: []
  };
}

/**
 * Create a Map from component array for efficient lookups
 *
 * Converts the component array into a Map for O(1) lookups by name.
 * This is used throughout the plugin for:
 * - Dependency resolution (checking if required components exist)
 * - Template analysis (getting component manifests)
 * - Validation (accessing component schemas)
 *
 * @param {Array} components - Array of components
 * @returns {Map} Component map keyed by name
 * @throws {Error} If duplicate component names are found
 */
function createComponentMap(components) {
  const componentMap = new Map();
  components.forEach(component => {
    // Detect duplicate component names across partials and sections
    // This prevents ambiguity in dependency resolution
    if (componentMap.has(component.name)) {
      throw new Error(`Duplicate component name: ${component.name}`);
    }

    // Map stores: componentName → full component object with path, styles, scripts, etc.
    componentMap.set(component.name, component);
  });
  return componentMap;
}

/**
 * Simple requirement validation for component dependencies
 *
 * Validates that components marked as 'requires' or 'dependencies' exist.
 * Components are namespaced (CSS) and wrapped in IIFEs (JS), so load order
 * doesn't affect functionality - only existence matters.
 */

/**
 * Validate that all required components exist in the component map
 * 
 * Checks both 'requires' (new format) and 'dependencies' (backward compatibility)
 * for component manifest declarations. Does not enforce ordering.
 * 
 * @param {Map} componentMap - Map of component names to component objects
 * @returns {Array<string>} Array of error messages (empty if all requirements met)
 */
function validateRequirements(componentMap) {
  const errors = [];
  componentMap.forEach(component => {
    // Check both 'requires' (new) and 'dependencies' (backward compat)
    const requirements = component.requires || component.dependencies || [];
    requirements.forEach(required => {
      if (!componentMap.has(required)) {
        errors.push(`Component "${component.name}" requires "${required}" which was not found`);
      }
    });
  });
  return errors;
}

/**
 * @typedef {Object} BundledAssets
 * @property {string|null} css - Bundled CSS content or null if no CSS
 * @property {string|null} js - Bundled JavaScript content or null if no JS
 */

/**
 * Bundle main entries and components using esbuild with plugins for modern, optimized output
 * - Uses esbuild.build() for full plugin ecosystem support
 * - Integrates PostCSS via esbuild-plugin-postcss
 * - Merges main entries with component assets into single output files
 * - Processing order: Main entries → Base components → Section components
 * - Supports tree shaking, minification (via minifyOutput flag), and modern JS output
 * 
 * @param {Array} baseComponents - Base/partial components
 * @param {Array} sectionComponents - Section components
 * @param {string} projectRoot - Project root directory for resolving paths
 * @param {Object} options - Plugin options including minifyOutput, PostCSS config, etc.
 * @returns {Promise<BundledAssets>} Promise resolving to merged main + component assets
 */
async function bundleWithESBuild(baseComponents, sectionComponents, projectRoot, options) {
  // Use Sets to automatically deduplicate shared CSS/JS files across components
  // Multiple components can reference same utility files (e.g., shared/button.css)
  const cssEntryPoints = new Set();
  const jsEntryPoints = new Set();

  // Track temp files for cleanup at the end
  const tempFiles = [];

  // Create temp directory for bundling operations
  const tempDir = fs__default["default"].mkdtempSync(path__default["default"].join(os__default["default"].tmpdir(), 'metalsmith-bundled-'));

  // Add main entries first (if specified)
  // If no main entries are specified, will bundle components only
  if (options.mainCSSEntry) {
    const mainCSSPath = path__default["default"].resolve(projectRoot, options.mainCSSEntry);
    if (fs__default["default"].existsSync(mainCSSPath)) {
      cssEntryPoints.add(mainCSSPath);
    }
  }
  if (options.mainJSEntry) {
    const mainJSPath = path__default["default"].resolve(projectRoot, options.mainJSEntry);
    if (fs__default["default"].existsSync(mainJSPath)) {
      jsEntryPoints.add(mainJSPath);
    }
  }

  // Collect component files
  // Base components first (for foundational styles), then sections
  const allComponents = [...baseComponents, ...sectionComponents];
  allComponents.forEach(component => {
    // Add CSS files (Set automatically deduplicates shared dependencies)
    component.styles.forEach(styleFile => {
      const filePath = path__default["default"].join(component.path, styleFile);
      if (fs__default["default"].existsSync(filePath)) {
        cssEntryPoints.add(filePath);
      }
    });

    // Add JS files (Set automatically deduplicates shared dependencies)
    component.scripts.forEach(scriptFile => {
      const filePath = path__default["default"].join(component.path, scriptFile);
      if (fs__default["default"].existsSync(filePath)) {
        jsEntryPoints.add(filePath);
      }
    });
  });

  // Configure esbuild plugins for enhanced CSS/JS processing
  const plugins = [];

  // Add PostCSS plugin if configured
  if (options.postcss && options.postcss.enabled) {
    plugins.push(postcssPlugin__default["default"].default({
      plugins: options.postcss.plugins || [],
      ...(options.postcss.options || {})
    }));
  }

  // Bundle CSS if we have any CSS files
  let cssContent = null;
  if (cssEntryPoints.size > 0) {
    try {
      // Concatenate all CSS files (main + components) into one temp file
      const cssContents = [];
      for (const cssFile of cssEntryPoints) {
        if (fs__default["default"].existsSync(cssFile)) {
          const content = fs__default["default"].readFileSync(cssFile, 'utf8');
          cssContents.push(content);
        }
      }
      if (cssContents.length > 0) {
        // 1. Concatenate main CSS with all component CSS
        const combinedCSS = cssContents.join('\n\n');

        // 2. Copy concatenated CSS to temp directory as main.css
        const tempMainCSS = path__default["default"].join(tempDir, 'main.css');
        fs__default["default"].writeFileSync(tempMainCSS, combinedCSS);
        tempFiles.push(tempMainCSS);

        // 3. Copy @import files to temp directory (if main CSS entry exists)
        if (options.mainCSSEntry && cssEntryPoints.size > 0) {
          const mainCSSFile = Array.from(cssEntryPoints)[0]; // Convert Set to array to get first item
          const mainCSSDir = path__default["default"].dirname(mainCSSFile);

          // Copy all CSS files from the same directory as main.css (for @import './file.css')
          const assetsFiles = fs__default["default"].readdirSync(mainCSSDir);
          assetsFiles.forEach(file => {
            if (file.endsWith('.css') && file !== path__default["default"].basename(mainCSSFile)) {
              fs__default["default"].copyFileSync(path__default["default"].join(mainCSSDir, file), path__default["default"].join(tempDir, file));
              tempFiles.push(path__default["default"].join(tempDir, file));
            }
          });

          // Also copy styles directory if it exists (for @import './styles/file.css')
          const stylesDir = path__default["default"].join(mainCSSDir, 'styles');
          if (fs__default["default"].existsSync(stylesDir)) {
            const tempStylesDir = path__default["default"].join(tempDir, 'styles');
            fs__default["default"].mkdirSync(tempStylesDir, {
              recursive: true
            });
            const styleFiles = fs__default["default"].readdirSync(stylesDir);
            styleFiles.forEach(file => {
              if (file.endsWith('.css')) {
                fs__default["default"].copyFileSync(path__default["default"].join(stylesDir, file), path__default["default"].join(tempStylesDir, file));
                tempFiles.push(path__default["default"].join(tempStylesDir, file));
              }
            });
          }
        }

        // 4. Process the concatenated CSS with esbuild to resolve @imports and apply minification
        const result = await esbuild.build({
          entryPoints: [tempMainCSS],
          bundle: true,
          // Enable bundling to resolve @import statements
          write: false,
          plugins,
          loader: {
            '.css': 'css'
          },
          minify: options.minifyOutput === true,
          logLevel: 'silent',
          absWorkingDir: tempDir // Use temp directory where files are copied
        });

        // 5. Extract processed CSS content (ready for output to build directory)
        if (result.outputFiles && result.outputFiles.length > 0) {
          cssContent = result.outputFiles[0].text;
        }

        // 6. Temp directory cleanup happens at the end of the function
      }
    } catch (error) {
      console.error('Error bundling CSS:', error.message);
    }
  }

  // Bundle JS if we have any JS files
  let jsContent = null;
  if (jsEntryPoints.size > 0) {
    try {
      // Create a temporary entry file in system temp directory
      // tempDir already created above
      const tempJSEntry = path__default["default"].join(tempDir, 'entry.js');
      tempFiles.push(tempDir); // Track temp directory for cleanup

      const jsEntryArray = Array.from(jsEntryPoints).filter(file => fs__default["default"].existsSync(file));

      // Check if we have any valid JS files
      if (jsEntryArray.length === 0) {
        jsContent = null;
      } else {
        // Separate main entry from component files
        let mainEntryFile = null;
        const componentFiles = [];
        jsEntryArray.forEach((file, index) => {
          const isMainEntry = index === 0 && options.mainJSEntry && path__default["default"].resolve(projectRoot, options.mainJSEntry) === file;
          if (isMainEntry) {
            mainEntryFile = file;
          } else {
            componentFiles.push(file);
          }
        });

        // Create entry file content
        let entryContent = '';

        // Import main entry if it exists
        if (mainEntryFile) {
          entryContent += `// Main entry: ${path__default["default"].relative(projectRoot, mainEntryFile)}\n`;
          entryContent += `import '${mainEntryFile}';\n\n`;
        }

        // For component files, we need to maintain isolation
        // Since ES modules are already isolated, we just import them
        componentFiles.forEach((file, index) => {
          entryContent += `// Component ${index}: ${path__default["default"].relative(projectRoot, file)}\n`;
          entryContent += `import '${file}';\n`;
        });

        // Write the entry file with import statements
        fs__default["default"].writeFileSync(tempJSEntry, entryContent);
        tempFiles.push(tempJSEntry); // Track for cleanup later

        // Verify file was created
        if (!fs__default["default"].existsSync(tempJSEntry)) {
          throw new Error(`Failed to create temp JS file at ${tempJSEntry}`);
        }
        const result = await esbuild.build({
          entryPoints: [tempJSEntry],
          bundle: true,
          write: false,
          plugins,
          format: 'iife',
          minify: options.minifyOutput === true,
          target: options.target || 'es2020',
          treeShaking: true,
          logLevel: 'silent',
          absWorkingDir: projectRoot // Set working directory for better path resolution
        });

        // Extract JS from the result
        if (result.outputFiles && result.outputFiles.length > 0) {
          jsContent = result.outputFiles[0].text;
        }

        // Don't clean up here - will clean up all temp files at the end
      }
    } catch (error) {
      console.error('Error bundling JS:', error.message);
    }
  }

  // Clean up all temp files and directories at the end
  tempFiles.forEach(tempPath => {
    try {
      if (fs__default["default"].existsSync(tempPath)) {
        const stats = fs__default["default"].statSync(tempPath);
        if (stats.isDirectory()) {
          // Remove directory and its contents
          fs__default["default"].rmSync(tempPath, {
            recursive: true,
            force: true
          });
        } else {
          // Remove file
          fs__default["default"].unlinkSync(tempPath);
        }
      }
    } catch {
      // Silently ignore cleanup errors - they're not critical
    }
  });
  return {
    css: cssContent,
    js: jsContent
  };
}

/**
 * @typedef {Object} ValidationRule
 * @property {string} [type] - Expected type (boolean, string, number, array, object)
 * @property {*} [const] - Single valid value
 * @property {Array} [enum] - Array of valid values
 * @property {string} [pattern] - Regex pattern the value must match
 * @property {*} [default] - Default value (for documentation)
 * @property {Object} [items] - Validation for array items
 * @property {Object} [properties] - Validation for object properties
 */

/**
 * @typedef {Object} ValidationConfig
 * @property {string[]} [required] - Array of required property paths
 * @property {Object<string, ValidationRule>} [properties] - Property validation rules
 */

/**
 * @typedef {Object} ValidationResult
 * @property {boolean} valid - Whether validation passed
 * @property {string} [error] - Error message if validation failed
 */

/**
 * @typedef {Object} ValidationError
 * @property {string} propertyPath - Path to the invalid property
 * @property {string} message - Error message
 * @property {*} value - The invalid value
 * @property {string} [tip] - Helpful tip for fixing the error
 */

/**
 * Get nested property value using dot notation
 * @param {Object} obj - Object to search
 * @param {string} path - Dot notation path (e.g., "containerFields.isAnimated")
 * @returns {*} Property value or undefined
 */
function getNestedProperty(obj, path) {
  if (!obj || typeof obj !== 'object') {
    return undefined;
  }
  return path.split('.').reduce((current, key) => {
    if (current && typeof current === 'object') {
      return current[key];
    }
    return undefined;
  }, obj);
}

/**
 * Check if property exists using dot notation
 * @param {Object} obj - Object to search
 * @param {string} path - Dot notation path
 * @returns {boolean} Whether property exists
 */
function hasNestedProperty(obj, path) {
  if (!obj || typeof obj !== 'object') {
    return false;
  }
  const keys = path.split('.');
  let current = obj;
  for (const key of keys) {
    if (!current || typeof current !== 'object' || !(key in current)) {
      return false;
    }
    current = current[key];
  }
  return true;
}

/**
 * Generate helpful tip for common validation errors
 * @param {*} value - The invalid value
 * @param {ValidationRule} rule - Validation rule
 * @returns {string|null} Helpful tip or null
 */
function generateTip(value, rule) {
  // String "false" or "true" when boolean expected
  if (rule.type === 'boolean' && typeof value === 'string') {
    if (value.toLowerCase() === 'false' || value.toLowerCase() === 'true') {
      return `String "${value}" evaluates to true in templates. Use boolean ${value.toLowerCase()} instead.`;
    }
  }

  // String number when number expected
  if (rule.type === 'number' && typeof value === 'string' && !isNaN(Number(value))) {
    return `String "${value}" should be a number. Remove quotes: ${value}`;
  }

  // Common misspellings for titleTag
  if (rule.enum && rule.enum.includes('h1') && typeof value === 'string') {
    const commonMisspellings = {
      'header': 'h1, h2, h3, h4, h5, or h6',
      'heading': 'h1, h2, h3, h4, h5, or h6',
      'title': 'h1, h2, h3, h4, h5, or h6'
    };
    if (commonMisspellings[value.toLowerCase()]) {
      return `Use ${commonMisspellings[value.toLowerCase()]} instead of "${value}".`;
    }
  }
  return null;
}

/**
 * Validate type constraint
 * @param {*} value - Value to validate
 * @param {ValidationRule} rule - Validation rule
 * @param {string} propertyPath - Property path for error messages
 * @returns {ValidationResult} Validation result
 */
function validateTypeConstraint(value, rule, propertyPath) {
  if (!rule.type) {
    return {
      valid: true
    };
  }
  const actualType = Array.isArray(value) ? 'array' : typeof value;
  if (actualType !== rule.type) {
    const tip = generateTip(value, rule);
    return {
      valid: false,
      error: `${propertyPath}: expected ${rule.type}, got ${actualType} "${value}"${tip ? `\nTip: ${tip}` : ''}`
    };
  }
  return {
    valid: true
  };
}

/**
 * Validate const constraint
 * @param {*} value - Value to validate
 * @param {ValidationRule} rule - Validation rule
 * @param {string} propertyPath - Property path for error messages
 * @returns {ValidationResult} Validation result
 */
function validateConstConstraint(value, rule, propertyPath) {
  if (rule.const === undefined) {
    return {
      valid: true
    };
  }
  if (value !== rule.const) {
    return {
      valid: false,
      error: `${propertyPath}: expected "${rule.const}", got "${value}"`
    };
  }
  return {
    valid: true
  };
}

/**
 * Validate enum constraint
 * @param {*} value - Value to validate
 * @param {ValidationRule} rule - Validation rule
 * @param {string} propertyPath - Property path for error messages
 * @returns {ValidationResult} Validation result
 */
function validateEnumConstraint(value, rule, propertyPath) {
  if (!rule.enum) {
    return {
      valid: true
    };
  }
  if (!rule.enum.includes(value)) {
    const tip = generateTip(value, rule);
    return {
      valid: false,
      error: `${propertyPath}: "${value}" is invalid. Must be one of: ${rule.enum.join(', ')}${tip ? `\nTip: ${tip}` : ''}`
    };
  }
  return {
    valid: true
  };
}

/**
 * Validate pattern constraint
 * @param {*} value - Value to validate
 * @param {ValidationRule} rule - Validation rule
 * @param {string} propertyPath - Property path for error messages
 * @returns {ValidationResult} Validation result
 */
function validatePatternConstraint(value, rule, propertyPath) {
  if (!rule.pattern) {
    return {
      valid: true
    };
  }
  if (typeof value !== 'string') {
    return {
      valid: false,
      error: `${propertyPath}: expected string for pattern validation, got ${typeof value}`
    };
  }
  const regex = new RegExp(rule.pattern);
  if (!regex.test(value)) {
    return {
      valid: false,
      error: `${propertyPath}: "${value}" does not match pattern ${rule.pattern}`
    };
  }
  return {
    valid: true
  };
}

/**
 * Validate array items
 * @param {Array} value - Array to validate
 * @param {ValidationRule} rule - Validation rule
 * @param {string} propertyPath - Property path for error messages
 * @returns {ValidationResult} Validation result
 */
function validateArrayItems(value, rule, propertyPath) {
  if (rule.type !== 'array' || !rule.items || !Array.isArray(value)) {
    return {
      valid: true
    };
  }
  for (let i = 0; i < value.length; i++) {
    const item = value[i];
    const itemPath = `${propertyPath}[${i}]`;
    if (rule.items.properties) {
      const itemValidation = validateObjectProperties(item, rule.items.properties, itemPath);
      if (!itemValidation.valid) {
        return itemValidation;
      }
    }

    // If items has type/enum/const rules, validate the item directly
    if (rule.items.type || rule.items.enum || rule.items.const !== undefined) {
      const itemResult = validateProperty(item, rule.items, itemPath);
      if (!itemResult.valid) {
        return itemResult;
      }
    }
  }
  return {
    valid: true
  };
}

/**
 * Validate a single property against its validation rule
 * @param {*} value - Property value to validate
 * @param {ValidationRule} rule - Validation rule
 * @param {string} propertyPath - Property path for error messages
 * @returns {ValidationResult} Validation result
 */
function validateProperty(value, rule, propertyPath) {
  if (value === undefined || value === null) {
    return {
      valid: true
    }; // Optional properties
  }

  // Run all validation constraints
  const typeResult = validateTypeConstraint(value, rule, propertyPath);
  if (!typeResult.valid) {
    return typeResult;
  }
  const constResult = validateConstConstraint(value, rule, propertyPath);
  if (!constResult.valid) {
    return constResult;
  }
  const enumResult = validateEnumConstraint(value, rule, propertyPath);
  if (!enumResult.valid) {
    return enumResult;
  }
  const patternResult = validatePatternConstraint(value, rule, propertyPath);
  if (!patternResult.valid) {
    return patternResult;
  }
  const arrayResult = validateArrayItems(value, rule, propertyPath);
  if (!arrayResult.valid) {
    return arrayResult;
  }
  return {
    valid: true
  };
}

/**
 * Validate object properties against validation rules
 * @param {Object} obj - Object to validate
 * @param {Object<string, ValidationRule>} properties - Property validation rules
 * @param {string} [basePath] - Base path for nested error messages
 * @returns {ValidationResult} Validation result
 */
function validateObjectProperties(obj, properties, basePath = '') {
  for (const [propertyPath, rule] of Object.entries(properties)) {
    const fullPath = basePath ? `${basePath}.${propertyPath}` : propertyPath;
    const value = getNestedProperty(obj, propertyPath);
    const result = validateProperty(value, rule, fullPath);
    if (!result.valid) {
      return result;
    }
  }
  return {
    valid: true
  };
}

/**
 * Validate required properties
 * @param {Object} obj - Object to validate
 * @param {string[]} required - Array of required property paths
 * @returns {ValidationResult} Validation result
 */
function validateRequiredProperties(obj, required) {
  for (const propertyPath of required) {
    if (!hasNestedProperty(obj, propertyPath)) {
      return {
        valid: false,
        error: `${propertyPath}: required property is missing`
      };
    }
  }
  return {
    valid: true
  };
}

/**
 * Validate a section against its component validation rules
 * @param {Object} section - Section data to validate
 * @param {ValidationConfig} validation - Validation configuration
 * @param {string} [context] - Context information for error messages
 * @returns {ValidationResult} Validation result
 */
function validateSection(section, validation, context = '') {
  if (!validation || typeof validation !== 'object') {
    return {
      valid: true
    }; // No validation rules
  }

  // Validate required properties
  if (validation.required && Array.isArray(validation.required)) {
    const requiredResult = validateRequiredProperties(section, validation.required);
    if (!requiredResult.valid) {
      const contextPrefix = context ? `${context}\n  ` : '';
      return {
        valid: false,
        error: `${contextPrefix}${requiredResult.error}`
      };
    }
  }

  // Validate properties
  if (validation.properties) {
    const propertiesResult = validateObjectProperties(section, validation.properties);
    if (!propertiesResult.valid) {
      const contextPrefix = context ? `${context}\n  ` : '';
      return {
        valid: false,
        error: `${contextPrefix}${propertiesResult.error}`
      };
    }
  }
  return {
    valid: true
  };
}

/**
 * Validate multiple sections with context information
 * @param {Array} sections - Array of sections to validate
 * @param {Function} getManifest - Function to get manifest for a section type
 * @param {string} [fileName] - File name for error context
 * @returns {Array<ValidationError>} Array of validation errors
 */
function validateSections(sections, getManifest, fileName = '') {
  const errors = [];
  if (!Array.isArray(sections)) {
    return errors;
  }
  sections.forEach((section, index) => {
    if (!section || typeof section !== 'object') {
      return;
    }
    const sectionType = section.sectionType;
    if (!sectionType) {
      return; // Skip sections without sectionType
    }
    try {
      const manifest = getManifest(sectionType);
      if (!manifest || !manifest.validation) {
        return; // No validation rules for this component
      }
      const context = fileName ? `Section ${index} (${sectionType}) in ${fileName}:` : `Section ${index} (${sectionType}):`;
      const result = validateSection(section, manifest.validation, context);
      if (!result.valid) {
        errors.push({
          propertyPath: `sections[${index}]`,
          message: result.error,
          value: section,
          sectionType,
          sectionIndex: index,
          fileName
        });
      }
    } catch (error) {
      // Skip validation if manifest loading fails
      console.warn(`Warning: Could not load manifest for section type "${sectionType}": ${error.message}`);
    }
  });
  return errors;
}

/**
 * Template Parser - Detects component imports in Nunjucks templates
 *
 * Parses Nunjucks template files to detect which components are actually used
 * via {% from "..." import ... %} statements. This enables tree-shaking of
 * unused components for optimal bundle sizes.
 */

/**
 * Regular expression to match Nunjucks import statements
 * Matches: {% from "components/_partials/button/button.njk" import button %}
 * Also matches compact syntax: {%from"path"import name%}
 * Captures: The full path to the component template
 */
const IMPORT_PATTERN = /\{%\s*from\s*["']([^"']+)["']\s*import\s+[^%]+\s*%\}/g;

/**
 * Regular expression to match Nunjucks include statements
 * Matches: {% include "components/sections/header/header.njk" %}
 * Also matches compact syntax: {%include"path"%}
 * Captures: The full path to the included template
 */
const INCLUDE_PATTERN = /\{%\s*include\s*["']([^"']+)["']\s*%\}/g;

/**
 * Extract component name from a Nunjucks import path
 *
 * Examples:
 *   "components/_partials/button/button.njk" → "button"
 *   "components/sections/hero/hero.njk" → "hero"
 *   "_partials/ctas/ctas.njk" → "ctas"
 *
 * @param {string} importPath - The path from the {% from "..." %} statement
 * @param {string[]} componentDirs - Component directory names to look for (_partials, sections)
 * @returns {string|null} - Component name or null if not a component import
 */
function extractComponentName(importPath, componentDirs) {
  // Split path into segments
  const segments = importPath.split('/');

  // Look for component directory markers (_partials, sections, etc.)
  for (let i = 0; i < segments.length; i++) {
    if (componentDirs.includes(segments[i])) {
      // Component name is the next segment after the directory marker
      if (i + 1 < segments.length) {
        return segments[i + 1];
      }
    }
  }
  return null;
}

/**
 * Parse a single template file for component imports
 *
 * @param {string} fileContent - Template file contents as string
 * @param {string[]} componentDirs - Component directory names (_partials, sections)
 * @returns {Set<string>} - Set of component names imported in this file
 */
function parseTemplateFile(fileContent, componentDirs) {
  const importedComponents = new Set();
  let match;

  // Check {% from "..." import ... %} statements
  IMPORT_PATTERN.lastIndex = 0;
  while ((match = IMPORT_PATTERN.exec(fileContent)) !== null) {
    const importPath = match[1];
    const componentName = extractComponentName(importPath, componentDirs);
    if (componentName) {
      importedComponents.add(componentName);
    }
  }

  // Check {% include "..." %} statements
  INCLUDE_PATTERN.lastIndex = 0;
  while ((match = INCLUDE_PATTERN.exec(fileContent)) !== null) {
    const includePath = match[1];
    const componentName = extractComponentName(includePath, componentDirs);
    if (componentName) {
      importedComponents.add(componentName);
    }
  }
  return importedComponents;
}

/**
 * Recursively walk any value (object or array) and collect all sectionType strings.
 *
 * This handles deeply nested component references such as tab panes, carousel slides,
 * accordion panels, or any other wrapper component that embeds sub-components via a
 * sectionType property at an arbitrary depth in the frontmatter tree.
 *
 * @param {*} value - Any frontmatter value to inspect
 * @param {Set<string>} result - Accumulator set for discovered sectionType values
 */
function collectSectionTypes(value, result) {
  if (!value || typeof value !== 'object') {
    return;
  }
  if (Array.isArray(value)) {
    value.forEach(item => collectSectionTypes(item, result));
    return;
  }
  if (value.sectionType && typeof value.sectionType === 'string') {
    result.add(value.sectionType);
  }
  Object.values(value).forEach(child => collectSectionTypes(child, result));
}

/**
 * Detect all used components across all template files and layout files
 *
 * Scans Metalsmith files object and layout directory for:
 * 1. Components in frontmatter — any sectionType value at any nesting depth
 * 2. Component imports in Nunjucks {% from "..." import ... %} statements
 * 3. Components in layout files via {% include "..." %} and {% from "..." import ... %}
 *
 * @param {Object} files - Metalsmith files object
 * @param {string[]} componentDirs - Component directory names (e.g., ['_partials', 'sections'])
 * @param {string|null} layoutDir - Path to layouts directory for scanning (or null to skip)
 * @returns {Set<string>} - Set of all component names used in templates
 */
function detectUsedComponents(files, componentDirs, layoutDir) {
  const allUsedComponents = new Set();

  // Process all template files in Metalsmith files object
  Object.keys(files).forEach(filepath => {
    // Only process template files
    if (!filepath.endsWith('.njk') && !filepath.endsWith('.html')) {
      return;
    }
    const file = files[filepath];

    // Walk the entire file frontmatter and collect every sectionType value.
    // This covers top-level sections as well as sub-components nested at any
    // depth (e.g. tab panes, carousel slides, accordion panels).
    collectSectionTypes(file, allUsedComponents);

    // Skip template parsing if no contents
    if (!file.contents) {
      return;
    }

    // Get file contents as string
    const content = Buffer.isBuffer(file.contents) ? file.contents.toString('utf8') : String(file.contents);

    // Parse Nunjucks imports and add components to the set
    const fileComponents = parseTemplateFile(content, componentDirs);
    fileComponents.forEach(component => allUsedComponents.add(component));
  });

  // Also scan layout files if layoutDir is provided
  if (layoutDir) {
    const layoutComponents = scanLayoutFiles(layoutDir, componentDirs);
    layoutComponents.forEach(component => allUsedComponents.add(component));
  }
  return allUsedComponents;
}

/**
 * Recursively scan layout directory for .njk template files and detect components
 * @param {string} layoutDir - Path to the layout directory
 * @param {string[]} componentDirs - Component directory names (_partials, sections)
 * @returns {Set<string>} - Set of component names found in layout files
 */
function scanLayoutFiles(layoutDir, componentDirs) {
  const components = new Set();

  // Check if directory exists
  if (!fs__default["default"].existsSync(layoutDir)) {
    return components;
  }
  function scanDirectory(dir) {
    const entries = fs__default["default"].readdirSync(dir, {
      withFileTypes: true
    });
    for (const entry of entries) {
      const fullPath = path__default["default"].join(dir, entry.name);
      if (entry.isDirectory()) {
        // Recursively scan subdirectories
        scanDirectory(fullPath);
      } else if (entry.isFile() && entry.name.endsWith('.njk')) {
        // Parse .njk template files
        try {
          const content = fs__default["default"].readFileSync(fullPath, 'utf8');
          const fileComponents = parseTemplateFile(content, componentDirs);
          fileComponents.forEach(comp => components.add(comp));
        } catch {
          // Silently skip files that can't be read
        }
      }
    }
  }
  scanDirectory(layoutDir);
  return components;
}

/**
 * Dependency Resolver - Resolves transitive component dependencies
 *
 * Given a set of directly used components, this module finds all transitively
 * required components by following the `requires` arrays in component manifests.
 *
 * Since components use IIFEs (JS) and are namespaced (CSS), load order doesn't
 * matter - we just need to ensure all required components are included.
 */

/**
 * Resolve all transitive dependencies for a set of components
 *
 * Starting with a set of directly used components, follows the dependency
 * chain to find all components that must be included. Uses breadth-first
 * traversal to collect all dependencies.
 *
 * @param {Set<string>} usedComponents - Set of component names directly used in templates
 * @param {Map<string, Object>} componentMap - Map of all available components
 * @returns {Set<string>} - Set of all components needed (used + dependencies)
 *
 * @example
 * // If "hero" requires "button" and "button" requires "icon"
 * // Input: new Set(['hero'])
 * // Output: new Set(['hero', 'button', 'icon'])
 */
function resolveAllDependencies(usedComponents, componentMap) {
  const resolved = new Set(usedComponents);
  const queue = [...usedComponents];
  while (queue.length > 0) {
    const currentName = queue.shift();
    const component = componentMap.get(currentName);

    // Skip if component not found (will be caught by validation later)
    if (!component) {
      continue;
    }

    // Get requirements (supports both 'requires' and legacy 'dependencies')
    const requirements = component.requires || component.dependencies || [];
    requirements.forEach(requiredName => {
      if (!resolved.has(requiredName)) {
        resolved.add(requiredName);
        queue.push(requiredName);
      }
    });
  }
  return resolved;
}

/**
 * Filter component list to only include needed components
 *
 * Takes a list of all available components and returns only those that
 * are needed (either used directly or required transitively).
 *
 * @param {Array<Object>} allComponents - Array of all available components
 * @param {Set<string>} neededComponents - Set of component names to include
 * @returns {Array<Object>} - Filtered array of component objects
 */
function filterNeededComponents(allComponents, neededComponents) {
  return allComponents.filter(component => neededComponents.has(component.name));
}

/**
 * Component Helper Utilities
 *
 * Small utility functions for working with component maps and manifests
 */

/**
 * Helper function to get component manifest for validation
 * @param {Map} componentMap - Map of component names to component objects
 * @param {string} sectionType - The section type to look up
 * @returns {Object} Component manifest object
 * @throws {Error} If component is not found
 */
function getManifest(componentMap, sectionType) {
  const component = componentMap.get(sectionType);
  if (!component) {
    throw new Error(`Component "${sectionType}" not found`);
  }
  return component;
}

/**
 * @typedef {Object} BundledAssets
 * @property {string|null} css - Bundled CSS content or null if no CSS
 * @property {string|null} js - Bundled JavaScript content or null if no JS
 */

/**
 * @typedef {import('./utils/options.js').BundledComponentsOptions} Options
 */

/**
 * A Metalsmith plugin that automatically discovers and bundles CSS and JavaScript files
 * from component-based architectures using esbuild. All component styles and scripts are
 * merged into the main CSS and JS files, creating a single bundle for each asset type.
 * Processing order: main entries → base components → section components.
 * Includes PostCSS support via esbuild plugins and validation for component properties.
 *
 * TWO-PHASE PLUGIN PATTERN:
 * Phase 1: Configuration phase - Normalizes and validates options once at plugin registration
 * Phase 2: Execution phase - Processes files during Metalsmith build (can run multiple times)
 *
 * Benefits:
 * - Options are validated once, not on every build
 * - Expensive setup operations (like parsing PostCSS config) happen once
 * - The returned plugin function is lightweight and optimized for repeated execution
 * - Supports Metalsmith's watch mode where plugin may be called multiple times
 *
 * @param {Options} [options] - Plugin options
 * @returns {import('metalsmith').Plugin} - Metalsmith plugin function
 */
function bundledComponents(options = {}) {
  // PHASE 1: Configuration - Normalize options once at plugin registration time
  // This runs only once when .use(bundledComponents(options)) is called
  options = normalizeOptions(options);

  /**
   * PHASE 2: Execution - The actual plugin function that processes files
   * This runs during the Metalsmith build and may be called multiple times in watch mode
   *
   * @param {Object} files - Metalsmith files object
   * @param {import('metalsmith').Metalsmith} metalsmith - Metalsmith instance
   * @param {Function} done - Callback function for async completion
   */
  function plugin(files, metalsmith, done) {
    const debug = metalsmith.debug('metalsmith-bundled-components');
    debug('Running with options: %O', options);
    async function processComponents() {
      /*
       * Component Discovery
       */
      // Use metalsmith.path() for consistent cross-platform path handling
      const basePath = metalsmith.path(options.basePath);
      const sectionsPath = metalsmith.path(options.sectionsPath);
      const layoutPath = metalsmith.path(options.layoutsPath);

      // Scans lib folder for all available components in
      // _partials and sections folders
      const allBaseComponents = collectComponents(basePath);
      const allSectionComponents = collectComponents(sectionsPath);
      const allComponents = [...allBaseComponents, ...allSectionComponents];
      debug('Partials path: %s', basePath);
      debug('Sections path: %s', sectionsPath);
      debug('Found all partials: %O', allBaseComponents.map(c => c.name));
      debug('Found all sections: %O', allSectionComponents.map(c => c.name));

      // Fail fast if no components found - this indicates a configuration error
      if (allComponents.length === 0) {
        const error = new Error(`No components found in specified directories.
  - basePath: ${options.basePath}
  - sectionsPath: ${options.sectionsPath}

This likely indicates a configuration error. Please verify:
  1. Component directories exist and contain components
  2. Paths are correct relative to project root
  3. Component folders follow expected structure (component-name/component-name.njk)`);
        throw error;
      }

      /*
       * Template Analysis - Tree Shaking
       * Detect which components are actually used in templates
       * (from page frontmatter, template content, and layout files)
       */

      // Extract directory names from paths, for example 'lib/components/_partials' → '_partials'
      const componentDirs = [path__default["default"].basename(options.basePath), path__default["default"].basename(options.sectionsPath)];

      // Scan all templates and layout files that are actually used
      const usedComponents = detectUsedComponents(files, componentDirs, layoutPath);
      debug('All components used (pages + layouts): %O', [...usedComponents]);

      // Create component map for available components
      const componentMap = createComponentMap(allComponents);
      debug('Component map created with %d available components', componentMap.size);

      /*
       * Resolve transitive dependencies
       */
      const neededComponents = resolveAllDependencies(usedComponents, componentMap);
      debug('Components needed (including dependencies): %O', [...neededComponents]);

      // Filter to only needed components
      const baseComponents = filterNeededComponents(allBaseComponents, neededComponents);
      const sectionComponents = filterNeededComponents(allSectionComponents, neededComponents);
      debug('Filtered partials to bundle: %O', baseComponents.map(c => c.name));
      debug('Filtered sections to bundle: %O', sectionComponents.map(c => c.name));

      /*
       * Validate section data if validation is enabled
       * Validates component data from page frontmatter against validation schemas 
       * defined in component manifest files.
       */
      if (options.validation.enabled) {
        debug('Validating section data...');
        const allValidationErrors = [];

        // Only validate content files (HTML/Markdown) that might contain sections 
        // frontmatter. Skip assets like images, CSS, JS files.
        Object.keys(files).filter(fileName => fileName.endsWith('.html') || fileName.endsWith('.htm') || fileName.endsWith('.md') || fileName.endsWith('.markdown')).forEach(fileName => {
          const file = files[fileName];
          const {
            contents,
            sections
          } = file;

          // Validate file.contents is a Buffer
          if (contents && !Buffer.isBuffer(contents)) {
            debug('Warning: file.contents is not a Buffer for %s', fileName);
          }

          // Collect all validation errors across all files. Allows reporting
          // multiple errors at once rather than failing on the first error.
          if (sections && Array.isArray(sections)) {
            const fileErrors = validateSections(sections, sectionType => getManifest(componentMap, sectionType), fileName);
            allValidationErrors.push(...fileErrors);
          }
        });

        // Handle validation errors
        if (allValidationErrors.length > 0) {
          const errorMessage = `Section Validation Errors:\n\n${allValidationErrors.map(error => `  ${error.message}`).join('\n\n')}`;
          console.error(errorMessage);
          if (options.validation.strict) {
            throw new Error('Section validation failed');
          } else {
            console.warn('\nValidation errors found but continuing build (strict mode disabled)');
            debug('Validation errors: %O', allValidationErrors);
          }
        } else {
          debug('All sections validated successfully');
        }
      }

      // Validate all required components exist (replaces complex dependency ordering)
      const requirementErrors = validateRequirements(componentMap);
      if (requirementErrors.length > 0) {
        console.error('Component requirement errors found:');
        requirementErrors.forEach(error => console.error(`  - ${error}`));
        throw new Error('Component requirement validation failed');
      }

      /*
       * Simple build order: base → sections (in filesystem discovery order)
       * No dependency resolution needed since components are namespaced (CSS)
       * and wrapped in IIFEs (JS), so load order doesn't affect functionality
       */
      const buildOrder = [...baseComponents.map(c => c.name), ...sectionComponents.map(c => c.name)];
      debug('Build order: %s', buildOrder.join(' → '));

      // PostCSS is handled via esbuild plugin
      if (options.postcss && options.postcss.enabled) {
        var _options$postcss$plug;
        debug('PostCSS processing enabled with %d plugins', ((_options$postcss$plug = options.postcss.plugins) == null ? void 0 : _options$postcss$plug.length) || 0);
      }

      /*
       * Bundle components and main entries using esbuild with plugins
       * Process in order: main entries → base components → section components
       */
      debug('Starting bundling process...');
      const bundledAssets = await bundleWithESBuild(baseComponents, sectionComponents, metalsmith.directory(), options);
      debug('Bundled assets completed: %O', {
        hasCss: !!bundledAssets.css,
        hasJs: !!bundledAssets.js
      });

      // Add bundled CSS (main + components) to Metalsmith files object
      // Instead of writing files directly to disk with fs.writeFileSync(),
      //  add them to the files object. Metalsmith then:
      // - Handles the actual writing to disk
      // - Allows other plugins to process these files further
      if (bundledAssets.css) {
        files[options.cssDest] = {
          contents: Buffer.isBuffer(bundledAssets.css) ? bundledAssets.css : Buffer.from(bundledAssets.css, 'utf8')
        };
        debug(`Added bundled CSS to ${options.cssDest}`);
      }

      // Add bundled JS (main + components) to Metalsmith files object
      if (bundledAssets.js) {
        files[options.jsDest] = {
          contents: Buffer.isBuffer(bundledAssets.js) ? bundledAssets.js : Buffer.from(bundledAssets.js, 'utf8')
        };
        debug(`Added bundled JS to ${options.jsDest}`);
      }
    }

    // Execute async processing and handle callback
    processComponents().then(() => done()).catch(error => done(error));
  }

  // Set function name for better debugging
  Object.defineProperty(plugin, 'name', {
    value: 'bundledComponents'
  });
  return plugin;
}

module.exports = bundledComponents;
