// metalsmith-search/lib/index.js
import * as cheerio from 'cheerio';

/**
 * Configuration utilities for plugin options:
 * deep merging, file pattern matching, and normalization.
 * Uses Metalsmith's native match() method to avoid external dependencies.
 */

/**
 * Default plugin options
 * All properties guaranteed to exist after merging with user options.
 * @type {Object}
 * @property {string} pattern - Files to process (HTML files)
 * @property {string[]} ignore - Files to exclude from processing
 * @property {string} indexPath - Output path for search index
 * @property {string[]} excludeSelectors - CSS selectors to exclude from content
 * @property {Object} fuseOptions - Fuse.js search configuration
 */
const defaultOptions = {
  pattern: '**/*.html',
  ignore: ['**/search-index.json'],
  indexPath: 'search-index.json',
  excludeSelectors: ['nav', 'header', 'footer'],
  // Fuse.js options
  fuseOptions: {
    keys: [{
      name: 'title',
      weight: 10
    },
    // Page title from <title> or <h1>
    {
      name: 'content',
      weight: 5
    },
    // All page text content
    {
      name: 'excerpt',
      weight: 3
    } // Auto-generated excerpt
    ],
    threshold: 0.3,
    includeScore: true,
    includeMatches: true,
    minMatchCharLength: 3 // Skip stop words (to, be, or, etc.)
  }
};

/**
 * Deep merge configuration objects without mutation
 * Recursively merges nested objects using reduce() and spread operators.
 * @param {Object} target - Base configuration object
 * @param {Object} source - Override configuration object
 * @returns {Object} Merged configuration
 */
const deepMerge = (target, source) => Object.keys(source).reduce((acc, key) => {
  var _source$key;
  return {
    ...acc,
    // If source value is a plain object, recursively merge it
    // Otherwise, use the source value directly (overwrites target)
    [key]: ((_source$key = source[key]) == null ? void 0 : _source$key.constructor) === Object ? deepMerge(target[key] || {}, source[key]) : source[key]
  };
}, {
  ...target
} // Start with a copy of target to avoid mutation
);

/**
 * Convert string or invalid value to array
 * @param {*} value - Value to normalize
 * @returns {Array} Array version of value
 */
function normalizeToArray(value) {
  if (typeof value === 'string') {
    return [value];
  }
  if (Array.isArray(value)) {
    return value;
  }
  return [];
}

/**
 * Normalize options by converting string values to arrays
 * @param {Object} options - Merged options (defaults + user overrides)
 * @returns {Object} Options with arrays for pattern, ignore, and excludeSelectors
 */
function normalizeOptions(options) {
  return {
    ...options,
    pattern: normalizeToArray(options.pattern),
    ignore: normalizeToArray(options.ignore),
    excludeSelectors: normalizeToArray(options.excludeSelectors)
  };
}

/**
 * Check if ignore patterns exist
 * @param {string[]} ignore - Array of ignore patterns
 * @returns {boolean} True if ignore patterns are present
 */
function hasIgnorePatterns(ignore) {
  return ignore.length > 0;
}

/**
 * Remove ignored files from matched files
 * @param {string[]} matchedFiles - Files matching pattern
 * @param {string[]} ignoredFiles - Files matching ignore patterns
 * @returns {string[]} Files to process (matched minus ignored)
 */
function filterIgnoredFiles(matchedFiles, ignoredFiles) {
  return matchedFiles.filter(filename => !ignoredFiles.includes(filename));
}

/**
 * Get files matching pattern while excluding ignored files
 * Uses Metalsmith's native match() method for consistency with other plugins.
 * @param {Object} files - Metalsmith files object
 * @param {Object} options - Normalized options with pattern and ignore arrays
 * @param {Object} metalsmith - Metalsmith instance
 * @returns {string[]} Files to process
 */
function validateFiles(files, options, metalsmith) {
  const {
    pattern,
    ignore
  } = options;
  const allFiles = Object.keys(files);
  const matchedFiles = metalsmith.match(pattern, allFiles);

  // Early return: no ignore patterns means no filtering needed
  if (!hasIgnorePatterns(ignore)) {
    return matchedFiles;
  }

  // Filter out ignored files
  const ignoredFiles = metalsmith.match(ignore, allFiles);
  return filterIgnoredFiles(matchedFiles, ignoredFiles);
}

/**
 * Anchor ID generation utility for section linking
 * Creates URL-safe anchor IDs for deep linking to page sections
 */

/**
 * Check if text is valid for anchor generation
 * @param {*} text - Text to validate
 * @returns {boolean} True if valid text input
 */
function isValidTextInput(text) {
  return text && typeof text === 'string';
}

/**
 * Clean text for URL-safe anchor generation
 * @param {string} text - Text to clean
 * @param {string} separator - Separator to use
 * @returns {string} Cleaned text
 */
function cleanTextForAnchor(text, separator) {
  return text.toLowerCase().replace(/<[^>]+>/g, '') // Remove HTML tags
  .replace(/[^\w\s-]/g, '') // Remove special characters
  .replace(/[\s_]+/g, separator) // Replace spaces with separator
  .replace(new RegExp(`${separator}+`, 'g'), separator) // Remove multiple separators
  .replace(new RegExp(`^${separator}+|${separator}+$`, 'g'), ''); // Trim separators
}

/**
 * Remove numbers from text if not allowed
 * @param {string} text - Text to process
 * @param {boolean} allowNumbers - Whether to allow numbers
 * @returns {string} Processed text
 */
function processNumbers(text, allowNumbers) {
  return allowNumbers ? text : text.replace(/\d/g, '');
}

/**
 * Truncate text to maximum length
 * @param {string} text - Text to truncate
 * @param {number} maxLength - Maximum length
 * @param {string} separator - Separator to clean from end
 * @returns {string} Truncated text
 */
function truncateText(text, maxLength, separator) {
  if (text.length <= maxLength) {
    return text;
  }
  return text.substring(0, maxLength).replace(new RegExp(`${separator}+$`), '');
}

/**
 * Add prefix to anchor if specified
 * @param {string} anchor - Base anchor
 * @param {string} prefix - Prefix to add
 * @param {string} separator - Separator to use
 * @returns {string} Anchor with prefix
 */
function addPrefix(anchor, prefix, separator) {
  return prefix ? `${prefix}${separator}${anchor}` : anchor;
}

/**
 * Add suffix to anchor if specified
 * @param {string} anchor - Base anchor
 * @param {string} suffix - Suffix to add
 * @param {string} separator - Separator to use
 * @returns {string} Anchor with suffix
 */
function addSuffix(anchor, suffix, separator) {
  return suffix ? `${anchor}${separator}${suffix}` : anchor;
}

/**
 * Ensure anchor is not empty
 * @param {string} anchor - Anchor to validate
 * @returns {string} Valid anchor (fallback to 'section' if empty)
 */
function ensureValidAnchor(anchor) {
  return anchor || 'section';
}

/**
 * Generate a URL-safe anchor ID from text
 * @param {string} text - Text to convert to anchor ID
 * @param {Object} options - Generation options
 * @returns {string} URL-safe anchor ID
 */
function generateAnchorId(text, options = {}) {
  if (!isValidTextInput(text)) {
    return 'section';
  }
  const config = {
    maxLength: 50,
    prefix: '',
    suffix: '',
    allowNumbers: true,
    separator: '-',
    ...options
  };
  let anchor = cleanTextForAnchor(text, config.separator);
  anchor = processNumbers(anchor, config.allowNumbers);
  anchor = truncateText(anchor, config.maxLength, config.separator);
  anchor = addPrefix(anchor, config.prefix, config.separator);
  anchor = addSuffix(anchor, config.suffix, config.separator);
  return ensureValidAnchor(anchor);
}

/**
 * HTML-first content extraction processor for search indexing
 *
 * Simplified approach:
 * - Extract all text content from the page
 * - Find all headings and ensure they have IDs (generate if missing)
 * - Return single page entry with headings metadata for client-side scroll-to
 */

/**
 * Extract searchable content from HTML file
 *
 * Returns a single search entry per page with:
 * - All text content
 * - Metadata from frontmatter
 * - List of headings with IDs for scroll-to functionality
 *
 * @param {Object} file - Metalsmith file object
 * @param {string} filename - File path
 * @param {Object} options - Processing options
 * @param {Object} metalsmith - Metalsmith instance
 * @returns {Array} Array with single search entry (or empty if no content)
 */
function extractSearchableContent(file, filename, options, metalsmith) {
  const debug = metalsmith.debug('metalsmith-search:extractor');
  try {
    // Validate file input
    if (!file || !Buffer.isBuffer(file.contents)) {
      debug(`Skipping ${filename}: invalid file object`);
      return [];
    }
    const html = file.contents.toString();
    if (!html.trim()) {
      debug(`Skipping ${filename}: empty content`);
      return [];
    }

    // Load HTML with Cheerio
    const $ = cheerio.load(html);

    // Remove excluded selectors (nav, header, footer, etc.)
    if (options.excludeSelectors && options.excludeSelectors.length > 0) {
      $(options.excludeSelectors.join(', ')).remove();
      debug(`Removed excluded selectors: ${options.excludeSelectors.join(', ')}`);
    }

    // Always remove scripts and styles
    $('script, style').remove();

    // Generate base URL for this file
    // Handle root index.html: index.html -> /
    // Handle nested index: foo/index.html -> /foo
    // Handle regular files: foo/bar.html -> /foo/bar
    let baseUrl = filename.replace(/\.html$/, '');
    if (baseUrl === 'index' || baseUrl.endsWith('/index')) {
      baseUrl = baseUrl.replace(/\/?index$/, '') || '/';
    }
    const cleanUrl = baseUrl.startsWith('/') ? baseUrl : `/${baseUrl}`;

    // Determine page title from HTML <title> tag or first <h1>
    const pageTitle = $('title').text().trim() || $('h1').first().text().trim() || 'Untitled';
    debug(`Processing ${filename} (URL: ${cleanUrl}, title: ${pageTitle})`);

    // Extract all headings and ensure they have IDs
    const headings = extractAndProcessHeadings($, debug);

    // Extract all text content
    const mainText = $.text().trim();
    if (!mainText) {
      debug(`Skipping ${filename}: no text content after processing`);
      return [];
    }

    // Create excerpt from content (first 250 chars)
    const excerpt = mainText.length > 300 ? `${mainText.substring(0, 250).replace(/\s+\S*$/, '')}...` : mainText;

    // Create single page entry
    const entry = {
      id: `page:${cleanUrl}`,
      type: 'page',
      url: cleanUrl,
      title: pageTitle,
      content: mainText,
      excerpt,
      headings,
      // Array of {level, id, title} for scroll-to functionality
      wordCount: mainText.split(/\s+/).filter(Boolean).length
    };
    debug(`Extracted page entry with ${headings.length} headings and ${entry.wordCount} words`);
    return [entry];
  } catch (error) {
    debug(`Error extracting content from ${filename}:`, error);
    return [];
  }
}

/**
 * Extract all headings (h1-h6) and ensure they have IDs
 *
 * For headings without IDs, generates URL-safe IDs from the heading text.
 * Returns metadata array for client-side scroll-to functionality.
 *
 * @param {Object} $ - Cheerio instance
 * @param {Function} debug - Debug logging function
 * @returns {Array} Array of {level: 'h2', id: 'section-id', title: 'Section Title'}
 */
function extractAndProcessHeadings($, debug) {
  const headings = [];
  const usedIds = new Set(); // Track used IDs to ensure uniqueness

  $('h1, h2, h3, h4, h5, h6').each((index, el) => {
    const $heading = $(el);
    const level = el.name; // 'h1', 'h2', etc.
    const title = $heading.text().trim();
    if (!title) {
      return; // Skip empty headings
    }

    // Get existing ID or generate one
    let id = $heading.attr('id');
    if (!id) {
      // Generate ID from heading text
      id = generateAnchorId(title);

      // Ensure uniqueness by appending number if needed
      let uniqueId = id;
      let counter = 1;
      while (usedIds.has(uniqueId)) {
        uniqueId = `${id}-${counter}`;
        counter++;
      }
      id = uniqueId;

      // Set the ID on the heading element
      $heading.attr('id', id);
      debug(`Generated ID '${id}' for ${level}: ${title}`);
    }
    usedIds.add(id);
    headings.push({
      level,
      id,
      title
    });
  });
  return headings;
}

/**
 * File Processing Utilities
 * Simplified HTML-first processing with no binary detection or priority sorting
 */

/**
 * Process a single file and extract search entries
 * @param {string} filename - Filename to process
 * @param {Object} files - Metalsmith files object
 * @param {Object} options - Processing options
 * @param {Function} debug - Debug function
 * @param {Object} metalsmith - Metalsmith instance
 * @returns {Promise<Array>} Array of search entries
 */
function processFile(filename, files, options, debug, metalsmith) {
  try {
    const file = files[filename];

    // Extract searchable content from the file
    const fileSearchEntries = extractSearchableContent(file, filename, options, metalsmith);
    debug(`Successfully processed: ${filename} (${fileSearchEntries.length} entries)`);
    return fileSearchEntries;
  } catch (error) {
    debug(`Error processing ${filename}:`, error);
    // Log but continue with other files
    return [];
  }
}

/**
 * Process all files and collect search entries
 * @param {string[]} filesToProcess - Array of filenames to process
 * @param {Object} files - Metalsmith files object
 * @param {Object} options - Processing options
 * @param {Function} debug - Debug function
 * @param {Object} metalsmith - Metalsmith instance
 * @returns {Promise<Array>} All search entries from all files
 */
function processAllFiles(filesToProcess, files, options, debug, metalsmith) {
  const allSearchEntries = [];

  // Process all files synchronously (Cheerio parsing is fast)
  for (const filename of filesToProcess) {
    const entries = processFile(filename, files, options, debug, metalsmith);
    allSearchEntries.push(...entries);
  }
  debug(`Extracted ${allSearchEntries.length} total search entries`);
  return allSearchEntries;
}

/**
 * Search index creation processor
 * Creates optimized search indexes using Fuse.js patterns
 */

/**
 * Create a search index from extracted content entries
 * @param {Array} searchEntries - Array of search entries
 * @param {Object} options - Index creation options
 * @returns {Promise<Object>} Search index object
 */
function createSearchIndex(searchEntries, options) {
  if (!Array.isArray(searchEntries) || searchEntries.length === 0) {
    return createEmptyIndex$1(options);
  }

  // Optimize entries for search
  const optimizedEntries = optimizeEntriesForSearch(searchEntries);

  // Create index structure
  const index = {
    version: '1.0.0',
    generator: 'metalsmith-search',
    generated: new Date().toISOString(),
    totalEntries: optimizedEntries.length,
    // Index configuration for client-side reconstruction
    config: {
      fuseOptions: options.fuseOptions
    },
    // Statistics for debugging and optimization
    stats: generateIndexStats(optimizedEntries),
    // The actual searchable data
    entries: optimizedEntries
  };
  return index;
}

/**
 * Optimize search entries for better search performance
 * @param {Array} entries - Raw search entries
 * @param {Object} options - Optimization options
 * @returns {Array} Optimized entries
 */
function optimizeEntriesForSearch(entries) {
  return entries.map((entry, index) => {
    const optimized = {
      // Unique identifier for each entry
      id: entry.id || `entry-${index}`,
      // Entry type and metadata
      type: entry.type || 'page',
      url: entry.url || '/',
      // Searchable content fields
      title: cleanText(entry.title || ''),
      content: cleanText(entry.content || ''),
      // Additional metadata for enhanced search
      ...(entry.description && {
        description: cleanText(entry.description)
      }),
      ...(entry.excerpt && {
        excerpt: cleanText(entry.excerpt)
      }),
      ...(entry.tags && {
        tags: entry.tags
      }),
      ...(entry.date && {
        date: entry.date
      }),
      ...(entry.author && {
        author: entry.author
      }),
      // Headings array for client-side scroll-to functionality
      // Format: [{level: 'h2', id: 'section-id', title: 'Section Title'}, ...]
      ...(entry.headings && entry.headings.length > 0 && {
        headings: entry.headings
      }),
      // Search relevance (initially 0, will be set by Fuse.js)
      score: 0
    };

    // Remove empty or undefined fields to reduce index size
    return removeEmptyFields(optimized);
  });
}

/**
 * Generate statistics about the search index
 * @param {Array} entries - Search entries
 * @returns {Object} Index statistics
 */
function generateIndexStats(entries) {
  const stats = {
    totalEntries: entries.length,
    entriesByType: {},
    averageContentLength: 0,
    totalContentLength: 0
  };
  let totalLength = 0;
  for (const entry of entries) {
    // Count by entry type
    stats.entriesByType[entry.type] = (stats.entriesByType[entry.type] || 0) + 1;

    // Calculate content length
    const contentLength = (entry.content || '').length;
    totalLength += contentLength;
  }
  stats.totalContentLength = totalLength;
  stats.averageContentLength = entries.length > 0 ? Math.round(totalLength / entries.length) : 0;
  return stats;
}

/**
 * Create an empty search index when no entries are found
 * @param {Object} options - Index options
 * @returns {Object} Empty index structure
 */
function createEmptyIndex$1(options) {
  return {
    version: '1.0.0',
    generator: 'metalsmith-search',
    generated: new Date().toISOString(),
    totalEntries: 0,
    config: {
      fuseOptions: options.fuseOptions
    },
    stats: {
      totalEntries: 0,
      entriesByType: {},
      averageContentLength: 0,
      totalContentLength: 0
    },
    entries: []
  };
}

/**
 * Clean text content for search optimization
 * @param {string} text - Raw text content
 * @returns {string} Cleaned text
 */
function cleanText(text) {
  if (!text || typeof text !== 'string') {
    return '';
  }
  return text.trim()
  // Normalize whitespace
  .replace(/\s+/g, ' ')
  // Remove excessive punctuation
  .replace(/[^\w\s\-.,!?;:'"()]/g, '')
  // Limit length to prevent index bloat (adjust as needed)
  .substring(0, 2000);
}

/**
 * Remove empty or undefined fields from an object
 * @param {Object} obj - Object to clean
 * @returns {Object} Cleaned object
 */
function removeEmptyFields(obj) {
  const cleaned = {};
  for (const [key, value] of Object.entries(obj)) {
    if (value !== null && value !== undefined && value !== '') {
      // Keep arrays even if empty (they might be intentionally empty)
      if (Array.isArray(value) || value) {
        cleaned[key] = value;
      }
    }
  }
  return cleaned;
}

/**
 * Search Index Helper Utilities
 * Functions for creating and managing search indexes
 */

/**
 * Create an empty search index and add it to files
 * @param {Object} files - Metalsmith files object
 * @param {Object} options - Plugin options
 * @param {Function} debug - Debug logging function
 * @returns {Promise<void>}
 */
function createEmptyIndex(files, options, debug) {
  debug('Creating empty search index');
  const emptySearchIndex = createSearchIndex([], options);
  files[options.indexPath] = {
    contents: Buffer.from(JSON.stringify(emptySearchIndex, null, 2)),
    mode: '0644'
  };
  debug(`Created empty search index at ${options.indexPath}`);
}

/**
 * Setup and validate files for processing
 * @param {Object} files - Metalsmith files object
 * @param {Object} options - Plugin options
 * @param {Object} metalsmith - Metalsmith instance
 * @param {Function} debug - Debug logging function
 * @returns {Promise<Object>} Object with setup results
 */
function setupFileProcessing(files, options, metalsmith, debug) {
  // Normalize and validate options
  const normalizedOptions = normalizeOptions(options);

  // Get files that match patterns (trusts pattern option, no binary detection)
  const filesToProcess = validateFiles(files, normalizedOptions, metalsmith);
  if (filesToProcess.length === 0) {
    createEmptyIndex(files, normalizedOptions, debug);
    return {
      shouldExit: true
    };
  }
  debug(`Processing ${filesToProcess.length} matched files`);
  return {
    shouldExit: false,
    normalizedOptions,
    filesToProcess
  };
}

/**
 * Create and save the final search index
 * @param {Array} searchEntries - All collected search entries
 * @param {Object} files - Metalsmith files object
 * @param {Object} options - Plugin options
 * @param {Function} debug - Debug logging function
 * @returns {Promise<void>}
 */
function createAndSaveIndex(searchEntries, files, options, debug) {
  // Create the search index
  const searchIndex = createSearchIndex(searchEntries, options);

  // Add the search index file to Metalsmith files
  files[options.indexPath] = {
    contents: Buffer.from(JSON.stringify(searchIndex, null, 2)),
    mode: '0644'
  };
  debug(`Created search index at ${options.indexPath}`);
}

/**
 * HTML-First Metalsmith Search Plugin
 *
 * Processes final rendered HTML files after layouts/templates for accurate search indexing.
 * Uses Cheerio for fast, accurate HTML parsing instead of expensive RegExp operations.
 *
 * This plugin uses Metalsmith's built-in methods:
 * - metalsmith.match(patterns, files) for pattern matching
 * - metalsmith.debug(namespace) for debug logging
 */

/**
 * Plugin options
 * @typedef {Object} Options
 * @property {string|string[]} [pattern] - HTML files to process (default: '**\/*.html')
 * @property {string|string[]} [ignore] - Files to ignore
 * @property {string} [indexPath] - Output path for search index (default: 'search-index.json')
 * @property {string[]} [excludeSelectors] - CSS selectors to exclude (e.g., ['nav', 'footer'])
 * @property {string[]} [contentFields] - Frontmatter fields to index
 * @property {Object} [fuseOptions] - Fuse.js configuration options
 */

/**
 * HTML-first Metalsmith search plugin with Fuse.js.
 * Processes final rendered HTML using Cheerio for accurate content extraction.
 * Best positioned AFTER layouts/templates in the Metalsmith pipeline.
 *
 * @param {Options} options - Plugin options
 * @returns {import('metalsmith').Plugin} Metalsmith plugin function
 */
function search(options = {}) {
  // Normalize options with defaults
  const config = deepMerge(defaultOptions, options);

  // Return the actual plugin function (two-phase pattern)
  const metalsmithPlugin = function (files, metalsmith, done) {
    const debug = metalsmith.debug('metalsmith-search');
    debug('Starting metalsmith-search with options:', config);
    try {
      // Setup and validate files for processing
      const setup = setupFileProcessing(files, config, metalsmith, debug);
      if (setup.shouldExit) {
        return done();
      }
      const {
        normalizedOptions,
        filesToProcess
      } = setup;

      // Process all files and collect search entries
      const allSearchEntries = processAllFiles(filesToProcess, files, normalizedOptions, debug, metalsmith);

      // Create and save the search index
      createAndSaveIndex(allSearchEntries, files, normalizedOptions, debug);
      debug('metalsmith-search completed successfully');
      done();
    } catch (error) {
      debug('metalsmith-search failed:', error);
      done(error);
    }
  };

  // Set function name for debugging (helps with stack traces and debugging)
  Object.defineProperty(metalsmithPlugin, 'name', {
    value: 'searchPlugin',
    configurable: true
  });
  return metalsmithPlugin;
}

export { search as default };
