// metalsmith-optimize-html/lib/index.cjs
'use strict';

/**
 * Whitespace optimization module
 * Handles complex whitespace optimization in HTML content while preserving
 * necessary formatting and spacing.
 *
 * @module whitespace-optimizer
 *
 * Core functionality:
 * - Collapses multiple whitespace characters to single space
 * - Removes unnecessary whitespace between HTML tags
 * - Preserves exact whitespace in special tags (pre, code, etc.)
 * - Maintains appropriate spacing around inline elements based on context
 * - Creates single-line output except for preserved content
 * - Handles nested inline elements correctly
 *
 * Processing steps:
 * 1. Preserve special tags (pre, code, etc.) with exact whitespace
 * 2. Process inline elements maintaining contextual spacing
 * 3. Handle remaining block-level content
 * 4. Restore preserved content
 */

const whitespaceOptimizer = {
  name: 'whitespace',
  /**
   * Optimizes HTML content by managing whitespace in various contexts
   * @param {string} content - HTML content to optimize
   * @returns {string} Optimized HTML content
   */
  optimize: (content, options = {}) => {
    // Check if content already contains placeholders (indicating double processing)
    const existingPlaceholders = content.match(/___(?:PRESERVE|INLINE)_\d+___/g);
    if (existingPlaceholders) {
      // Return content unchanged to avoid double-processing
      return content;
    }
    // Tags that must maintain exact internal whitespace
    // These are typically code-related or formatting-specific elements
    const preserveTags = ['pre', 'code', 'textarea', 'script', 'style', ...(options.excludeTags || []) // Additional exclusions, like web-components
    ];

    // Inline elements that may affect text flow and spacing
    // These elements should maintain spacing based on their context
    const inlineTags = ['a', 'span', 'em', 'strong', 'b', 'i', 'u', 's',
    // Basic text formatting
    'small', 'mark', 'sub', 'sup',
    // Text presentation
    'time', 'cite', 'abbr',
    // Semantic elements
    'label', 'svg' // UI and graphics elements
    ];

    // Storage for content that must be preserved or processed separately
    const preserved = [];
    const inlineElements = [];

    // STEP 1: Preserve special tags
    // Matches complete tag pairs with their content and preserves them exactly
    let html = content.replace(new RegExp(`(<(${preserveTags.join('|')})[^>]*>[\\s\\S]*?</\\2>)`, 'gi'), match => {
      const index = preserved.length;
      preserved.push(match);
      return `___PRESERVE_${index}___`;
    });

    // STEP 2: Handle inline elements
    // Complex regex pattern to match inline elements with their surrounding context
    const inlinePattern = new RegExp(`(\\s*)` +
    // Leading whitespace
    `(<(${inlineTags.join('|')})[^>]*>)` +
    // Opening tag with attributes
    `([^<]*(?:(?!</\\3)[\\s\\S])*?)` +
    // Content (including nested elements)
    `(</\\3>)` +
    // Closing tag
    `(\\s*)`,
    // Trailing whitespace
    'gi');

    // Process nested inline elements iteratively from inside out
    let lastHtml;
    do {
      lastHtml = html;
      html = html.replace(inlinePattern, (match, beforeSpace, openTag, tagName, content, closeTag, afterSpace) => {
        // Skip processing if content contains placeholder patterns
        // This prevents inline processing from interfering with PRESERVE placeholders
        if (content.match(/___(?:PRESERVE|INLINE|EXCLUDE)_\d+___/)) {
          return match; // Return unchanged
        }

        // Normalize internal whitespace while preserving content
        const normalizedContent = content.replace(/\s+/g, ' ').trim();
        const normalized = `${openTag}${normalizedContent}${closeTag}`;
        const index = inlineElements.length;
        inlineElements.push(normalized);

        // Maintain contextual spacing - keep single space only where space existed
        const leadSpace = beforeSpace.length > 0 ? ' ' : '';
        const trailSpace = afterSpace.length > 0 ? ' ' : '';
        return `${leadSpace}___INLINE_${index}___${trailSpace}`;
      });
    } while (html !== lastHtml); // Continue until all nested elements are processed

    // STEP 3: Process remaining block-level content
    // Split content into tags and text, process text portions
    const blockParts = html.split(/(<\/?[^>]+>)/g);
    html = blockParts.map(part => {
      if (part.startsWith('<')) {
        return part;
      } // Keep tags unchanged
      return part.replace(/\s+/g, ' ').trim(); // Normalize text content
    }).join('').trim();

    // STEP 4: Restore preserved content
    // Use a more robust restoration approach that handles all edge cases
    let result = html;

    // Create a map of all placeholders and their content for efficient lookup
    const placeholderMap = new Map();

    // Add preserved tags to map
    for (let i = 0; i < preserved.length; i++) {
      placeholderMap.set(`___PRESERVE_${i}___`, preserved[i]);
    }

    // Add inline elements to map
    for (let i = 0; i < inlineElements.length; i++) {
      placeholderMap.set(`___INLINE_${i}___`, inlineElements[i]);
    }

    // Replace all placeholders in a single pass using a regex
    // This ensures placeholders are replaced even if the content contains other placeholder patterns
    result = result.replace(/___(?:PRESERVE|INLINE)_\d+___/g, match => {
      const replacement = placeholderMap.get(match);
      if (replacement) {
        return replacement;
      }
      // This shouldn't happen in normal operation
      return match;
    });
    return result;
  }
};

/**
 * Factory functions for creating HTML optimizer objects
 * @module optimizer-factory
 */

/**
 * @typedef {Object} Optimizer
 * @property {string} name - The name of the optimizer
 * @property {Function} optimize - The function that optimizes HTML content
 */

/**
 * Creates a standard HTML optimizer
 *
 * @param {string} name - The optimizer name
 * @param {string} optionFlag - The option flag that enables this optimizer
 * @param {Function} optimizeImpl - The function that implements the optimization
 * @returns {Optimizer} - A standard optimizer object
 */
function createOptimizer(name, optionFlag, optimizeImpl) {
  return {
    name,
    /**
     * Standard optimizer function that checks if it's enabled before proceeding
     *
     * @param {string} content - HTML content to optimize
     * @param {Object} options - Configuration options
     * @returns {string} - Optimized HTML content
     */
    optimize(content, options = {}) {
      // Skip if explicitly disabled
      if (options[optionFlag] === false) {
        return content;
      }

      // Perform optimization
      return optimizeImpl(content, options);
    }
  };
}

/**
 * Creates a regex-based HTML optimizer
 *
 * @param {string} name - The optimizer name
 * @param {string} optionFlag - The option flag that enables this optimizer
 * @param {RegExp} pattern - The regex pattern to match
 * @param {Function} replacer - The replacement function or string
 * @returns {Optimizer} - A regex-based optimizer
 */
function createRegexOptimizer(name, optionFlag, pattern, replacer) {
  return createOptimizer(name, optionFlag, content => content.replace(pattern, replacer));
}

/**
 * Creates a tag-level optimizer that processes HTML tags individually
 *
 * @param {string} name - The optimizer name
 * @param {string} optionFlag - The option flag that enables this optimizer
 * @param {Function} tagProcessor - Function that processes a single tag
 * @returns {Optimizer} - A tag-level optimizer
 */
function createTagOptimizer(name, optionFlag, tagProcessor) {
  // Regex to match HTML tags (excluding closing tags)
  const TAG_REGEX = /<[^>]+>/g;
  return createOptimizer(name, optionFlag, content => {
    return content.replace(TAG_REGEX, tag => {
      // Skip closing tags
      if (tag.startsWith('</')) {
        return tag;
      }

      // Process tag
      return tagProcessor(tag);
    });
  });
}

/**
 * Creates an attribute-level optimizer that processes specific attributes
 *
 * @param {string} name - The optimizer name
 * @param {string} optionFlag - The option flag that enables this optimizer
 * @param {Function} shouldProcessAttribute - Function that determines if an attribute should be processed
 * @param {Function} attributeProcessor - Function that processes an attribute's value
 * @returns {Optimizer} - An attribute-level optimizer
 */
function createAttributeOptimizer(name, optionFlag, shouldProcessAttribute, attributeProcessor) {
  // Regex to match HTML attributes
  const ATTRIBUTE_REGEX = /\s([^\s=]+)(?:=["']([^"']*)["'])?/g;
  return createTagOptimizer(name, optionFlag, tag => {
    return tag.replace(ATTRIBUTE_REGEX, (match, name, value) => {
      // Skip if this attribute shouldn't be processed
      if (!shouldProcessAttribute(name, value)) {
        return match;
      }

      // Process attribute
      return attributeProcessor(match, name, value);
    });
  });
}

/**
 * Comment optimization module
 * Removes HTML comments when enabled
 *
 * @module comment-optimizer
 *
 * Implements a simple HTML comment removal:
 * - Removes all <!-- --> style comments from HTML
 * - Preserves conditional comments by default (can be configured)
 * - Enable with removeComments: true
 */

// Regex to match HTML comments
const COMMENT_REGEX = /<!--[\s\S]*?-->/g;

/**
 * Creates the comment optimizer using the factory
 */
const commentOptimizer = createRegexOptimizer('comment',
// Optimizer name
'removeComments',
// Option flag
COMMENT_REGEX,
// Pattern to match
'' // Replace with empty string
);

/**
 * Empty attributes optimization module
 * Removes empty attributes while preserving specific ones like alt and value
 * @module empty-attributes-optimizer
 *   - Removes attributes with empty values
 *   - Handles both quoted and unquoted empty values
 *   - Enable with removeEmptyAttributes: true
 */

const emptyAttributesOptimizer = {
  name: 'emptyAttributes',
  optimize: (content, {
    removeEmptyAttributes = false
  } = {}) => {
    if (!removeEmptyAttributes) {
      return content;
    }
    const preserveEmpty = ['alt', 'value'];
    return content.replace(/(<[^>]+?)(\s+[^=>\s]+(?:=["']\s*["']))+(?=\s*\/?>)/g, (match, prefix) => {
      // Keep the tag start, process attributes
      return prefix + match.slice(prefix.length).replace(/\s+[^=>\s]+(?:=["']\s*["'])/g, attr => {
        // Check if this is an attribute we should preserve
        const attrName = attr.match(/\s+([^=>\s]+)/)[1];
        return preserveEmpty.includes(attrName) ? attr : '';
      }).trim();
    });
  }
};

/**
 * Boolean attributes optimization module
 *
 * @module boolean-attributes-optimizer
 *
 * Normalizes HTML boolean attributes according to the HTML5 specification:
 * - Converts attribute="attribute" to just attribute (e.g., checked="checked" → checked)
 * - Handles all standard HTML5 boolean attributes
 * - Removes attributes with value="false"
 * - Preserves attributes with non-standard values
 *
 * @example
 * // Input
 * <input type="checkbox" checked="checked" disabled="disabled">
 * <button disabled="false">Button</button>
 *
 * // Output
 * <input type="checkbox" checked disabled>
 * <button>Button</button>
 */

// Complete list of HTML5 boolean attributes
const BOOLEAN_ATTRIBUTES = new Set(['allowfullscreen', 'async', 'autofocus', 'autoplay', 'checked', 'controls', 'default', 'defer', 'disabled', 'formnovalidate', 'hidden', 'ismap', 'itemscope', 'loop', 'multiple', 'muted', 'nomodule', 'novalidate', 'open', 'playsinline', 'readonly', 'required', 'reversed', 'selected', 'truespeed']);

/**
 * Determine if an attribute is a boolean attribute that should be processed
 *
 * @param {string} name - Attribute name
 * @returns {boolean} - True if the attribute should be processed
 */
function isBooleanAttribute(name) {
  return BOOLEAN_ATTRIBUTES.has(name);
}

/**
 * Process a boolean attribute
 *
 * @param {string} match - The original attribute match
 * @param {string} name - The attribute name
 * @param {string} value - The attribute value
 * @returns {string} - The processed attribute
 */
function processBooleanAttribute(match, name, value) {
  // Handle different attribute value cases
  if (value === 'false') {
    // Remove attributes with value="false"
    return '';
  } else if (value === 'true' || value === name || value === '') {
    // Normalize to attribute-only form
    return ` ${name}`;
  }

  // Preserve original for non-standard values
  return match;
}

/**
 * Create the boolean attributes optimizer
 */
const booleanAttributesOptimizer = createAttributeOptimizer('booleanAttributes', 'normalizeBooleanAttributes', isBooleanAttribute, processBooleanAttribute);

/**
 * URL attributes optimization module
 * Cleans up URL values in attributes like href, src, action
 * @module url-attributes-optimizer
 */

const urlAttributesOptimizer = {
  name: 'urlAttributes',
  optimize: (content, {
    cleanUrlAttributes = false
  } = {}) => {
    if (!cleanUrlAttributes) {
      return content;
    }
    const urlAttributes = new Set([
    // Standard attributes
    'href', 'src', 'action', 'srcset', 'data',
    // Meta tag URLs
    'content',
    // SVG attributes
    'xmlns', 'xlink:href']);

    // Process tag by tag
    return content.replace(/<[^>]+>/g, tag => {
      // Don't process closing tags
      if (tag.startsWith('</')) {
        return tag;
      }

      // Clean URL attributes
      return tag.replace(/\s([^\s=]+)=["']([^"']*)["']/g, (match, name, value) => {
        if (!urlAttributes.has(name)) {
          return match;
        }

        // For meta tags, only process URL-specific tags
        if (name === 'content' && !tag.includes('og:url') && !tag.includes('twitter:url')) {
          return match;
        }

        // For xlink:href with #, preserve the value
        if (name === 'xlink:href' && value.startsWith('#')) {
          return ` ${name}="${value.trim()}"`;
        }

        // Clean and normalize URL
        const cleanedValue = value.trim().replace(/\s+/g, '').replace(/^https?:/, '');
        return ` ${name}="${cleanedValue}"`;
      });
    });
  }
};

/**
 * Data attributes optimization module
 * Cleans up HTML5 data-* attributes
 * @module data-attributes-optimizer
 *   - Cleans data-* attribute values
 *   - Normalizes whitespace in data attribute values
 *   - Maintains valid JSON in data attributes
 *   - Enable with cleanDataAttributes: true
 */

const dataAttributesOptimizer = {
  name: 'dataAttributes',
  optimize: (content, {
    cleanDataAttributes = false
  } = {}) => {
    if (!cleanDataAttributes) {
      return content;
    }
    return content.replace(/<[^>]+>/g, tag => {
      if (tag.startsWith('</')) {
        return tag;
      }
      return tag.replace(/\s(data-[^\s=]+)=(["'])([\s\S]*?)\2/g, (match, name, quote, value) => {
        // Remove empty data attributes
        if (!value.trim()) {
          return '';
        }

        // Try to parse and normalize JSON content
        const trimmedValue = value.trim();
        if (trimmedValue.startsWith('{') || trimmedValue.startsWith('[')) {
          try {
            const jsonValue = JSON.parse(trimmedValue);
            // Use single quotes for JSON to avoid escaping
            return ` ${name}='${JSON.stringify(jsonValue)}'`;
          } catch {
            // Not valid JSON, handle as regular string
          }
        }

        // Handle numeric values
        if (!isNaN(trimmedValue) && trimmedValue.length > 0) {
          return ` ${name}="${trimmedValue}"`;
        }

        // Handle boolean values
        if (trimmedValue.toLowerCase() === 'true') {
          return ` ${name}="true"`;
        }
        if (trimmedValue.toLowerCase() === 'false') {
          return ` ${name}="false"`;
        }

        // Regular strings - normalize spaces
        return ` ${name}="${trimmedValue.replace(/\s+/g, ' ')}"`;
      });
    });
  }
};

/**
 * Tag spaces optimization module
 * Removes unnecessary spaces inside HTML tags
 * @module tag-spaces-optimizer
 *   - Removes extra spaces inside HTML tags
 *   - Normalizes spaces between attributes
 *   - Preserves spaces in attribute values
 *   - Enable with removeTagSpaces: true
 */

const tagSpacesOptimizer = {
  /**
   * Optimizer name
   * @type {string}
   */
  name: 'tag-spaces',
  /**
   * Optimizes HTML content by removing extra spaces inside tags
   * - Collapses multiple spaces between attributes to single space
   * - Removes spaces before closing > or />
   * - Preserves spaces inside attribute values
   *
   * @param {string} content - HTML content to optimize
   * @param {Object} options - Optimization options
   * @returns {string} Optimized HTML content
   */
  optimize: content => {
    return content.replace(/<([^>]*)>/g, (match, tagContent) => {
      // Normalize spaces in attribute values
      const normalizedTag = tagContent.replace(/=\s*"/g, '="') // Remove spaces before attribute values
      .replace(/"\s+/g, '" ') // Normalize spaces after attribute values
      .replace(/\s+/g, ' ') // Collapse multiple spaces to single
      .trim(); // Remove leading/trailing spaces

      return `<${normalizedTag}>`;
    });
  }
};

/**
 * Default attributes optimization module
 * Removes default attribute values that browsers assume anyway
 * @module default-attributes-optimizer
 *   - Removes common default attributes:
 *     - script[type="text/javascript"]
 *     - style[type="text/css"]
 *     - link[type="text/css"]
 *     - form[method="get"]
 *     - input[type="text"]
 *   - Enable with removeDefaultAttributes: true
 */

const defaultAttributesOptimizer = {
  name: 'defaultAttributes',
  optimize: (content, {
    removeDefaultAttributes = false
  } = {}) => {
    if (!removeDefaultAttributes) {
      return content;
    }

    // Map of elements to their default attributes
    const defaultAttributes = {
      script: {
        type: 'text/javascript'
      },
      style: {
        type: 'text/css'
      },
      link: {
        type: 'text/css'
      },
      form: {
        method: 'get'
      },
      input: {
        type: 'text'
      }
    };

    // Process tag by tag
    return content.replace(/<[^>]+>/g, tag => {
      // Don't process closing tags
      if (tag.startsWith('</')) {
        return tag;
      }

      // Get tag name
      const tagMatch = tag.match(/^<([^\s>]+)/);
      if (!tagMatch) {
        return tag;
      }
      const tagName = tagMatch[1].toLowerCase();

      // If tag has default attributes defined
      if (defaultAttributes[tagName]) {
        const defaults = defaultAttributes[tagName];
        // Remove each default attribute if present
        Object.entries(defaults).forEach(([attr, value]) => {
          const pattern = new RegExp(`\\s${attr}=["']${value}["']`, 'i');
          tag = tag.replace(pattern, '');
        });
      }
      return tag;
    });
  }
};

/**
 * Doctype optimization module
 * Replaces any doctype declaration with simple HTML5 doctype
 * @module doctype-optimizer
 *   - Replaces any DOCTYPE declaration with simple HTML5 DOCTYPE
 *   - Handles multiple DOCTYPEs in a document
 *   - Always adds DOCTYPE if one existed
 *   - Enable with simplifyDoctype: true
 */

const doctypeOptimizer = {
  name: 'doctype',
  optimize: (content, {
    simplifyDoctype = false
  } = {}) => {
    if (!simplifyDoctype) {
      return content;
    }

    // Remove all existing doctypes
    let result = content.replace(/<!DOCTYPE[^>]*>/gi, '');

    // Add single HTML5 doctype at the start if there was at least one doctype
    if (content.match(/<!DOCTYPE[^>]*>/i)) {
      result = `<!DOCTYPE html>${result}`;
    }
    return result;
  }
};

/**
 * Protocol optimization module
 * Removes http:// and https:// from URLs while preserving external links
 * @module protocols-optimizer
 *   - Removes http:// and https:// from URLs
 *   - Replaces with protocol-relative // URLs
 *   - Preserves protocols in links with rel="external"
 *   - Enable with removeProtocols: true
 */

const protocolsOptimizer = {
  name: 'protocols',
  optimize: (content, {
    removeProtocols = false
  } = {}) => {
    if (!removeProtocols) {
      return content;
    }

    // URL attributes are defined in the regex pattern below

    return content.replace(/<[^>]+>/g, tag => {
      // Skip tags with rel="external"
      if (tag.includes('rel="external"')) {
        return tag;
      }

      // Process URL attributes
      return tag.replace(/(\s(?:href|src|content|action)=["'])(https?:\/\/[^"']+)(["'])/gi, (match, prefix, url, quote) => `${prefix}//${url.replace(/^https?:\/\//, '')}${quote}`);
    });
  }
};

/**
 * Safe attribute quote removal optimizer
 * Removes quotes from HTML attributes when it's safe to do so
 * @module safe-quote-removal-optimizer
 */
const safeQuoteRemovalOptimizer = {
  /**
   * Optimizer name
   * @type {string}
   */
  name: 'safeQuoteRemoval',
  /**
   * Removes quotes from attribute values when safe
   * - Only removes quotes if value contains no special characters
   * - Preserves quotes for values with spaces, brackets, etc.
   * - Preserves quotes for URL values (//example.com)
   * - Processes each HTML tag separately to avoid false matches
   *
   * @param {string} content - HTML content to optimize
   * @param {Object} options - Optimization options
   * @param {boolean} [options.safeRemoveAttributeQuotes=false] - Whether to remove safe quotes
   * @returns {string} Optimized HTML content
   *
   * @example
   * // Safe to remove quotes:
   * <div class="container"> -> <div class=container>
   *
   * // Quotes preserved:
   * <div class="multiple classes">
   * <div data-json='{"key":"value"}'>
   * <a href="//example.com">
   */
  optimize: (content, {
    safeRemoveAttributeQuotes = false
  } = {}) => {
    if (!safeRemoveAttributeQuotes) {
      return content;
    }

    // URL attributes that should keep their quotes
    const urlAttributes = new Set(['href', 'src', 'action', 'content', 'srcset', 'xmlns', 'xlink:href']);
    return content.replace(/<[^>]+>/g, tag => {
      return tag.replace(/(\s)([\w-]+)=(["'])([^"'<>`\s{}()[\]?=]+)\3/g, (match, space, name, quote, value) => {
        // Preserve quotes for URL attributes
        if (urlAttributes.has(name) && (value.startsWith('//') || value.includes(':'))) {
          return match;
        }
        return `${space}${name}=${value}`;
      });
    });
  }
};

/**
 * Optimizer Registry for metalsmith-optimize-html
 *
 * Centralizes the registration and loading of HTML optimizers
 * using a functional approach.
 *
 * @module optimizer-registry
 */

/**
 * @typedef {Object} Optimizer
 * @property {string} name - Unique name of the optimizer
 * @property {Function} optimize - The optimization function
 */

/**
 * Map of option flag names to optimizer instances
 * Using direct imports instead of dynamic imports for better compatibility
 * @type {Record<string, Optimizer>}
 */
const OPTIMIZERS = {
  // Core optimizer - always loaded
  whitespace: whitespaceOptimizer,
  // Optional optimizers - loaded based on options
  removeComments: commentOptimizer,
  removeEmptyAttributes: emptyAttributesOptimizer,
  normalizeBooleanAttributes: booleanAttributesOptimizer,
  cleanUrlAttributes: urlAttributesOptimizer,
  cleanDataAttributes: dataAttributesOptimizer,
  removeTagSpaces: tagSpacesOptimizer,
  removeDefaultAttributes: defaultAttributesOptimizer,
  simplifyDoctype: doctypeOptimizer,
  removeProtocols: protocolsOptimizer,
  safeRemoveAttributeQuotes: safeQuoteRemovalOptimizer
};

// Cache of loaded optimizers
const optimizers = new Map();

/**
 * Get an optimizer by its flag name
 * Now synchronously returns the optimizer from the OPTIMIZERS map
 *
 * @param {string} flag - The flag name
 * @returns {Optimizer|null} - The optimizer or null if not found
 */
function getOptimizer(flag) {
  // Return from cache if already loaded
  if (optimizers.has(flag)) {
    return optimizers.get(flag);
  }

  // Check if this optimizer exists
  if (!OPTIMIZERS[flag]) {
    return null;
  }

  // Get the optimizer and cache it
  const optimizer = OPTIMIZERS[flag];
  optimizers.set(flag, optimizer);
  return optimizer;
}

/**
 * Load optimizers based on configuration
 * Now synchronously builds and returns the optimizers array
 *
 * @param {Object} options - Configuration options
 * @returns {Optimizer[]} Array of initialized optimizers
 */
function loadOptimizers(options) {
  const loadedOptimizers = [];
  try {
    // Always load core whitespace optimizer
    const whitespaceOptimizer = getOptimizer('whitespace');
    if (whitespaceOptimizer) {
      loadedOptimizers.push(whitespaceOptimizer);
    }

    // Load optional optimizers based on options
    // Only load optimizers that are explicitly enabled (truthy), not undefined
    const flagsToLoad = Object.keys(OPTIMIZERS).filter(flag => flag !== 'whitespace' && options[flag] === true); // Skip core optimizer and only load explicitly enabled flags

    // Load optimizers and filter out nulls
    flagsToLoad.forEach(flag => {
      const optimizer = getOptimizer(flag);
      if (optimizer) {
        loadedOptimizers.push(optimizer);
      }
    });
  } catch (error) {
    console.error('Error loading optimizers:', error);
  }
  return loadedOptimizers;
}

/**
 * Clear the optimizer cache - useful for testing
 */
function clearCache() {
  optimizers.clear();
}
var optimizerRegistry = {
  getOptimizer,
  loadOptimizers,
  clearCache
};

/**
 * Configuration validation for metalsmith-optimize-html
 * @module validate-options
 */

/**
 * @typedef {Object} ValidationResult
 * @property {boolean} valid - Whether the options are valid
 * @property {string[]} errors - Array of error messages if invalid
 */

/**
 * Validate that a value is a boolean or undefined
 *
 * @param {any} value - The value to check
 * @param {string} name - Option name for error message
 * @returns {string|null} - Error message or null if valid
 */
function validateBoolean(value, name) {
  if (value === undefined || typeof value === 'boolean') {
    return null;
  }
  return `Option "${name}" must be a boolean, got ${typeof value}: ${value}`;
}

/**
 * Validate that a value is a string or undefined
 *
 * @param {any} value - The value to check
 * @param {string} name - Option name for error message
 * @returns {string|null} - Error message or null if valid
 */
function validateString(value, name) {
  if (value === undefined || typeof value === 'string') {
    return null;
  }
  return `Option "${name}" must be a string, got ${typeof value}: ${value}`;
}

/**
 * Validate that a value is an array of strings or undefined
 *
 * @param {any} value - The value to check
 * @param {string} name - Option name for error message
 * @returns {string|null} - Error message or null if valid
 */
function validateStringArray(value, name) {
  if (value === undefined) {
    return null;
  }
  if (!Array.isArray(value)) {
    return `Option "${name}" must be an array, got ${typeof value}: ${value}`;
  }
  for (let i = 0; i < value.length; i++) {
    const item = value[i];
    if (typeof item !== 'string') {
      return `Option "${name}" must contain only strings, item at index ${i} is ${typeof item}: ${item}`;
    }
  }
  return null;
}

/**
 * Option validators mapped by option name
 * @type {Object.<string, function>}
 */
const VALIDATORS = {
  // Core options
  pattern: validateString,
  excludeTags: validateStringArray,
  aggressive: validateBoolean,
  // Feature options
  removeComments: validateBoolean,
  removeTagSpaces: validateBoolean,
  normalizeBooleanAttributes: validateBoolean,
  cleanUrlAttributes: validateBoolean,
  removeProtocols: validateBoolean,
  removeDefaultAttributes: validateBoolean,
  cleanDataAttributes: validateBoolean,
  simplifyDoctype: validateBoolean,
  safeRemoveAttributeQuotes: validateBoolean,
  removeEmptyAttributes: validateBoolean
};

/**
 * Validate user options against the expected schema
 *
 * @param {Object} options - User options to validate
 * @returns {ValidationResult} - Validation result
 */
function validateOptions(options) {
  const errors = [];

  // Ensure options is an object
  if (!options || typeof options !== 'object' || Array.isArray(options)) {
    return {
      valid: false,
      errors: ['Options must be an object']
    };
  }

  // Check for unknown options
  for (const key of Object.keys(options)) {
    if (!Object.prototype.hasOwnProperty.call(VALIDATORS, key)) {
      errors.push(`Unknown option "${key}"`);
    }
  }

  // Validate each option
  for (const [name, validator] of Object.entries(VALIDATORS)) {
    if (Object.prototype.hasOwnProperty.call(options, name)) {
      const error = validator(options[name], name);
      if (error) {
        errors.push(error);
      }
    }
  }
  return {
    valid: errors.length === 0,
    errors
  };
}

/**
 * Format validation errors into a readable message
 *
 * @param {string[]} errors - Array of error messages
 * @returns {string} - Formatted error message
 */
function formatValidationErrors(errors) {
  return ['Invalid options for metalsmith-optimize-html:', ...errors.map(err => `  - ${err}`)].join('\n');
}
var validator = {
  validateOptions,
  formatValidationErrors
};

/**
 * File filtering utilities for metalsmith-optimize-html
 * @module utils/file-filters
 */

/**
 * Filter files by pattern before processing for better performance
 * @param {import('metalsmith').Files} files - Metalsmith files object
 * @param {string} pattern - Glob pattern to match
 * @param {import('metalsmith')} metalsmith - Metalsmith instance for matching
 * @returns {string[]} Array of matching filenames
 */
function getMatchingFiles(files, pattern, metalsmith) {
  const filenames = Object.keys(files);
  return metalsmith.match(pattern, filenames);
}

/**
 * Check if file has valid contents for processing
 * Validates Buffer type, non-zero length, and basic content structure
 * @param {Object} file - Metalsmith file object
 * @returns {boolean} True if file can be processed
 */
function isProcessableFile(file) {
  // Check if file object exists
  if (!file || typeof file !== 'object') {
    return false;
  }

  // Check if contents property exists and is a Buffer
  if (!file.contents || !Buffer.isBuffer(file.contents)) {
    return false;
  }

  // Check if buffer is not empty
  if (file.contents.length === 0) {
    return false;
  }

  // Validate that buffer can be safely converted to string
  try {
    const content = file.contents.toString('utf8');
    // Basic validation - ensure it's not just null bytes or invalid characters
    if (content.length === 0 || content === '\0'.repeat(content.length)) {
      return false;
    }
    return true;
  } catch {
    // Buffer contains invalid UTF-8 sequences
    return false;
  }
}

/**
 * Content processing utilities for metalsmith-optimize-html
 * @module utils/content-processor
 */

/**
 * Process content with optimizers, handling excluded tags if specified
 * @param {string} content - The HTML content to process
 * @param {import('../index.js').Optimizer[]} optimizers - The array of optimizers to apply
 * @param {import('../index.js').Options} options - Configuration options
 * @returns {string} - The processed HTML content
 */
function processContent(content, optimizers, options) {
  var _options$excludeTags;
  let result;

  // If we have tags to exclude from processing
  if (((_options$excludeTags = options.excludeTags) == null ? void 0 : _options$excludeTags.length) > 0) {
    const preserved = [];
    const excludePattern = new RegExp(`<(${options.excludeTags.join('|')})[^>]*>[\\s\\S]*?</\\1>`, 'gi');

    // Preserve excluded tags
    content = content.replace(excludePattern, match => {
      preserved.push(match);
      return `___EXCLUDE_${preserved.length - 1}___`;
    });

    // Apply optimizers
    result = optimizers.reduce((text, optimizer) => optimizer.optimize(text, options), content);

    // Restore excluded content
    result = preserved.reduce((text, preservedContent, i) => text.replace(`___EXCLUDE_${i}___`, preservedContent), result);
  } else {
    // Normal optimization without exclusions
    result = optimizers.reduce((text, optimizer) => optimizer.optimize(text, options), content);
  }
  return result;
}

/**
 * Metalsmith plugin for HTML optimization
 * @module metalsmith-optimize-html
 */

// Configuration defaults
const BASE_OPTIONS = {
  pattern: '**/*.html',
  excludeTags: []
};
const DEFAULT_AGGRESSIVE_OPTIONS = {
  removeComments: true,
  removeTagSpaces: true,
  normalizeBooleanAttributes: true,
  cleanUrlAttributes: true,
  removeProtocols: true,
  removeDefaultAttributes: true,
  cleanDataAttributes: true,
  simplifyDoctype: true,
  safeRemoveAttributeQuotes: true
};

/**
 * @typedef {Object} Optimizer
 * @property {string} name - The name of the optimizer
 * @property {Function} optimize - The optimization function that processes HTML content
 */

/**
 * @typedef {Object} Options
 * @property {string} [pattern='**\/*.html'] - Glob pattern for matching files
 * @property {string[]} [excludeTags=[]] - HTML tags to exclude from optimization
 * @property {boolean} [aggressive=false] - Enable all optimizations
 * @property {boolean} [removeComments=false] - Remove HTML comments
 * @property {boolean} [removeTagSpaces=false] - Remove spaces around tags
 * @property {boolean} [normalizeBooleanAttributes=false] - Normalize boolean attributes
 * @property {boolean} [cleanUrlAttributes=false] - Clean URL attributes
 * @property {boolean} [removeProtocols=false] - Remove protocols from URLs
 * @property {boolean} [removeDefaultAttributes=false] - Remove default attributes
 * @property {boolean} [cleanDataAttributes=false] - Clean data attributes
 * @property {boolean} [simplifyDoctype=false] - Simplify DOCTYPE declaration
 * @property {boolean} [safeRemoveAttributeQuotes=false] - Safely remove attribute quotes
 */

/**
 * Process a single file with error handling and validation
 * @param {string} filename - The filename being processed
 * @param {Object} file - The Metalsmith file object
 * @param {Array} activeOptimizers - Array of optimizer functions
 * @param {Object} mergedOptions - Merged optimization options
 * @param {Function} debug - Debug logging function
 * @returns {boolean} True if file was processed successfully
 */
function processFile(filename, file, activeOptimizers, mergedOptions, debug) {
  // Validate file.contents before processing
  if (!isProcessableFile(file)) {
    debug('skipping %s: contents is not processable', filename);
    return false;
  }

  // Early filtering: skip empty files only
  if (!file.contents || file.contents.length === 0) {
    debug('skipping %s: empty file', filename);
    return false;
  }

  // Destructure file properties for cleaner access
  const {
    contents
  } = file;

  // Get content and process it with error handling
  let content;
  try {
    content = contents.toString('utf8');
  } catch (error) {
    debug('skipping %s: failed to decode content as UTF-8 - %s', filename, error.message);
    return false;
  }

  // Early filtering: skip files that don't contain HTML tags
  if (!content.includes('<') || !content.includes('>')) {
    debug('skipping %s: no HTML tags detected', filename);
    return false;
  }

  // Process content with error handling
  let optimizedContent;
  try {
    optimizedContent = processContent(content, activeOptimizers, mergedOptions);
  } catch (error) {
    debug('skipping %s: optimization failed - %s', filename, error.message);
    return false;
  }

  // Validate optimized content before updating file
  if (typeof optimizedContent !== 'string' || optimizedContent.length === 0) {
    debug('skipping %s: optimization produced invalid content', filename);
    return false;
  }

  // Update file with optimized content
  try {
    file.contents = Buffer.from(optimizedContent, 'utf8');
  } catch (error) {
    debug('skipping %s: failed to create buffer from optimized content - %s', filename, error.message);
    return false;
  }
  debug('processed %s (%d bytes -> %d bytes)', filename, content.length, optimizedContent.length);
  return true;
}

/**
 * Creates a Metalsmith plugin for HTML optimization
 *
 * @param {Options} userOptions - Configuration options
 * @returns {import('metalsmith').Plugin} Metalsmith plugin function
 * @throws {Error} If options are invalid
 */
function optimizeHTML(userOptions = {}) {
  // Validate user options
  const validation = validator.validateOptions(userOptions);
  if (!validation.valid) {
    throw new Error(validator.formatValidationErrors(validation.errors));
  }

  // Merge options with defaults
  const options = {
    ...BASE_OPTIONS,
    ...(userOptions.aggressive ? DEFAULT_AGGRESSIVE_OPTIONS : {}),
    ...userOptions
  };

  // Cache optimizers for reuse across multiple files
  let optimizers;

  /**
   * The plugin function - Two-phase processing:
   * Phase 1: Filter and match files for optimal performance
   * Phase 2: Process matched files with optimizers
   * @param {import('metalsmith').Files} files - Metalsmith files object
   * @param {import('metalsmith')} metalsmith - Metalsmith instance
   * @param {Function} done - Callback function
   */
  const plugin = function (files, metalsmith, done) {
    try {
      // Load optimizers on first run - now synchronous
      if (!optimizers) {
        optimizers = optimizerRegistry.loadOptimizers(options);
      }
      const debug = metalsmith.debug ? metalsmith.debug('metalsmith-optimize-html') : () => {};
      debug('running with options: %O', options);
      const activeOptimizers = plugin._testOptimizers || optimizers;

      // Phase 1: Filter files before expensive transformations
      const matchingFiles = getMatchingFiles(files, options.pattern, metalsmith);
      debug('matched %d files for processing', matchingFiles.length);

      // Access site-wide metadata for potential configuration
      const metadata = metalsmith.metadata();
      const siteOptions = metadata.htmlOptimization || {};
      const mergedOptions = {
        ...options,
        ...siteOptions
      };

      // Phase 2: Process each matching file
      for (const filename of matchingFiles) {
        const file = files[filename];
        processFile(filename, file, activeOptimizers, mergedOptions, debug);
      }
      done();
    } catch (error) {
      done(error);
    }
  };

  // Property for testing
  plugin._testOptimizers = null;

  // Set function name for better debugging
  Object.defineProperty(plugin, 'name', {
    value: 'optimizeHTML'
  });
  return plugin;
}

module.exports = optimizeHTML;
