// metalsmith-seo/lib/index.cjs
'use strict';

var cheerio = require('cheerio');
var path = require('path');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var path__default = /*#__PURE__*/_interopDefaultLegacy(path);

/**
 * @fileoverview Utility functions for object manipulation.
 */

/**
 * Safely gets a nested property value from an object using dot notation.
 * @param {Object} obj - The object to query
 * @param {string} pathStr - The property path (e.g., 'a.b.c')
 * @param {*} [defaultValue] - Value to return if the property is undefined
 * @returns {*} The property value or defaultValue
 */
const get = (obj, pathStr, defaultValue) => {
  if (!obj || !pathStr || typeof pathStr !== "string") {
    return defaultValue;
  }
  const keys = pathStr.split(".");
  let result = obj;
  for (const key of keys) {
    var _result;
    result = (_result = result) == null ? void 0 : _result[key];
    if (result === undefined) {
      return defaultValue;
    }
  }
  return result;
};

/**
 * @fileoverview Unified metadata extractor that processes frontmatter for SEO optimization.
 *
 * Metadata extraction priority order:
 * 1. seo object (highest priority - explicit SEO overrides)
 * 2. card object (for blog posts and content cards)
 * 3. root level properties
 * 4. configured defaults
 * 5. auto-generated content (lowest priority)
 */

/**
 * Extracts a property value following the priority chain:
 * 1. seo object (highest priority)
 * 2. card object
 * 3. root level (using fallback path)
 * 4. default value
 *
 * @param {Object} seoData - The seo object from frontmatter
 * @param {Object} frontmatter - The complete frontmatter object
 * @param {string|Object} propertyConfig - Property name as string, or object with paths for each location
 *   If string: used for all locations
 *   If object: { seo: 'seoPropertyName', card: 'cardPropertyName', root: 'rootPropertyName' }
 * @param {string} [fallbackPath] - Additional fallback path at root level
 * @param {*} [defaultValue] - Default value if not found anywhere
 * @returns {*} The extracted value
 */
function extractWithPriority(seoData, frontmatter, propertyConfig, fallbackPath, defaultValue) {
  let seoProperty, cardProperty, rootProperty;
  if (typeof propertyConfig === "string") {
    // Use same property name for all locations
    seoProperty = propertyConfig;
    cardProperty = propertyConfig;
    rootProperty = fallbackPath || propertyConfig;
  } else {
    // Use specific property names for each location
    seoProperty = propertyConfig.seo;
    cardProperty = propertyConfig.card;
    rootProperty = propertyConfig.root || fallbackPath;
  }

  // 1. Check seo object first (highest priority)
  if (seoProperty) {
    const seoValue = get(seoData, seoProperty);
    if (seoValue !== undefined && seoValue !== null) {
      return seoValue;
    }
  }

  // 2. Check card object
  if (cardProperty) {
    const cardValue = get(frontmatter, `card.${cardProperty}`);
    if (cardValue !== undefined && cardValue !== null) {
      return cardValue;
    }
  }

  // 3. Check root level
  if (rootProperty) {
    const rootValue = get(frontmatter, rootProperty);
    if (rootValue !== undefined && rootValue !== null) {
      return rootValue;
    }
  }

  // 4. Return default value
  return defaultValue;
}

/**
 * @typedef {Object} ExtractedMetadata
 * @property {string} title - Page title
 * @property {string} description - Page description
 * @property {string} [image] - Social sharing image URL
 * @property {string} [canonicalURL] - Canonical URL
 * @property {string} [robots] - Robots directive
 * @property {boolean} [noIndex] - Whether page should be excluded from indexing
 * @property {string} [type] - Content type (article, page, product, etc.)
 * @property {Date|string} [publishDate] - Publication date
 * @property {Date|string} [modifiedDate] - Last modified date
 * @property {string} [author] - Author name
 * @property {string[]} [keywords] - Keywords/tags
 * @property {number} [wordCount] - Word count from content
 * @property {number} [readingTime] - Estimated reading time in minutes
 */

/**
 * @typedef {Object} MetadataOptions
 * @property {string} hostname - Base hostname for generating URLs
 * @property {string} [seoProperty='seo'] - Frontmatter property containing SEO data
 * @property {Object} [defaults] - Default values for missing metadata
 * @property {Object} [fallbacks] - Fallback property mappings
 * @property {number} [wordsPerMinute=200] - Reading speed for calculating reading time
 */

/**
 * Extracts and normalizes metadata from file frontmatter
 * @param {string} filePath - File path relative to source
 * @param {Object} frontmatter - File frontmatter and content
 * @param {MetadataOptions} options - Extraction options
 * @returns {ExtractedMetadata} Extracted and normalized metadata
 */
function extractMetadata(filePath, frontmatter, options) {
  const {
    hostname,
    seoProperty = "seo",
    defaults = {},
    fallbacks = {},
    wordsPerMinute = 200
  } = options;

  // Ensure hostname is a string
  const hostnameStr = String(hostname || "");

  // Get SEO-specific metadata (highest priority)
  const seoData = get(frontmatter, seoProperty) || {};

  // Extract metadata with fallback chain: seo -> defaults -> fallbacks -> auto-generated
  const metadata = {
    // Core SEO properties
    title: extractTitle(seoData, frontmatter, defaults, fallbacks),
    description: extractDescription(seoData, frontmatter, defaults, fallbacks),
    image: extractImage(seoData, frontmatter, defaults, fallbacks, hostnameStr),
    canonicalURL: extractCanonicalURL(seoData, filePath, hostnameStr),
    robots: extractRobots(seoData, defaults),
    noIndex: extractNoIndex(seoData),
    type: extractContentType(seoData, frontmatter),
    // Dates
    publishDate: extractPublishDate(seoData, frontmatter, fallbacks),
    modifiedDate: extractModifiedDate(seoData, frontmatter, fallbacks),
    // Author and content metadata
    author: extractAuthor(seoData, frontmatter, fallbacks, defaults),
    keywords: extractKeywords(seoData, frontmatter, fallbacks),
    // Content analysis
    wordCount: extractWordCount(frontmatter),
    readingTime: null // Will be calculated after word count
  };

  // Calculate reading time
  if (metadata.wordCount) {
    metadata.readingTime = calculateReadingTime(metadata.wordCount, wordsPerMinute);
  }
  return metadata;
}

/**
 * Extract title with fallback chain
 */
function extractTitle(seoData, frontmatter, defaults, fallbacks) {
  const extracted = extractWithPriority(seoData, frontmatter, "title", fallbacks.title || "title");
  return extracted || defaults.title || "Untitled";
}

/**
 * Extract description with fallback chain
 */
function extractDescription(seoData, frontmatter, defaults, fallbacks) {
  // Note: card uses "excerpt" while seo uses "description"
  const extracted = extractWithPriority(seoData, frontmatter, {
    seo: "description",
    card: "excerpt",
    // card uses "excerpt" instead of "description"
    root: fallbacks.description || "excerpt"
  });
  return extracted || defaults.description || autoGenerateDescription(frontmatter.contents);
}

/**
 * Extract image with fallback chain
 */
function extractImage(seoData, frontmatter, defaults, fallbacks, hostname) {
  // Check both "image" and "socialImage" in seo
  const seoImage = seoData.image || seoData.socialImage;
  const image = seoImage || extractWithPriority({},
  // Empty seo since we already checked it
  frontmatter, "image", fallbacks.image || "featured_image") || defaults.socialImage;

  // Convert relative URLs to absolute
  if (image && !image.startsWith("http")) {
    return `${hostname.replace(/\/$/, "")}${image.startsWith("/") ? "" : "/"}${image}`;
  }
  return image;
}

/**
 * Extract canonical URL
 */
function extractCanonicalURL(seoData, filePath, hostname) {
  if (seoData.canonicalURL) {
    return seoData.canonicalURL;
  }

  // Auto-generate from file path
  const cleanPath = filePath.replace(/\.html?$/, "").replace(/\/index$/, "");
  return `${hostname.replace(/\/$/, "")}/${cleanPath}`.replace(/\/+/g, "/");
}

/**
 * Extract robots directive
 */
function extractRobots(seoData, defaults) {
  return seoData.robots || defaults.robots || "index,follow";
}

/**
 * Extract noIndex flag
 */
function extractNoIndex(seoData) {
  return Boolean(seoData.noIndex);
}

/**
 * Auto-detect content type
 */
function extractContentType(seoData, frontmatter) {
  if (seoData.type) {
    return seoData.type;
  }

  // Use our centralized extraction for date and author
  const hasDate = extractWithPriority(seoData, frontmatter, "date", "date");
  const hasAuthor = extractWithPriority(seoData, frontmatter, "author", "author");
  const hasTags = get(frontmatter, "tags") || get(frontmatter, "card.tags");

  // Auto-detect based on frontmatter properties
  if (hasDate && (hasAuthor || hasTags)) {
    return "article";
  }
  if (frontmatter.price || frontmatter.sku) {
    return "product";
  }
  if (frontmatter.address || frontmatter.phone) {
    return "local-business";
  }
  return "page";
}

/**
 * Extract publish date
 */
function extractPublishDate(seoData, frontmatter, fallbacks) {
  const date = extractWithPriority(seoData, frontmatter, {
    seo: "publishDate",
    card: "date",
    // card uses "date" instead of "publishDate"
    root: fallbacks.publishDate || "date"
  });
  return normalizeDate(date);
}

/**
 * Extract modified date
 */
function extractModifiedDate(seoData, frontmatter, fallbacks) {
  const date = get(seoData, "modifiedDate") || get(frontmatter, fallbacks.modifiedDate || "updated");
  return normalizeDate(date);
}

/**
 * Extract author information
 */
function extractAuthor(seoData, frontmatter, fallbacks, defaults) {
  // Try the standard extraction first
  let author = extractWithPriority(seoData, frontmatter, "author", fallbacks.author || "author.name");

  // If not found and we used author.name as fallback, also try just "author"
  if (!author && (!fallbacks.author || fallbacks.author === "author.name")) {
    author = get(frontmatter, "author");
  }

  // Final fallback to site-wide default author
  if (!author) {
    author = defaults.siteOwner || defaults.author;
  }

  // Handle array of authors (join them)
  if (Array.isArray(author)) {
    return author.join(", ");
  }
  return author;
}

/**
 * Extract keywords/tags
 */
function extractKeywords(seoData, frontmatter, fallbacks) {
  const keywords = extractWithPriority(seoData, frontmatter, {
    seo: "keywords",
    card: "tags",
    // card might use "tags" instead of "keywords"
    root: fallbacks.keywords || "tags"
  });
  if (Array.isArray(keywords)) {
    return keywords;
  }
  if (typeof keywords === "string") {
    return keywords.split(",").map(k => k.trim());
  }
  return [];
}

/**
 * Extract word count from content
 */
function extractWordCount(frontmatter) {
  if (!Buffer.isBuffer(frontmatter.contents)) {
    return 0;
  }
  const content = frontmatter.contents.toString();
  const words = content.replace(/<[^>]*>/g, "") // Remove HTML tags
  .replace(/\s+/g, " ") // Normalize whitespace
  .trim().split(" ").filter(word => word.length > 0);
  return words.length;
}

/**
 * Calculate reading time from word count
 * @param {number} wordCount - Number of words in content
 * @param {number} [wordsPerMinute=200] - Reading speed in words per minute
 * @returns {number} Reading time in minutes
 */
function calculateReadingTime(wordCount, wordsPerMinute = 200) {
  return Math.ceil(wordCount / wordsPerMinute);
}

/**
 * Auto-generate description from content
 */
function autoGenerateDescription(contents, maxLength = 160) {
  if (!Buffer.isBuffer(contents)) {
    return "";
  }
  const content = contents.toString().replace(/<[^>]*>/g, "") // Remove HTML tags
  .replace(/\s+/g, " ") // Normalize whitespace
  .trim();
  if (content.length <= maxLength) {
    return content;
  }

  // Find the last complete sentence within the limit
  const truncated = content.substring(0, maxLength);
  const lastSentence = truncated.lastIndexOf(".");
  const lastSpace = truncated.lastIndexOf(" ");
  if (lastSentence > maxLength * 0.6) {
    return content.substring(0, lastSentence + 1);
  }
  if (lastSpace > maxLength * 0.8) {
    return `${content.substring(0, lastSpace)}...`;
  }
  return `${truncated}...`;
}

/**
 * Normalize date to ISO string
 */
function normalizeDate(date) {
  if (!date) {
    return null;
  }
  if (date instanceof Date) {
    return date.toISOString();
  }
  if (typeof date === "string") {
    const parsed = new Date(date);
    if (!isNaN(parsed.getTime())) {
      return parsed.toISOString();
    }
  }
  return null;
}

/**
 * @fileoverview Generator for basic HTML meta tags.
 */

/**
 * @typedef {Object} MetaTagsResult
 * @property {string} title - Title tag content
 * @property {Array<Object>} metaTags - Array of meta tag objects
 * @property {Array<Object>} linkTags - Array of link tag objects
 */

/**
 * Generates basic HTML meta tags from extracted metadata
 * @param {Object} metadata - Extracted metadata object
 * @param {Object} siteConfig - Site-wide configuration
 * @returns {MetaTagsResult} Generated meta tags
 */
function generateMetaTags(metadata, siteConfig = {}) {
  const metaTags = [];
  const linkTags = [];

  // Basic meta tags
  if (metadata.description) {
    metaTags.push({
      name: "description",
      content: metadata.description
    });
  }
  if (metadata.keywords && metadata.keywords.length > 0) {
    metaTags.push({
      name: "keywords",
      content: Array.isArray(metadata.keywords) ? metadata.keywords.join(", ") : metadata.keywords
    });
  }

  // Robots directive
  const robots = generateRobotsDirective(metadata, siteConfig);
  if (robots) {
    metaTags.push({
      name: "robots",
      content: robots
    });
  }

  // Author
  if (metadata.author) {
    metaTags.push({
      name: "author",
      content: metadata.author
    });
  }

  // Viewport (configurable for mobile optimization)
  const viewportContent = siteConfig.viewport || "width=device-width, initial-scale=1.0";
  metaTags.push({
    name: "viewport",
    content: viewportContent
  });

  // Canonical URL
  if (metadata.canonicalURL) {
    linkTags.push({
      rel: "canonical",
      href: metadata.canonicalURL
    });
  }

  // Theme color (if specified in site config)
  if (siteConfig.themeColor) {
    metaTags.push({
      name: "theme-color",
      content: siteConfig.themeColor
    });
  }

  // Additional technical meta tags
  addTechnicalMetaTags(metaTags, metadata, siteConfig);
  return {
    title: metadata.title,
    metaTags,
    linkTags
  };
}

/**
 * Generates robots directive based on metadata and configuration
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 * @returns {string|null} Robots directive string
 */
function generateRobotsDirective(metadata, siteConfig) {
  if (metadata.noIndex) {
    return "noindex,nofollow";
  }
  if (metadata.robots) {
    return metadata.robots;
  }

  // Default based on content type
  switch (metadata.type) {
    case "article":
      return "index,follow";
    case "page":
      return "index,follow";
    case "product":
      return "index,follow";
    default:
      return siteConfig.defaultRobots || "index,follow";
  }
}

/**
 * Adds technical meta tags for advanced SEO
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function addTechnicalMetaTags(metaTags, metadata, siteConfig) {
  // Content language
  if (siteConfig.language) {
    metaTags.push({
      httpEquiv: "content-language",
      content: siteConfig.language
    });
  }

  // Publisher
  if (siteConfig.publisher) {
    metaTags.push({
      name: "publisher",
      content: siteConfig.publisher
    });
  }

  // Copyright
  if (siteConfig.copyright) {
    metaTags.push({
      name: "copyright",
      content: siteConfig.copyright
    });
  }

  // Googlebot specific directives
  const googlebotDirectives = [];
  if (siteConfig.maxSnippet) {
    googlebotDirectives.push(`max-snippet:${siteConfig.maxSnippet}`);
  }
  if (siteConfig.maxImagePreview) {
    googlebotDirectives.push(`max-image-preview:${siteConfig.maxImagePreview}`);
  }
  if (siteConfig.maxVideoPreview) {
    googlebotDirectives.push(`max-video-preview:${siteConfig.maxVideoPreview}`);
  }
  if (googlebotDirectives.length > 0) {
    metaTags.push({
      name: "googlebot",
      content: googlebotDirectives.join(",")
    });
  }

  // Article-specific meta tags
  if (metadata.type === "article") {
    addArticleMetaTags(metaTags, metadata);
  }
}

/**
 * Adds article-specific meta tags
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 */
function addArticleMetaTags(metaTags, metadata) {
  if (metadata.publishDate) {
    const date = new Date(metadata.publishDate);
    metaTags.push({
      name: "article:published_time",
      content: date.toISOString()
    });
  }
  if (metadata.modifiedDate) {
    const date = new Date(metadata.modifiedDate);
    metaTags.push({
      name: "article:modified_time",
      content: date.toISOString()
    });
  }
  if (metadata.author) {
    metaTags.push({
      name: "article:author",
      content: metadata.author
    });
  }
  if (metadata.keywords && metadata.keywords.length > 0) {
    metadata.keywords.forEach(keyword => {
      metaTags.push({
        name: "article:tag",
        content: keyword
      });
    });
  }
}

/**
 * @fileoverview Generator for Open Graph meta tags.
 */

// Image type mapping for Open Graph images
const IMAGE_TYPE_MAP = {
  jpg: "image/jpeg",
  jpeg: "image/jpeg",
  png: "image/png",
  gif: "image/gif",
  webp: "image/webp",
  svg: "image/svg+xml"
};

// Content type to Open Graph type mapping
const OG_TYPE_MAP = {
  article: "article",
  product: "product",
  profile: "profile",
  page: "website",
  "local-business": "business.business"
};

/**
 * @typedef {Object} OpenGraphResult
 * @property {Array<Object>} metaTags - Array of Open Graph meta tag objects
 */

/**
 * Generates Open Graph meta tags from extracted metadata
 * @param {Object} metadata - Extracted metadata object
 * @param {Object} siteConfig - Site-wide configuration
 * @returns {OpenGraphResult} Generated Open Graph tags
 */
function generateOpenGraphTags(metadata, siteConfig = {}) {
  const metaTags = [];

  // Core Open Graph tags
  addCoreOpenGraphTags(metaTags, metadata, siteConfig);

  // Type-specific tags
  switch (metadata.type) {
    case "article":
      addArticleOpenGraphTags(metaTags, metadata);
      break;
    case "product":
      addProductOpenGraphTags(metaTags, metadata);
      break;
    case "profile":
      addProfileOpenGraphTags(metaTags, metadata);
      break;
  }

  // Site-specific tags
  addSiteOpenGraphTags(metaTags, siteConfig);
  return {
    metaTags
  };
}

/**
 * Adds core Open Graph tags (required for all content)
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function addCoreOpenGraphTags(metaTags, metadata, siteConfig) {
  // og:title (required)
  if (metadata.title) {
    metaTags.push({
      property: "og:title",
      content: metadata.title
    });
  }

  // og:type (required)
  const ogType = mapContentTypeToOGType(metadata.type);
  metaTags.push({
    property: "og:type",
    content: ogType
  });

  // og:url (required)
  if (metadata.canonicalURL) {
    metaTags.push({
      property: "og:url",
      content: metadata.canonicalURL
    });
  }

  // og:description
  if (metadata.description) {
    metaTags.push({
      property: "og:description",
      content: metadata.description
    });
  }

  // og:image (recommended)
  if (metadata.image) {
    metaTags.push({
      property: "og:image",
      content: metadata.image
    });

    // Add image metadata if available
    addImageMetadata(metaTags, metadata, siteConfig);
  }

  // og:site_name
  if (siteConfig.siteName) {
    metaTags.push({
      property: "og:site_name",
      content: siteConfig.siteName
    });
  }

  // og:locale
  const locale = siteConfig.locale || siteConfig.language || "en_US";
  metaTags.push({
    property: "og:locale",
    content: locale
  });
}

/**
 * Adds image metadata for Open Graph
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function addImageMetadata(metaTags, metadata, siteConfig) {
  // Default image dimensions for social sharing
  const defaultWidth = siteConfig.ogImageWidth || 1200;
  const defaultHeight = siteConfig.ogImageHeight || 630;
  metaTags.push({
    property: "og:image:width",
    content: String(defaultWidth)
  });
  metaTags.push({
    property: "og:image:height",
    content: String(defaultHeight)
  });

  // Image alt text (accessibility)
  const imageAlt = metadata.imageAlt || metadata.title || "Image";
  metaTags.push({
    property: "og:image:alt",
    content: imageAlt
  });

  // Image type (if known)
  if (metadata.image) {
    const imageType = getImageType(metadata.image);
    if (imageType) {
      metaTags.push({
        property: "og:image:type",
        content: imageType
      });
    }
  }
}

/**
 * Adds article-specific Open Graph tags
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 */
function addArticleOpenGraphTags(metaTags, metadata) {
  // article:published_time
  if (metadata.publishDate) {
    metaTags.push({
      property: "article:published_time",
      content: new Date(metadata.publishDate).toISOString()
    });
  }

  // article:modified_time
  if (metadata.modifiedDate) {
    metaTags.push({
      property: "article:modified_time",
      content: new Date(metadata.modifiedDate).toISOString()
    });
  }

  // article:author
  if (metadata.author) {
    metaTags.push({
      property: "article:author",
      content: metadata.author
    });
  }

  // article:section
  if (metadata.section) {
    metaTags.push({
      property: "article:section",
      content: metadata.section
    });
  }

  // article:tag (multiple tags)
  if (metadata.keywords && metadata.keywords.length > 0) {
    metadata.keywords.forEach(tag => {
      metaTags.push({
        property: "article:tag",
        content: tag
      });
    });
  }

  // Reading time (custom property)
  if (metadata.readingTime) {
    metaTags.push({
      property: "article:reading_time",
      content: metadata.readingTime
    });
  }
}

/**
 * Adds product-specific Open Graph tags
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 */
function addProductOpenGraphTags(metaTags, metadata) {
  // product:brand
  if (metadata.brand) {
    metaTags.push({
      property: "product:brand",
      content: metadata.brand
    });
  }

  // product:availability
  if (metadata.availability) {
    metaTags.push({
      property: "product:availability",
      content: metadata.availability
    });
  }

  // product:condition
  if (metadata.condition) {
    metaTags.push({
      property: "product:condition",
      content: metadata.condition
    });
  }

  // product:price
  if (metadata.price) {
    metaTags.push({
      property: "product:price:amount",
      content: String(metadata.price)
    });
  }

  // product:price:currency
  if (metadata.currency) {
    metaTags.push({
      property: "product:price:currency",
      content: metadata.currency
    });
  }
}

/**
 * Adds profile-specific Open Graph tags
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 */
function addProfileOpenGraphTags(metaTags, metadata) {
  // profile:first_name
  if (metadata.firstName) {
    metaTags.push({
      property: "profile:first_name",
      content: metadata.firstName
    });
  }

  // profile:last_name
  if (metadata.lastName) {
    metaTags.push({
      property: "profile:last_name",
      content: metadata.lastName
    });
  }

  // profile:username
  if (metadata.username) {
    metaTags.push({
      property: "profile:username",
      content: metadata.username
    });
  }
}

/**
 * Adds site-specific Open Graph tags
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} siteConfig - Site configuration
 */
function addSiteOpenGraphTags(metaTags, siteConfig) {
  // Facebook App ID
  if (siteConfig.facebookAppId) {
    metaTags.push({
      property: "fb:app_id",
      content: siteConfig.facebookAppId
    });
  }

  // Facebook Admins
  if (siteConfig.facebookAdmins) {
    const admins = Array.isArray(siteConfig.facebookAdmins) ? siteConfig.facebookAdmins : [siteConfig.facebookAdmins];
    admins.forEach(admin => {
      metaTags.push({
        property: "fb:admins",
        content: String(admin)
      });
    });
  }
}

/**
 * Maps content types to Open Graph types
 * @param {string} contentType - Internal content type
 * @returns {string} Open Graph type
 */
function mapContentTypeToOGType(contentType) {
  return OG_TYPE_MAP[contentType] || "website";
}

/**
 * Determines image MIME type from file extension
 * @param {string} imageUrl - Image URL
 * @returns {string|null} MIME type
 */
function getImageType(imageUrl) {
  var _imageUrl$split$pop;
  const extension = (_imageUrl$split$pop = imageUrl.split(".").pop()) == null ? void 0 : _imageUrl$split$pop.toLowerCase();
  return IMAGE_TYPE_MAP[extension] || null;
}

/**
 * @fileoverview Generator for Twitter Card meta tags.
 */

/**
 * @typedef {Object} TwitterCardResult
 * @property {Array<Object>} metaTags - Array of Twitter Card meta tag objects
 */

/**
 * Generates Twitter Card meta tags from extracted metadata
 * @param {Object} metadata - Extracted metadata object
 * @param {Object} siteConfig - Site-wide configuration
 * @returns {TwitterCardResult} Generated Twitter Card tags
 */
function generateTwitterCardTags(metadata, siteConfig = {}) {
  const metaTags = [];

  // Determine card type
  const cardType = determineCardType(metadata, siteConfig);

  // Core Twitter Card tags
  addCoreTwitterTags(metaTags, metadata, siteConfig, cardType);

  // Card type-specific tags
  switch (cardType) {
    case "summary_large_image":
      addSummaryLargeImageTags(metaTags, metadata);
      break;
    case "summary":
      addSummaryTags(metaTags, metadata);
      break;
    case "app":
      addAppTags(metaTags, metadata, siteConfig);
      break;
    case "player":
      addPlayerTags(metaTags, metadata);
      break;
    default:
      // Default to summary
      addSummaryTags(metaTags, metadata);
      break;
  }
  return {
    metaTags
  };
}

/**
 * Determines the appropriate Twitter Card type
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 * @returns {string} Twitter Card type
 */
function determineCardType(metadata, siteConfig) {
  // Allow explicit override in site config or metadata
  if (metadata.twitterCard) {
    return metadata.twitterCard;
  }
  if (siteConfig.twitterCardType) {
    return siteConfig.twitterCardType;
  }

  // Auto-determine based on content
  if (metadata.image) {
    return "summary_large_image";
  }
  if (metadata.type === "video" || metadata.videoUrl) {
    return "player";
  }
  if (metadata.type === "app" || metadata.appId) {
    return "app";
  }
  return "summary";
}

/**
 * Adds core Twitter Card tags (common to all card types)
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 * @param {string} cardType - Twitter card type
 */
function addCoreTwitterTags(metaTags, metadata, siteConfig, cardType) {
  // twitter:card (required)
  metaTags.push({
    name: "twitter:card",
    content: cardType
  });

  // twitter:site (recommended)
  if (siteConfig.twitterSite) {
    metaTags.push({
      name: "twitter:site",
      content: ensureTwitterHandle(siteConfig.twitterSite)
    });
  }

  // twitter:creator (optional)
  const creator = metadata.twitterCreator || metadata.author || siteConfig.twitterCreator;
  if (creator) {
    metaTags.push({
      name: "twitter:creator",
      content: ensureTwitterHandle(creator)
    });
  }

  // twitter:title (falls back to og:title)
  if (metadata.title) {
    // Twitter titles should be shorter than OG titles
    const twitterTitle = truncateTitle(metadata.title, 70);
    metaTags.push({
      name: "twitter:title",
      content: twitterTitle
    });
  }

  // twitter:description (falls back to og:description)
  if (metadata.description) {
    // Twitter descriptions should be shorter
    const maxLength = siteConfig.twitterDescriptionLength || 200;
    const twitterDescription = truncateDescription(metadata.description, maxLength);
    metaTags.push({
      name: "twitter:description",
      content: twitterDescription
    });
  }
}

/**
 * Adds tags specific to summary_large_image cards
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 */
function addSummaryLargeImageTags(metaTags, metadata) {
  if (metadata.image) {
    metaTags.push({
      name: "twitter:image",
      content: metadata.image
    });

    // Image alt text (accessibility)
    const imageAlt = metadata.imageAlt || metadata.title || "Image";
    metaTags.push({
      name: "twitter:image:alt",
      content: imageAlt
    });
  }
}

/**
 * Adds tags specific to summary cards
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 */
function addSummaryTags(metaTags, metadata) {
  if (metadata.image) {
    metaTags.push({
      name: "twitter:image",
      content: metadata.image
    });

    // Image alt text (accessibility)
    const imageAlt = metadata.imageAlt || metadata.title || "Image";
    metaTags.push({
      name: "twitter:image:alt",
      content: imageAlt
    });
  }
}

/**
 * Adds tags specific to app cards
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function addAppTags(metaTags, metadata, siteConfig) {
  // iOS app
  if (metadata.iosAppId || siteConfig.iosAppId) {
    metaTags.push({
      name: "twitter:app:id:iphone",
      content: metadata.iosAppId || siteConfig.iosAppId
    });
  }
  if (metadata.iosAppUrl || siteConfig.iosAppUrl) {
    metaTags.push({
      name: "twitter:app:url:iphone",
      content: metadata.iosAppUrl || siteConfig.iosAppUrl
    });
  }

  // Android app
  if (metadata.androidAppId || siteConfig.androidAppId) {
    metaTags.push({
      name: "twitter:app:id:googleplay",
      content: metadata.androidAppId || siteConfig.androidAppId
    });
  }
  if (metadata.androidAppUrl || siteConfig.androidAppUrl) {
    metaTags.push({
      name: "twitter:app:url:googleplay",
      content: metadata.androidAppUrl || siteConfig.androidAppUrl
    });
  }

  // App name
  if (metadata.appName || siteConfig.appName) {
    metaTags.push({
      name: "twitter:app:name:iphone",
      content: metadata.appName || siteConfig.appName
    });
    metaTags.push({
      name: "twitter:app:name:googleplay",
      content: metadata.appName || siteConfig.appName
    });
  }
}

/**
 * Adds tags specific to player cards
 * @param {Array} metaTags - Meta tags array to modify
 * @param {Object} metadata - Extracted metadata
 */
function addPlayerTags(metaTags, metadata) {
  if (metadata.videoUrl) {
    metaTags.push({
      name: "twitter:player",
      content: metadata.videoUrl
    });
  }
  if (metadata.videoWidth) {
    metaTags.push({
      name: "twitter:player:width",
      content: String(metadata.videoWidth)
    });
  }
  if (metadata.videoHeight) {
    metaTags.push({
      name: "twitter:player:height",
      content: String(metadata.videoHeight)
    });
  }
  if (metadata.videoStreamUrl) {
    metaTags.push({
      name: "twitter:player:stream",
      content: metadata.videoStreamUrl
    });
  }

  // Player image (thumbnail)
  if (metadata.image) {
    metaTags.push({
      name: "twitter:image",
      content: metadata.image
    });
  }
}

/**
 * Ensures Twitter handle has @ prefix
 * @param {string} handle - Twitter handle
 * @returns {string} Formatted Twitter handle
 */
function ensureTwitterHandle(handle) {
  if (typeof handle !== "string") {
    return String(handle);
  }
  const cleaned = handle.trim();
  return cleaned.startsWith("@") ? cleaned : `@${cleaned}`;
}

/**
 * Truncates title to fit Twitter's recommendations
 * @param {string} title - Original title
 * @param {number} maxLength - Maximum length
 * @returns {string} Truncated title
 */
function truncateTitle(title, maxLength = 70) {
  if (title.length <= maxLength) {
    return title;
  }

  // Find the last space before the limit
  const truncated = title.substring(0, maxLength);
  const lastSpace = truncated.lastIndexOf(" ");
  if (lastSpace > maxLength * 0.8) {
    return `${title.substring(0, lastSpace)}...`;
  }
  return `${truncated}...`;
}

/**
 * Truncates description to fit Twitter's recommendations
 * @param {string} description - Original description
 * @param {number} maxLength - Maximum length
 * @returns {string} Truncated description
 */
function truncateDescription(description, maxLength = 200) {
  if (description.length <= maxLength) {
    return description;
  }

  // Find the last complete sentence within the limit
  const truncated = description.substring(0, maxLength);
  const lastSentence = truncated.lastIndexOf(".");
  const lastSpace = truncated.lastIndexOf(" ");
  if (lastSentence > maxLength * 0.6) {
    return description.substring(0, lastSentence + 1);
  }
  if (lastSpace > maxLength * 0.8) {
    return `${description.substring(0, lastSpace)}...`;
  }
  return `${truncated}...`;
}

/**
 * @fileoverview Generator for JSON-LD structured data.
 */

/**
 * @typedef {Object} JsonLdResult
 * @property {Array<Object>} schemas - Array of JSON-LD schema objects
 * @property {string} html - HTML script tag containing JSON-LD
 */

/**
 * Generates JSON-LD structured data from extracted metadata
 * @param {Object} metadata - Extracted metadata object
 * @param {Object} siteConfig - Site-wide configuration
 * @param {string} filePath - File path for context
 * @returns {JsonLdResult} Generated JSON-LD schemas
 */
function generateJsonLd(metadata, siteConfig = {}, filePath = "") {
  const schemas = [];

  // Generate schemas based on content type and configuration
  generateWebSiteSchema(schemas, siteConfig);
  generateContentSchema(schemas, metadata, siteConfig);
  generateBreadcrumbSchema(schemas, filePath, siteConfig);
  generateOrganizationSchema(schemas, siteConfig);

  // Convert schemas to JSON-LD context
  const jsonLdSchemas = schemas.map(schema => ({
    "@context": "https://schema.org",
    ...schema
  }));

  // Generate HTML script tag
  const html = generateJsonLdHtml(jsonLdSchemas);
  return {
    schemas: jsonLdSchemas,
    html
  };
}

/**
 * Generates WebSite schema
 * @param {Array} schemas - Schemas array to modify
 * @param {Object} siteConfig - Site configuration
 */
function generateWebSiteSchema(schemas, siteConfig) {
  if (!siteConfig.name && !siteConfig.siteName) {
    return;
  }
  const webSiteSchema = {
    "@type": "WebSite",
    name: siteConfig.siteName || siteConfig.name,
    url: siteConfig.hostname || siteConfig.url
  };

  // Add search action if configured
  if (siteConfig.searchUrl) {
    webSiteSchema.potentialAction = {
      "@type": "SearchAction",
      target: {
        "@type": "EntryPoint",
        urlTemplate: `${siteConfig.searchUrl}?q={search_term_string}`
      },
      "query-input": "required name=search_term_string"
    };
  }

  // Add alternate name
  if (siteConfig.alternateNames) {
    webSiteSchema.alternateName = Array.isArray(siteConfig.alternateNames) ? siteConfig.alternateNames : [siteConfig.alternateNames];
  }
  schemas.push(webSiteSchema);
}

/**
 * Generates content-specific schema based on type
 * @param {Array} schemas - Schemas array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function generateContentSchema(schemas, metadata, siteConfig) {
  switch (metadata.type) {
    case "article":
      generateArticleSchema(schemas, metadata, siteConfig);
      break;
    case "product":
      generateProductSchema(schemas, metadata);
      break;
    case "local-business":
      generateLocalBusinessSchema(schemas, metadata, siteConfig);
      break;
    case "page":
    default:
      generateWebPageSchema(schemas, metadata, siteConfig);
      break;
  }
}

/**
 * Generates Article schema
 * @param {Array} schemas - Schemas array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function generateArticleSchema(schemas, metadata, siteConfig) {
  const articleSchema = {
    "@type": "Article",
    headline: metadata.title,
    url: metadata.canonicalURL
  };

  // Description
  if (metadata.description) {
    articleSchema.description = metadata.description;
  }

  // Image
  if (metadata.image) {
    articleSchema.image = {
      "@type": "ImageObject",
      url: metadata.image
    };
  }

  // Dates
  if (metadata.publishDate) {
    articleSchema.datePublished = new Date(metadata.publishDate).toISOString();
  }
  if (metadata.modifiedDate) {
    articleSchema.dateModified = new Date(metadata.modifiedDate).toISOString();
  }

  // Author
  if (metadata.author) {
    articleSchema.author = {
      "@type": "Person",
      name: metadata.author
    };
  }

  // Publisher (organization)
  if (siteConfig.organization) {
    articleSchema.publisher = generatePublisherSchema(siteConfig.organization);
  }

  // Article section
  if (metadata.section) {
    articleSchema.articleSection = metadata.section;
  }

  // Keywords
  if (metadata.keywords && metadata.keywords.length > 0) {
    articleSchema.keywords = metadata.keywords;
  }

  // Word count
  if (metadata.wordCount) {
    articleSchema.wordCount = metadata.wordCount;
  }

  // Reading time
  if (metadata.readingTime) {
    articleSchema.timeRequired = metadata.readingTime;
  }
  schemas.push(articleSchema);
}

/**
 * Generates Product schema
 * @param {Array} schemas - Schemas array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function generateProductSchema(schemas, metadata) {
  const productSchema = {
    "@type": "Product",
    name: metadata.title,
    description: metadata.description
  };

  // Image
  if (metadata.image) {
    productSchema.image = metadata.image;
  }

  // Brand
  if (metadata.brand) {
    productSchema.brand = {
      "@type": "Brand",
      name: metadata.brand
    };
  }

  // SKU
  if (metadata.sku) {
    productSchema.sku = metadata.sku;
  }

  // Offers
  if (metadata.price) {
    productSchema.offers = {
      "@type": "Offer",
      price: metadata.price,
      priceCurrency: metadata.currency || "USD",
      availability: `https://schema.org/${metadata.availability || "InStock"}`,
      condition: `https://schema.org/${metadata.condition || "NewCondition"}`
    };
  }

  // Aggregate rating
  if (metadata.rating) {
    productSchema.aggregateRating = {
      "@type": "AggregateRating",
      ratingValue: metadata.rating,
      ratingCount: metadata.ratingCount || 1
    };
  }
  schemas.push(productSchema);
}

/**
 * Generates LocalBusiness schema
 * @param {Array} schemas - Schemas array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function generateLocalBusinessSchema(schemas, metadata, siteConfig) {
  const businessSchema = {
    "@type": "LocalBusiness",
    name: metadata.title || siteConfig.businessName,
    description: metadata.description
  };

  // Address
  if (metadata.address || siteConfig.address) {
    const address = metadata.address || siteConfig.address;
    businessSchema.address = {
      "@type": "PostalAddress",
      streetAddress: address.streetAddress,
      addressLocality: address.city,
      addressRegion: address.state,
      postalCode: address.postalCode,
      addressCountry: address.country
    };
  }

  // Geographic coordinates
  if (metadata.latitude && metadata.longitude) {
    businessSchema.geo = {
      "@type": "GeoCoordinates",
      latitude: metadata.latitude,
      longitude: metadata.longitude
    };
  }

  // Contact information
  if (metadata.phone || siteConfig.phone) {
    businessSchema.telephone = metadata.phone || siteConfig.phone;
  }
  if (metadata.email || siteConfig.email) {
    businessSchema.email = metadata.email || siteConfig.email;
  }

  // Opening hours
  if (metadata.openingHours || siteConfig.openingHours) {
    businessSchema.openingHours = metadata.openingHours || siteConfig.openingHours;
  }

  // Image
  if (metadata.image) {
    businessSchema.image = metadata.image;
  }
  schemas.push(businessSchema);
}

/**
 * Generates WebPage schema
 * @param {Array} schemas - Schemas array to modify
 * @param {Object} metadata - Extracted metadata
 * @param {Object} siteConfig - Site configuration
 */
function generateWebPageSchema(schemas, metadata, siteConfig) {
  const webPageSchema = {
    "@type": "WebPage",
    name: metadata.title,
    url: metadata.canonicalURL
  };
  if (metadata.description) {
    webPageSchema.description = metadata.description;
  }
  if (metadata.image) {
    webPageSchema.image = metadata.image;
  }

  // Author (even for regular pages, if available)
  if (metadata.author) {
    webPageSchema.author = {
      "@type": "Person",
      name: metadata.author
    };
  }

  // Part of website
  if (siteConfig.siteName) {
    webPageSchema.isPartOf = {
      "@type": "WebSite",
      name: siteConfig.siteName,
      url: siteConfig.hostname
    };
  }
  schemas.push(webPageSchema);
}

/**
 * Generates BreadcrumbList schema
 * @param {Array} schemas - Schemas array to modify
 * @param {string} filePath - File path
 * @param {Object} siteConfig - Site configuration
 */
function generateBreadcrumbSchema(schemas, filePath, siteConfig) {
  if (!filePath || !siteConfig.hostname) {
    return;
  }
  const pathSegments = filePath.replace(/\.html?$/, "").replace(/\/index$/, "").split("/").filter(segment => segment.length > 0);
  if (pathSegments.length === 0) {
    return;
  }
  const breadcrumbItems = [];
  let currentPath = "";

  // Add home
  breadcrumbItems.push({
    "@type": "ListItem",
    position: 1,
    name: "Home",
    item: siteConfig.hostname
  });

  // Add path segments
  pathSegments.forEach((segment, index) => {
    currentPath += `/${segment}`;
    const name = segment.replace(/-/g, " ").replace(/\b\w/g, l => l.toUpperCase());
    breadcrumbItems.push({
      "@type": "ListItem",
      position: index + 2,
      name: name,
      item: `${siteConfig.hostname}${currentPath}`
    });
  });
  const breadcrumbSchema = {
    "@type": "BreadcrumbList",
    itemListElement: breadcrumbItems
  };
  schemas.push(breadcrumbSchema);
}

/**
 * Generates Organization schema
 * @param {Array} schemas - Schemas array to modify
 * @param {Object} siteConfig - Site configuration
 */
function generateOrganizationSchema(schemas, siteConfig) {
  if (!siteConfig.organization) {
    return;
  }
  const org = siteConfig.organization;
  const organizationSchema = {
    "@type": "Organization",
    name: org.name,
    url: org.url || siteConfig.hostname
  };
  if (org.logo) {
    organizationSchema.logo = {
      "@type": "ImageObject",
      url: org.logo
    };
  }
  if (org.description) {
    organizationSchema.description = org.description;
  }
  if (org.sameAs) {
    organizationSchema.sameAs = Array.isArray(org.sameAs) ? org.sameAs : [org.sameAs];
  }
  if (org.contactPoint) {
    organizationSchema.contactPoint = {
      "@type": "ContactPoint",
      telephone: org.contactPoint.telephone,
      contactType: org.contactPoint.contactType || "customer service"
    };
  }
  schemas.push(organizationSchema);
}

/**
 * Generates publisher schema for articles
 * @param {Object} organization - Organization configuration
 * @returns {Object} Publisher schema
 */
function generatePublisherSchema(organization) {
  const publisher = {
    "@type": "Organization",
    name: organization.name
  };
  if (organization.logo) {
    publisher.logo = {
      "@type": "ImageObject",
      url: organization.logo
    };
  }
  return publisher;
}

/**
 * Converts JSON-LD schemas to HTML script tag
 * @param {Array} schemas - Array of JSON-LD schemas
 * @returns {string} HTML script tag
 */
function generateJsonLdHtml(schemas) {
  if (schemas.length === 0) {
    return "";
  }

  // If multiple schemas, use @graph
  const jsonLd = schemas.length === 1 ? schemas[0] : {
    "@context": "https://schema.org",
    "@graph": schemas.map(schema => {
      // eslint-disable-next-line no-unused-vars
      const {
        "@context": _,
        ...rest
      } = schema;
      return rest;
    })
  };
  const jsonString = JSON.stringify(jsonLd, null, 2);
  return `<script type="application/ld+json">\n${jsonString}\n</script>`;
}

/**
 * @fileoverview HTML injection utilities for strategic head element placement.
 */

/**
 * Updates or creates the title tag
 * @param {string} html - The HTML content to modify
 * @param {string} title - The title content
 * @returns {string} Modified HTML content
 */
function updateTitle(html, title) {
  const $ = cheerio.load(html, {
    decodeEntities: false,
    lowerCaseAttributeNames: false
  });
  let $head = $("head");
  if ($head.length === 0) {
    $("body").before("<head></head>");
    $head = $("head");
  }
  const $title = $head.find("title");
  if ($title.length > 0) {
    $title.text(title);
  } else {
    $head.prepend(`<title>${escapeHtml(title)}</title>`);
  }
  return $.html();
}

/**
 * Adds or updates a meta tag
 * @param {string} html - The HTML content to modify
 * @param {string} name - The meta tag name or property
 * @param {string} content - The meta tag content
 * @param {string} [type='name'] - The attribute type: 'name', 'property', 'http-equiv'
 * @returns {string} Modified HTML content
 */
function updateMetaTag(html, name, content, type = "name") {
  const $ = cheerio.load(html, {
    decodeEntities: false,
    lowerCaseAttributeNames: false
  });
  let $head = $("head");
  if ($head.length === 0) {
    $("body").before("<head></head>");
    $head = $("head");
  }

  // Find existing meta tag
  const selector = `meta[${type}="${name}"]`;
  const $existing = $head.find(selector);
  if ($existing.length > 0) {
    // Update existing meta tag
    $existing.attr("content", content);
  } else {
    // Create new meta tag
    const metaTag = `<meta ${type}="${escapeHtml(name)}" content="${escapeHtml(content)}">`;

    // Insert after existing meta tags but before other elements
    const $lastMeta = $head.find("meta").last();
    if ($lastMeta.length > 0) {
      $lastMeta.after(metaTag);
    } else {
      // Insert after title if it exists, otherwise at the start
      const $title = $head.find("title");
      if ($title.length > 0) {
        $title.after(metaTag);
      } else {
        $head.prepend(metaTag);
      }
    }
  }
  return $.html();
}

/**
 * Adds or updates a link tag
 * @param {string} html - The HTML content to modify
 * @param {string} rel - The link relationship
 * @param {string} href - The link URL
 * @param {Object} [attributes={}] - Additional attributes
 * @returns {string} Modified HTML content
 */
function updateLinkTag(html, rel, href, attributes = {}) {
  const $ = cheerio.load(html, {
    decodeEntities: false,
    lowerCaseAttributeNames: false
  });
  let $head = $("head");
  if ($head.length === 0) {
    $("body").before("<head></head>");
    $head = $("head");
  }

  // Find existing link tag
  const $existing = $head.find(`link[rel="${rel}"]`);

  // Build attributes string
  const attrs = Object.entries(attributes).map(([key, value]) => `${key}="${escapeHtml(value)}"`).join(" ");
  const linkTag = `<link rel="${escapeHtml(rel)}" href="${escapeHtml(href)}"${attrs ? ` ${attrs}` : ""}>`;
  if ($existing.length > 0) {
    // Replace existing link tag
    $existing.replaceWith(linkTag);
  } else {
    // Add new link tag after other links
    const $lastLink = $head.find("link").last();
    if ($lastLink.length > 0) {
      $lastLink.after(linkTag);
    } else {
      // Insert after meta tags
      const $lastMeta = $head.find("meta").last();
      if ($lastMeta.length > 0) {
        $lastMeta.after(linkTag);
      } else {
        $head.append(linkTag);
      }
    }
  }
  return $.html();
}

/**
 * Adds a script tag to the head (for JSON-LD structured data)
 * @param {string} html - The HTML content to modify
 * @param {string} scriptContent - The script content
 * @param {string} [type='application/ld+json'] - The script type
 * @param {string} [position='end'] - Where to place the script
 * @returns {string} Modified HTML content
 */
function addScript(html, scriptContent, type = "application/ld+json", position = "end") {
  const $ = cheerio.load(html, {
    decodeEntities: false,
    lowerCaseAttributeNames: false
  });
  let $head = $("head");
  if ($head.length === 0) {
    $("body").before("<head></head>");
    $head = $("head");
  }
  const scriptTag = `<script type="${escapeHtml(type)}">${scriptContent}</script>`;
  if (position === "end") {
    $head.append(scriptTag);
  } else {
    $head.prepend(scriptTag);
  }
  return $.html();
}

/**
 * Escapes HTML special characters
 * @param {string} str - String to escape
 * @returns {string} Escaped string
 */
function escapeHtml(str) {
  if (typeof str !== "string") {
    return String(str);
  }
  return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#39;");
}

/**
 * Removes existing SEO meta tags (for clean injection)
 * @param {string} html - The HTML content to modify
 * @param {string[]} [tags] - Specific tags to remove
 * @returns {string} Modified HTML content
 */
function removeExistingMetaTags(html, tags = []) {
  const $ = cheerio.load(html, {
    decodeEntities: false,
    lowerCaseAttributeNames: false
  });
  const defaultTags = ["description", "keywords", "robots", "canonical"];
  const ogTags = ["og:title", "og:description", "og:image", "og:url", "og:type", "og:site_name"];
  const twitterTags = ["twitter:card", "twitter:title", "twitter:description", "twitter:image", "twitter:site", "twitter:creator"];
  const tagsToRemove = tags.length > 0 ? tags : [...defaultTags, ...ogTags, ...twitterTags];
  tagsToRemove.forEach(tag => {
    $(`meta[name="${tag}"]`).remove();
    $(`meta[property="${tag}"]`).remove();
  });

  // Remove existing JSON-LD scripts
  $('script[type="application/ld+json"]').remove();

  // Remove existing canonical link
  $('link[rel="canonical"]').remove();
  return $.html();
}

/**
 * @fileoverview Unified head optimizer that orchestrates all SEO metadata generation.
 */

/**
 * @typedef {Object} HeadOptimizationOptions
 * @property {string} hostname - Base hostname for generating URLs
 * @property {string} [seoProperty='seo'] - Frontmatter property containing SEO data
 * @property {Object} [defaults] - Default values for missing metadata
 * @property {Object} [fallbacks] - Fallback property mappings
 * @property {Object} [social] - Social media configuration
 * @property {Object} [jsonLd] - JSON-LD configuration
 * @property {boolean} [cleanExisting=true] - Remove existing SEO tags before injection
 * @property {boolean} [generateSitemap=true] - Whether to include files in sitemap
 * @property {number} [wordsPerMinute=200] - Reading speed for calculating reading time
 */

/**
 * @typedef {Object} OptimizationResult
 * @property {string} html - Modified HTML content
 * @property {Object} metadata - Extracted metadata
 * @property {Object} generated - Generated SEO content
 */

/**
 * Processes a file and optimizes its head section with SEO metadata
 * @param {string} filePath - File path relative to source
 * @param {Object} frontmatter - File frontmatter and content
 * @param {HeadOptimizationOptions} options - Optimization options
 * @returns {OptimizationResult} Optimization result
 */
async function optimizeHead(filePath, frontmatter, options) {
  const {
    hostname,
    seoProperty = "seo",
    defaults = {},
    fallbacks = {},
    social = {},
    jsonLd = {},
    cleanExisting = true,
    generateSitemap = true,
    wordsPerMinute = 200
  } = options;

  // Skip non-HTML files
  if (!isHtmlFile(filePath) || !Buffer.isBuffer(frontmatter.contents)) {
    return {
      html: frontmatter.contents.toString(),
      metadata: null,
      generated: null
    };
  }

  // Extract unified metadata
  const metadata = extractMetadata(filePath, frontmatter, {
    hostname,
    seoProperty,
    defaults,
    fallbacks,
    wordsPerMinute
  });

  // Check if file should be excluded from SEO processing
  if (metadata.noIndex && !generateSitemap) {
    return {
      html: frontmatter.contents.toString(),
      metadata,
      generated: null
    };
  }

  // Generate all SEO content
  const generated = await generateAllSeoContent(metadata, {
    hostname,
    social,
    jsonLd,
    filePath
  });

  // Inject SEO content into HTML
  let html = frontmatter.contents.toString();
  html = await injectSeoContent(html, metadata, generated, {
    cleanExisting
  });
  return {
    html,
    metadata,
    generated
  };
}

/**
 * Generates all SEO content from metadata
 * @param {Object} metadata - Extracted metadata
 * @param {Object} config - Generation configuration
 * @returns {Object} Generated SEO content
 */
function generateAllSeoContent(metadata, config) {
  const {
    hostname,
    social,
    jsonLd,
    filePath
  } = config;

  // Site configuration combining hostname with social/jsonLd configs
  const siteConfig = {
    hostname,
    ...social,
    ...jsonLd
  };

  // Generate meta tags
  const metaResult = generateMetaTags(metadata, siteConfig);

  // Generate Open Graph tags
  const ogResult = generateOpenGraphTags(metadata, siteConfig);

  // Generate Twitter Card tags
  const twitterResult = generateTwitterCardTags(metadata, siteConfig);

  // Generate JSON-LD structured data
  const jsonLdResult = generateJsonLd(metadata, siteConfig, filePath);
  return {
    meta: metaResult,
    openGraph: ogResult,
    twitter: twitterResult,
    jsonLd: jsonLdResult
  };
}

/**
 * Injects SEO content into HTML head section
 * @param {string} html - Original HTML content
 * @param {Object} metadata - Extracted metadata
 * @param {Object} generated - Generated SEO content
 * @param {Object} options - Injection options
 * @returns {string} Modified HTML content
 */
function injectSeoContent(html, metadata, generated, options = {}) {
  const {
    cleanExisting = true
  } = options;
  let modifiedHtml = html;

  // Clean existing SEO tags if requested
  if (cleanExisting) {
    modifiedHtml = removeExistingMetaTags(modifiedHtml);
  }

  // Update title tag
  if (metadata.title) {
    modifiedHtml = updateTitle(modifiedHtml, metadata.title);
  }

  // Inject meta tags (critical tags first)
  const criticalMetaTags = getCriticalMetaTags(generated.meta.metaTags);
  const otherMetaTags = getOtherMetaTags(generated.meta.metaTags);

  // Inject critical meta tags early
  for (const tag of criticalMetaTags) {
    if (tag.name) {
      modifiedHtml = updateMetaTag(modifiedHtml, tag.name, tag.content, "name");
    } else if (tag.httpEquiv) {
      modifiedHtml = updateMetaTag(modifiedHtml, tag.httpEquiv, tag.content, "http-equiv");
    }
  }

  // Inject link tags (after critical meta tags)
  for (const link of generated.meta.linkTags) {
    modifiedHtml = updateLinkTag(modifiedHtml, link.rel, link.href, Object.fromEntries(Object.entries(link).filter(([key]) => !["rel", "href"].includes(key))));
  }

  // Inject remaining meta tags
  for (const tag of otherMetaTags) {
    if (tag.name) {
      modifiedHtml = updateMetaTag(modifiedHtml, tag.name, tag.content, "name");
    }
  }

  // Inject Open Graph tags
  for (const tag of generated.openGraph.metaTags) {
    modifiedHtml = updateMetaTag(modifiedHtml, tag.property, tag.content, "property");
  }

  // Inject Twitter Card tags
  for (const tag of generated.twitter.metaTags) {
    modifiedHtml = updateMetaTag(modifiedHtml, tag.name, tag.content, "name");
  }

  // Inject JSON-LD structured data (at the end for optimal loading)
  if (generated.jsonLd.html) {
    const jsonLdContent = generated.jsonLd.html.replace(/<script[^>]*>|<\/script>/g, "");
    modifiedHtml = addScript(modifiedHtml, jsonLdContent, "application/ld+json", "end");
  }
  return modifiedHtml;
}

/**
 * Determines critical meta tags that should be injected early
 * @param {Array} metaTags - All meta tags
 * @returns {Array} Critical meta tags
 */
function getCriticalMetaTags(metaTags) {
  const criticalTags = ["viewport", "charset", "description", "robots"];
  return metaTags.filter(tag => criticalTags.includes(tag.name) || criticalTags.includes(tag.httpEquiv));
}

/**
 * Gets non-critical meta tags
 * @param {Array} metaTags - All meta tags
 * @returns {Array} Non-critical meta tags
 */
function getOtherMetaTags(metaTags) {
  const criticalTags = ["viewport", "charset", "description", "robots"];
  return metaTags.filter(tag => !criticalTags.includes(tag.name) && !criticalTags.includes(tag.httpEquiv));
}

/**
 * Checks if a file is an HTML file that should be processed
 * @param {string} filePath - File path
 * @returns {boolean} Whether file should be processed
 */
function isHtmlFile(filePath) {
  return /\.html?$/i.test(filePath);
}

/**
 * Batch processes multiple files for SEO optimization
 * @param {Object} files - Metalsmith files object
 * @param {HeadOptimizationOptions} options - Optimization options
 * @returns {Promise<Object>} Results for each processed file
 */
async function batchOptimizeHeads(files, options) {
  const results = {};
  const fileList = Object.keys(files);

  // Process files in parallel batches
  const batchSize = options.batchSize || 10;
  for (let i = 0; i < fileList.length; i += batchSize) {
    const batch = fileList.slice(i, i + batchSize);
    const batchPromises = batch.map(async filePath => {
      try {
        const result = await optimizeHead(filePath, files[filePath], options);

        // Update file contents if HTML was modified
        if (result.html !== files[filePath].contents.toString()) {
          files[filePath].contents = Buffer.from(result.html);
        }

        // Store metadata for potential use by other plugins
        if (result.metadata) {
          files[filePath].seoMetadata = result.metadata;
        }
        return {
          filePath,
          result
        };
      } catch (error) {
        console.error(`SEO optimization failed for ${filePath}:`, error);
        return {
          filePath,
          error
        };
      }
    });
    const batchResults = await Promise.all(batchPromises);

    // Collect results
    batchResults.forEach(({
      filePath,
      result,
      error
    }) => {
      results[filePath] = error ? {
        error
      } : result;
    });
  }
  return results;
}

/**
 * @fileoverview URL building and validation utilities for sitemap generation.
 */

/**
 * Determines whether a file should be included in the sitemap.
 * Files are included if they match the pattern and are not marked as private.
 * @param {string} file - File path relative to source directory
 * @param {Object} frontmatter - File metadata and frontmatter
 * @param {Object} metalsmith - Metalsmith instance
 * @param {string} pattern - Glob pattern to match files
 * @param {string} privateProperty - Property name to check if file should be excluded
 * @returns {boolean} True if file should be processed, false otherwise
 */
function checkFile(file, frontmatter, metalsmith, pattern, privateProperty) {
  // Only process files that match the pattern
  const matchResult = metalsmith.match(pattern, file);
  if (!matchResult || matchResult.length === 0) {
    return false;
  }

  // Don't process private files
  if (get(frontmatter, privateProperty)) {
    return false;
  }
  return true;
}

/**
 * Constructs the final URL for a file based on configuration options.
 * Handles canonical URL overrides and index file omission.
 * @param {string} file - File path relative to source directory
 * @param {Object} frontmatter - File metadata and frontmatter
 * @param {Object} options - URL building options
 * @param {string} options.urlProperty - Property name to read canonical URL from file metadata
 * @param {boolean} options.omitIndex - Whether to omit index.html from URLs
 * @returns {string} Final URL for the sitemap entry
 */
function buildUrl(file, frontmatter, options) {
  const {
    urlProperty,
    omitIndex
  } = options;

  // Frontmatter settings take precedence
  const canonicalUrl = get(frontmatter, urlProperty);
  if (typeof canonicalUrl === "string") {
    return canonicalUrl;
  }

  // Remove index.html if necessary
  const indexFile = "index.html";
  if (omitIndex && path__default["default"].basename(file) === indexFile) {
    return replaceBackslash(file.slice(0, 0 - indexFile.length));
  }

  // Otherwise just use 'file'
  return replaceBackslash(file);
}

/**
 * Normalizes file paths by replacing backslashes with forward slashes.
 * Ensures cross-platform compatibility for URLs.
 * @param {string} url - File path that may contain backslashes
 * @returns {string} Normalized path with forward slashes
 */
function replaceBackslash(url) {
  return url.replace(/\\/g, "/");
}

/**
 * @fileoverview SEO-focused priority and changefreq calculation utilities.
 *
 * Uses sensible defaults based on URL hierarchy and content type,
 * while avoiding flawed assumptions about content length or complex patterns.
 */

/**
 * Calculates SEO-focused priority based on URL hierarchy and content type.
 * Avoids flawed assumptions about content length or complex pattern matching.
 *
 * @param {string} file - File path relative to source directory
 * @param {Object} frontmatter - File metadata and frontmatter
 * @param {Object} options - Calculation options
 * @returns {number} Calculated priority between 0.1 and 1.0
 */
function calculatePriority(file) {
  // Homepage gets highest priority
  if (file === "index.html" || file === "index.htm") {
    return 1.0;
  }

  // URL depth - shallower pages are generally more important
  const pathDepth = file.split(path__default["default"].sep).length;
  let priority;
  if (pathDepth === 1) {
    priority = 0.8; // Root level pages
  } else if (pathDepth === 2) {
    priority = 0.6; // One level deep
  } else if (pathDepth === 3) {
    priority = 0.4; // Two levels deep
  } else {
    priority = 0.3; // Deeper pages
  }

  // Section-level index pages get priority boost
  if (file.endsWith("index.html") && pathDepth > 1) {
    priority = Math.min(priority + 0.2, 1.0);
  }
  return Math.min(Math.max(priority, 0.1), 1.0);
}

/**
 * Calculates sensible change frequency based on URL hierarchy and content type.
 * @param {string} file - File path relative to source directory
 * @param {Object} frontmatter - File metadata and frontmatter
 * @param {Object} options - Calculation options
 * @returns {string} Calculated change frequency
 */
function calculateChangefreq(file, frontmatter, options) {
  // Homepage and main index pages change frequently
  if (file === "index.html" || file === "index.htm") {
    return "weekly";
  }

  // Section index pages update when new content is added
  if (file.endsWith("index.html")) {
    return "monthly";
  }

  // Check for last modified date to make informed decisions
  const {
    modifiedProperty,
    lastmod
  } = options;
  const fileLastmod = get(frontmatter, modifiedProperty) || lastmod;
  if (fileLastmod) {
    const modDate = fileLastmod instanceof Date ? fileLastmod : new Date(fileLastmod);
    const now = new Date();
    const daysSinceModified = (now - modDate) / (1000 * 60 * 60 * 24);
    if (daysSinceModified < 30) {
      return "monthly"; // Recently modified content
    } else if (daysSinceModified < 365) {
      return "yearly"; // Older content
    }
  }

  // Default for most content pages
  return "yearly";
}

/**
 * @fileoverview XML generation utilities for sitemap creation.
 *
 * This module contains pure functions for generating XML content, separated from
 * the main sitemap processing logic for better maintainability and testability.
 *
 * ## Security Focus
 *
 * All XML generation functions in this module prioritize security through proper
 * character escaping to prevent XML injection attacks and ensure valid output.
 */

/**
 * Escapes special XML characters to prevent XML injection and ensure valid output.
 *
 * This function implements the five standard XML character entity references
 * as defined in the XML 1.0 specification. It's critical for security and
 * correctness when embedding user-controlled content in XML.
 *
 * ## Security Considerations
 *
 * XML character escaping prevents:
 * - XML injection attacks where malicious XML could be injected
 * - XML parsing errors from unescaped special characters
 * - XSS attacks if the XML is later processed in a web context
 *
 * ## Performance Notes
 *
 * This implementation uses simple string replacement rather than regex for
 * better performance with large strings. The order of replacements is important:
 * ampersand (&) must be escaped first to avoid double-escaping.
 *
 * @param {string} str - String to escape for safe XML inclusion
 * @returns {string} XML-safe string with special characters escaped, empty string if input is not a string
 *
 * @example
 * escapeXML('AT&T <mobile> "quotes" & \'apostrophes\'');
 * // Returns: 'AT&amp;T &lt;mobile&gt; &quot;quotes&quot; &amp; &apos;apostrophes&apos;'
 *
 * @example
 * escapeXML(null); // Returns: ''
 * escapeXML(123);  // Returns: ''
 */
function escapeXML(str) {
  if (typeof str !== "string") {
    return "";
  }
  return str.replace(/&/g, "&amp;") // Must be first to avoid double-escaping
  .replace(/</g, "&lt;") // Less-than sign
  .replace(/>/g, "&gt;") // Greater-than sign
  .replace(/"/g, "&quot;") // Double quote
  .replace(/'/g, "&apos;"); // Apostrophe/single quote
}

/**
 * Generates complete XML sitemap content from processed link data.
 *
 * This is the core XML generation function that builds a standards-compliant
 * sitemap.xml file. It handles all the formatting details and edge cases
 * to ensure the output matches the behavior of the original sitemap library.
 *
 * ## XML Structure
 *
 * The generated XML follows the sitemaps.org protocol structure:
 * ```xml
 * <?xml version="1.0" encoding="UTF-8"?>
 * <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
 *   <url>
 *     <loc>https://example.com/page</loc>
 *     <lastmod>2024-01-15T00:00:00.000Z</lastmod>
 *     <changefreq>weekly</changefreq>
 *     <priority>0.8</priority>
 *     <xhtml:link rel="alternate" hreflang="fr" href="https://example.com/fr/page"/>
 *   </url>
 * </urlset>
 * ```
 *
 * ## URL Handling Logic
 *
 * URLs are processed with the following logic:
 * 1. **Absolute URLs**: Used as-is if they start with 'http'
 * 2. **Relative URLs**: Combined with hostname, handling trailing/leading slashes
 * 3. **Path normalization**: Ensures no double slashes in the final URL
 *
 * ## Element Generation Rules
 *
 * - **`<loc>`**: Always present, contains the full absolute URL
 * - **`<lastmod>`**: Only included if lastmod value exists and is valid
 * - **`<changefreq>`**: Only included if changefreq is specified
 * - **`<priority>`**: Only included if priority is not undefined/null
 * - **`<xhtml:link>`**: Only included for entries with alternate language links
 *
 * ## Namespace Handling
 *
 * The implementation uses minimal namespace declarations to match the original
 * library behavior. The xhtml namespace is not declared in the root element
 * but is used directly in xhtml:link elements (browsers handle this correctly).
 *
 * ## Performance Optimizations
 *
 * - String concatenation is used instead of DOM manipulation for speed
 * - Single pass through the links array
 * - Minimal string allocations
 * - No external dependencies or complex parsing
 *
 * @param {Array<Object>} links - Array of processed URL entries with metadata
 * @param {string} links[].url - Relative or absolute URL for the page
 * @param {string} [links[].lastmod] - ISO date string for last modification
 * @param {string} [links[].changefreq] - Change frequency (daily, weekly, etc.)
 * @param {number|string} [links[].priority] - Priority value (0.0 to 1.0)
 * @param {Array<Object>} [links[].links] - Array of alternate language links
 * @param {string} links[].links[].lang - Language code (e.g., 'en', 'fr')
 * @param {string} links[].links[].url - Absolute URL for the alternate version
 * @param {string} hostname - Base hostname for relative URL resolution (e.g., 'https://example.com')
 * @returns {string} Complete XML sitemap content ready for file output
 *
 * @example
 * // Basic sitemap with simple URLs
 * const links = [
 *   { url: 'index.html', priority: 1.0, changefreq: 'daily' },
 *   { url: 'about.html', priority: 0.8, lastmod: '2024-01-15T00:00:00.000Z' }
 * ];
 * const xml = generateSitemapXML(links, 'https://example.com');
 *
 * @example
 * // Advanced sitemap with alternate language links
 * const links = [
 *   {
 *     url: 'index.html',
 *     priority: 1.0,
 *     links: [
 *       { lang: 'en', url: 'https://example.com/' },
 *       { lang: 'fr', url: 'https://example.com/fr/' }
 *     ]
 *   }
 * ];
 * const xml = generateSitemapXML(links, 'https://example.com');
 */
function generateSitemapXML(links, hostname) {
  // Start building the XML - always use basic namespace like the old library
  let xml = '<?xml version="1.0" encoding="UTF-8"?>';
  xml += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">';

  // Process each URL entry
  for (const link of links) {
    xml += "<url>";

    // Build the complete URL with proper hostname handling
    const fullUrl = link.url.startsWith("http") ? link.url : `${hostname.replace(/\/$/, "")}/${link.url.replace(/^\//, "")}`;
    xml += `<loc>${escapeXML(fullUrl)}</loc>`;

    // Add optional lastmod element
    if (link.lastmod) {
      xml += `<lastmod>${escapeXML(String(link.lastmod))}</lastmod>`;
    }

    // Add optional changefreq element
    if (link.changefreq) {
      xml += `<changefreq>${escapeXML(String(link.changefreq))}</changefreq>`;
    }

    // Add optional priority element with proper decimal formatting
    if (link.priority !== undefined && link.priority !== null) {
      // Ensure priority is formatted with decimal point (e.g., 1.0 not 1)
      const priorityStr = typeof link.priority === "number" ? link.priority.toFixed(1) : String(link.priority);
      xml += `<priority>${escapeXML(priorityStr)}</priority>`;
    }

    // Add alternate language links (hreflang) if present
    if (link.links && Array.isArray(link.links)) {
      for (const altLink of link.links) {
        if (altLink.lang && altLink.url) {
          xml += `<xhtml:link rel="alternate" hreflang="${escapeXML(altLink.lang)}" href="${escapeXML(altLink.url)}"/>`;
        }
      }
    }
    xml += "</url>";
  }
  xml += "</urlset>";
  return xml;
}

/*
 * ## Implementation Notes & Design Decisions
 *
 * This XML generation module was extracted from the main sitemap processor to:
 * - Improve maintainability by separating concerns
 * - Enable easier testing of XML generation logic
 * - Keep the main sitemap processor focused on file processing
 * - Allow reuse of XML utilities in other contexts if needed
 *
 * ### Why Not Use a Template Engine?
 *
 * Initially, we considered using template engines like Nunjucks for XML generation.
 * However, for sitemap generation, direct string building proved superior because:
 *
 * - **Simplicity**: Sitemaps have a straightforward, predictable structure
 * - **Performance**: String concatenation is faster than template compilation
 * - **Dependencies**: Avoids adding another external dependency
 * - **Control**: Full control over output formatting and edge cases
 * - **Size**: Keeps the implementation lightweight and focused
 *
 * ### Why Not Use XML Libraries?
 *
 * XML libraries (like xmlbuilder, xml2js) were avoided because:
 * - They add significant overhead for our simple use case
 * - Sitemap XML structure is static and predictable
 * - Manual escaping gives us explicit control over security
 * - Performance is better with direct string building
 * - Reduces bundle size and dependency complexity
 *
 * ### Security Considerations
 *
 * This module prioritizes security through:
 * - Proper XML character escaping in all output
 * - Input validation and type checking
 * - No use of dangerous functions like eval() or innerHTML
 * - Clear separation of data and markup
 *
 * ### Future Maintenance
 *
 * If you need to modify this XML generation:
 * 1. Always escape user-controlled content with escapeXML()
 * 2. Test with real XML validators after changes
 * 3. Maintain compatibility with existing test expectations
 * 4. Consider security implications of any new XML features
 * 5. Keep functions pure (no side effects) for easier testing
 */

/**
 * @fileoverview Lightweight sitemap processor for generating sitemap.xml files.
 *
 * Sitemaps are fundamentally simple XML files with a well-defined structure.
 * Rather than using a heavy external library with streaming APIs and complex
 * features we don't need, this implementation generates XML directly using
 * string concatenation.
 *
 * ## XML Format Compliance
 *
 * This implementation generates valid XML sitemaps according to the sitemaps.org protocol:
 * - Standard XML declaration with UTF-8 encoding
 * - Proper namespace declarations for sitemap and xhtml (when needed)
 * - XML character escaping for security (prevents XXE and other XML attacks)
 * - Support for all standard sitemap elements: loc, lastmod, changefreq, priority
 * - Support for alternate language links (hreflang) via xhtml:link elements
 *
 * ## Attribution
 *
 * Inspired by and adapted from metalsmith-sitemap by Segment.io
 * Original: https://github.com/segmentio/metalsmith-sitemap
 * License: MIT
 */

/**
 * @typedef {Object} SitemapOptions
 * @property {string} hostname - Base hostname for all URLs in the sitemap
 * @property {Date|string} [lastmod] - Default last modified date for all files
 * @property {string} [links] - Property name to read additional links from file metadata
 * @property {'always'|'hourly'|'daily'|'weekly'|'monthly'|'yearly'|'never'} [changefreq] - Default change frequency
 * @property {boolean} [omitIndex] - Whether to omit index.html from URLs
 * @property {string} [output='sitemap.xml'] - Output filename for the sitemap
 * @property {string} [pattern] - Glob pattern to match files for inclusion (default: all HTML files)
 * @property {string|number} [priority] - Default priority for all URLs (0.0 to 1.0)
 * @property {string} [urlProperty='canonical'] - Property name to read canonical URL from file metadata
 * @property {string} [modifiedProperty='lastmod'] - Property name to read last modified date from file metadata
 * @property {string} [privateProperty='private'] - Property name to check if file should be excluded
 * @property {string} [priorityProperty='priority'] - Property name to read priority from file metadata
 * @property {boolean} [auto=false] - Enable automatic priority and changefreq calculation based on content analysis
 */

/**
 * Processes Metalsmith files and generates a sitemap.xml file.
 *
 * This is the main entry point for sitemap generation. It:
 * 1. Filters files based on pattern matching and privacy settings
 * 2. Extracts metadata (lastmod, changefreq, priority) from frontmatter
 * 3. Handles automatic priority/changefreq calculation if enabled
 * 4. Processes alternate language links (hreflang)
 * 5. Generates the final XML content and adds it to the files object
 *
 * ## File Processing Logic
 *
 * Files are processed in a specific order to ensure consistent output:
 * - Root level files first (depth 1), then subdirectories
 * - Alphabetical sorting within the same directory depth
 * - Only processes files with Buffer contents (skips virtual files)
 * - Respects pattern matching and privacy exclusions
 *
 * ## Date Handling
 *
 * All dates are normalized to midnight UTC. This ensures consistent output
 * regardless of timezone:
 * - JavaScript Date objects: Converted to ISO string with time set to 00:00:00.000Z
 * - String dates: Parsed and normalized to midnight UTC
 * - Invalid dates: Skipped (no lastmod element generated)
 *
 * ## Priority Formatting
 *
 * Numeric priorities are formatted with one decimal place to match original behavior:
 * - 1 becomes "1.0"
 * - 0.5 becomes "0.5"
 * - String priorities are passed through unchanged
 *
 * ## Auto-calculation Features
 *
 * When auto=true, priority and changefreq are calculated based on:
 * - File path depth (shallower = higher priority)
 * - Content type patterns (/blog/, /products/, etc.)
 * - Last modification date (recent = higher priority)
 * - Content length (longer = slightly higher priority)
 *
 * @param {Object} files - Metalsmith files object containing all site files
 * @param {Object} metalsmith - Metalsmith instance for utilities like pattern matching
 * @param {SitemapOptions} options - Configuration options for sitemap generation
 * @returns {Promise<void>} Resolves when sitemap has been generated and added to files
 *
 * @example
 * // Basic usage
 * await processSitemap(files, metalsmith, {
 *   hostname: 'https://example.com',
 *   pattern: '**\/*.html'
 * });
 *
 * @example
 * // With auto-calculation and custom settings
 * await processSitemap(files, metalsmith, {
 *   hostname: 'https://example.com',
 *   auto: true,
 *   output: 'custom-sitemap.xml',
 *   omitIndex: true,
 *   links: 'alternateLinks' // Property name for hreflang links
 * });
 */
function processSitemap(files, metalsmith, options) {
  return new Promise((resolve, reject) => {
    try {
      const {
        auto = false,
        changefreq,
        hostname,
        lastmod,
        links: linksOption,
        omitIndex,
        output = "sitemap.xml",
        pattern = "**/*.html",
        priority,
        urlProperty = "canonical",
        modifiedProperty = "lastmod",
        privateProperty = "private",
        priorityProperty = "priority"
      } = options;
      const links = [];

      // Custom sorting to ensure consistent output order
      // Files in root directory come first, then subdirectories
      const sortedFiles = Object.keys(files).sort((a, b) => {
        const aDepth = a.split(path__default["default"].sep).length;
        const bDepth = b.split(path__default["default"].sep).length;
        if (aDepth !== bDepth) {
          return aDepth - bDepth; // Shallower paths first
        }
        return a.localeCompare(b); // Alphabetical within same depth
      });
      sortedFiles.forEach(function (file) {
        // Get the current file's frontmatter
        const frontmatter = files[file];

        // Validate file.contents is a Buffer before processing
        if (!Buffer.isBuffer(frontmatter.contents)) {
          return;
        }

        // Only process files that pass the check
        if (!checkFile(file, frontmatter, metalsmith, pattern, privateProperty)) {
          return;
        }

        // Get lastmod value and format it properly
        let lastmodValue = get(frontmatter, modifiedProperty) || lastmod;
        if (lastmodValue instanceof Date) {
          // Format date as ISO string to match old library behavior
          // The old sitemap library normalized times to midnight UTC
          const d = new Date(lastmodValue);
          d.setUTCHours(0, 0, 0, 0);
          lastmodValue = d.toISOString();
        } else if (typeof lastmodValue === "string") {
          // Parse the date string
          const parsed = new Date(lastmodValue);
          if (!isNaN(parsed.getTime())) {
            // Normalize to midnight UTC like the old library did
            parsed.setUTCHours(0, 0, 0, 0);
            lastmodValue = parsed.toISOString();
          }
        }

        // Create the sitemap entry (reject keys with falsy values)
        let entryChangefreq, entryPriority;
        if (auto) {
          /**
           * Auto mode
           * Calculate values, ignore global and frontmatter settings!
           */
          entryChangefreq = calculateChangefreq(file, frontmatter, {
            modifiedProperty,
            lastmod
          });
          entryPriority = calculatePriority(file);
        } else {
          /**
           * Manual mode
           * Use global defaults, allow frontmatter overrides
           */
          entryChangefreq = frontmatter.changefreq || changefreq;
          entryPriority = get(frontmatter, priorityProperty) || priority;
        }
        const entry = Object.fromEntries(Object.entries({
          changefreq: entryChangefreq,
          priority: entryPriority,
          lastmod: lastmodValue,
          links: linksOption ? get(frontmatter, linksOption) : undefined
        }).filter(([value]) => value !== undefined));

        // Add the url (which is allowed to be falsy)
        entry.url = buildUrl(file, frontmatter, {
          urlProperty,
          omitIndex
        });

        // Add the entry to the links array
        links.push(entry);
      });

      // Generate sitemap XML content
      const sitemapContent = generateSitemapXML(links, hostname);

      // Add the sitemap file to the files object
      files[output] = {
        contents: Buffer.from(sitemapContent, "utf-8")
      };
      resolve();
    } catch (error) {
      reject(new Error(`Failed to generate sitemap: ${error.message}`));
    }
  });
}

/**
 * @fileoverview Robots.txt generation and coordination for Metalsmith SEO
 * Handles existing robots.txt files intelligently and generates basic ones when needed
 */

/**
 * @typedef {Object} RobotsOptions
 * @property {string} hostname - Base hostname for sitemap URL
 * @property {string} [sitemapFile='sitemap.xml'] - Name of the sitemap file to reference
 * @property {boolean} [generateRobots=true] - Whether to generate robots.txt if none exists
 * @property {boolean} [addSitemapReference=true] - Whether to add sitemap reference to existing robots.txt
 * @property {Array<string>} [disallowPaths=[]] - Additional paths to disallow
 * @property {string} [userAgent='*'] - User agent for robots directives
 */

/**
 * Processes robots.txt file - generates if missing, updates if needed
 * @param {Object} files - Metalsmith files object
 * @param {Object} metalsmith - Metalsmith instance
 * @param {RobotsOptions} options - Configuration options
 * @returns {void}
 */
function processRobots(files, metalsmith, options) {
  const {
    hostname,
    sitemapFile = "sitemap.xml",
    generateRobots = true,
    addSitemapReference = true,
    disallowPaths = [],
    userAgent = "*"
  } = options;
  const robotsFile = "robots.txt";
  const existingRobots = files[robotsFile];
  // Ensure hostname is a string
  const hostnameStr = String(hostname || "");
  const sitemapUrl = `${hostnameStr.replace(/\/$/, "")}/${sitemapFile}`;
  if (existingRobots) {
    // Handle existing robots.txt file
    if (addSitemapReference) {
      const content = existingRobots.contents.toString();

      // Check if sitemap is already referenced
      if (!content.includes("Sitemap:") && !content.includes("sitemap:")) {
        // Add sitemap reference
        const updatedContent = `${content.trim()}\n\nSitemap: ${sitemapUrl}\n`;
        existingRobots.contents = Buffer.from(updatedContent);

        // Only log in non-test environments
        const isTest = process.env.NODE_ENV === "test" || process.env.METALSMITH_ENV === "test";
        if (!isTest) {
          console.warn("[metalsmith-seo] Added sitemap reference to existing robots.txt");
        }
      }
    }
  } else if (generateRobots) {
    // Generate basic robots.txt file
    const robotsContent = generateBasicRobots({
      userAgent,
      disallowPaths,
      sitemapUrl
    });
    files[robotsFile] = {
      contents: Buffer.from(robotsContent),
      mode: "0644"
    };

    // Only log in non-test environments
    const isTest = process.env.NODE_ENV === "test" || process.env.METALSMITH_ENV === "test";
    if (!isTest) {
      console.warn("[metalsmith-seo] Generated robots.txt with sitemap reference");
    }
  }
}

/**
 * Generates basic robots.txt content using a simple template
 * @param {Object} options - Generation options
 * @param {string} options.userAgent - User agent directive
 * @param {Array<string>} options.disallowPaths - Paths to disallow
 * @param {string} options.sitemapUrl - Full URL to sitemap
 * @returns {string} Robots.txt content
 */
function generateBasicRobots({
  userAgent,
  disallowPaths,
  sitemapUrl
}) {
  const disallowDirectives = disallowPaths.length > 0 ? disallowPaths.map(path => `Disallow: ${path}`).join("\n") : "Disallow:";
  return `User-agent: ${userAgent}
${disallowDirectives}

Sitemap: ${sitemapUrl}
`;
}

/**
 * @fileoverview Configuration building utilities for the SEO plugin.
 * Extracts and merges configuration from multiple sources with proper fallback handling.
 * @author Werner Glinka
 */

/**
 * Get nested property from an object using dot notation path
 * @param {Object} obj - The object to query
 * @param {string} path - The path to the property (e.g., 'site' or 'data.site')
 * @returns {*} The value at the path, or undefined if not found
 */
function getNestedProperty(obj, path) {
  if (!path || !obj) {
    return undefined;
  }
  const parts = path.split(".");
  let current = obj;
  for (const part of parts) {
    if (current === null || current === undefined || typeof current !== "object") {
      return undefined;
    }
    current = current[part];
  }
  return current;
}

/**
 * Helper to conditionally add properties to an object
 * @param {Object} source - Source object to read from
 * @param {Object} mapping - Mapping of source keys to target keys
 * @returns {Object} Object with only defined properties
 */
function pickDefined(source, mapping) {
  const result = {};
  for (const [sourceKey, targetKey] of Object.entries(mapping)) {
    if (source[sourceKey] !== undefined) {
      result[targetKey] = source[sourceKey];
    }
  }
  return result;
}

/**
 * Attempts to deduce site metadata from the index file when no site metadata exists
 * @param {Object} files - Metalsmith files object
 * @param {string} seoProperty - The SEO property name to check
 * @param {Object} pluginDefaults - Plugin default values
 * @returns {Object} Object with deduced siteName and description
 */
function deduceSiteMetadata(files, seoProperty, pluginDefaults) {
  let deducedSiteName = "";
  let deducedDescription = "";
  const indexFile = files["index.html"] || files["index.md"];
  if (indexFile) {
    // Deduce site name if no plugin default title exists
    if (!(pluginDefaults != null && pluginDefaults.title)) {
      var _indexFile$seoPropert;
      // Check multiple possible locations for title in order of preference:
      // 1. [seoProperty].title (SEO-specific title, e.g., seo.title)
      // 2. title (root-level title)
      deducedSiteName = ((_indexFile$seoPropert = indexFile[seoProperty]) == null ? void 0 : _indexFile$seoPropert.title) || indexFile.title;
    }

    // Deduce description if no plugin default description exists
    if (!(pluginDefaults != null && pluginDefaults.description)) {
      var _indexFile$seoPropert2;
      // Check multiple possible locations for description in order of preference:
      // 1. [seoProperty].description (SEO-specific description, e.g., seo.description)
      // 2. description (root-level description)
      // 3. excerpt (common in blog posts)
      deducedDescription = ((_indexFile$seoPropert2 = indexFile[seoProperty]) == null ? void 0 : _indexFile$seoPropert2.description) || indexFile.description || indexFile.excerpt;
    }
  }
  return {
    deducedSiteName,
    deducedDescription
  };
}

/**
 * Builds the complete configuration object by merging plugin options, site metadata, and deduced values
 * @param {Object} pluginOptions - Options passed to the plugin
 * @param {Object} siteMetadata - Site metadata from Metalsmith
 * @param {Object} files - Metalsmith files object
 * @param {string} seoProperty - The SEO property name
 * @returns {Object} Complete merged configuration
 */
function buildConfig(pluginOptions, siteMetadata, files, seoProperty) {
  var _pluginOptions$defaul, _pluginOptions$defaul2, _pluginOptions$defaul3, _pluginOptions$defaul4, _pluginOptions$social, _pluginOptions$social2, _pluginOptions$social3, _pluginOptions$social4, _pluginOptions$jsonLd;
  const hasSiteMetadata = Object.keys(siteMetadata).length > 0;
  let deducedSiteName = "";
  let deducedDescription = "";

  // Attempt to deduce site name and description when no site metadata exists
  // and no plugin defaults are provided. These fallbacks help provide reasonable
  // values for social media and SEO tags
  if (!hasSiteMetadata) {
    const deduced = deduceSiteMetadata(files, seoProperty, pluginOptions.defaults);
    deducedSiteName = deduced.deducedSiteName;
    deducedDescription = deduced.deducedDescription;
  }

  // Merge configurations: deduced < site defaults < plugin options
  const config = {
    // Site-wide defaults from site.json or deduced
    hostname: siteMetadata.url || pluginOptions.hostname,
    defaults: {
      title: ((_pluginOptions$defaul = pluginOptions.defaults) == null ? void 0 : _pluginOptions$defaul.title) || siteMetadata.title || deducedSiteName,
      description: ((_pluginOptions$defaul2 = pluginOptions.defaults) == null ? void 0 : _pluginOptions$defaul2.description) || siteMetadata.description || deducedDescription,
      socialImage: ((_pluginOptions$defaul3 = pluginOptions.defaults) == null ? void 0 : _pluginOptions$defaul3.socialImage) || siteMetadata.socialImage || siteMetadata.defaultImage || "",
      siteOwner: ((_pluginOptions$defaul4 = pluginOptions.defaults) == null ? void 0 : _pluginOptions$defaul4.siteOwner) || siteMetadata.siteOwner || "",
      ...(pluginOptions.defaults || {})
    },
    social: {
      siteName: ((_pluginOptions$social = pluginOptions.social) == null ? void 0 : _pluginOptions$social.siteName) || siteMetadata.name || siteMetadata.title || deducedSiteName,
      locale: ((_pluginOptions$social2 = pluginOptions.social) == null ? void 0 : _pluginOptions$social2.locale) || siteMetadata.locale || "en_US",
      twitterSite: ((_pluginOptions$social3 = pluginOptions.social) == null ? void 0 : _pluginOptions$social3.twitterSite) || siteMetadata.twitter,
      facebookAppId: ((_pluginOptions$social4 = pluginOptions.social) == null ? void 0 : _pluginOptions$social4.facebookAppId) || siteMetadata.facebookAppId,
      ...(siteMetadata.social || {}),
      ...(pluginOptions.social || {})
    },
    jsonLd: {
      organization: ((_pluginOptions$jsonLd = pluginOptions.jsonLd) == null ? void 0 : _pluginOptions$jsonLd.organization) || siteMetadata.organization,
      ...(siteMetadata.jsonLd || {}),
      ...(pluginOptions.jsonLd || {})
    },
    // Plugin-specific options (not typically in site.json)
    seoProperty,
    enableSitemap: pluginOptions.enableSitemap !== undefined ? pluginOptions.enableSitemap : true,
    enableRobots: pluginOptions.enableRobots !== undefined ? pluginOptions.enableRobots : true,
    batchSize: pluginOptions.batchSize || 10,
    // Fallback mappings
    fallbacks: pluginOptions.fallbacks || {
      title: "title",
      description: "excerpt",
      image: "featured_image",
      author: "author"
    },
    // Complete sitemap configuration - merge all sources once with defaults
    // Priority: pluginOptions.sitemap > pluginOptions (legacy) > siteMetadata.sitemap > defaults
    sitemap: {
      // Defaults
      output: "sitemap.xml",
      pattern: "**/*.html",
      omitIndex: false,
      urlProperty: "canonical",
      modifiedProperty: "lastmod",
      privateProperty: "private",
      priorityProperty: "priority",
      auto: true,
      // Merge site metadata
      ...(siteMetadata.sitemap || {}),
      // Apply legacy root-level options (for backward compatibility)
      ...pickDefined(pluginOptions, {
        output: "output",
        pattern: "pattern",
        omitIndex: "omitIndex",
        changefreq: "changefreq",
        priority: "priority",
        lastmod: "lastmod",
        links: "links",
        urlProperty: "urlProperty",
        modifiedProperty: "modifiedProperty",
        privateProperty: "privateProperty",
        priorityProperty: "priorityProperty",
        auto: "auto"
      }),
      // Override with explicit sitemap config (highest priority)
      ...(pluginOptions.sitemap || {})
    },
    // Complete robots configuration with defaults
    robots: {
      // Defaults
      generateRobots: true,
      addSitemapReference: true,
      disallowPaths: [],
      userAgent: "*",
      // Merge site metadata
      ...(siteMetadata.robots || {}),
      // Override with explicit robots config
      ...(pluginOptions.robots || {})
    }
  };
  return config;
}

/**
 * Validates the final configuration and throws helpful errors if required values are missing
 * @param {Object} config - The built configuration object
 * @param {string} metadataPath - The metadata path used for error messaging
 * @throws {Error} When hostname is not provided
 */
function validateConfig(config, metadataPath) {
  if (!config.hostname) {
    const metadataHint = metadataPath === "site" ? "site.url" : `${metadataPath}.url`;
    throw new Error(`[metalsmith-seo] hostname is required (set in plugin options or ${metadataHint} in metadata)`);
  }
}

/**
 * @fileoverview Comprehensive Metalsmith SEO plugin with head optimization, social media tags, and structured data.
 * @author Werner Glinka
 */

/**
 * @typedef {Object} SeoOptions
 * @property {string} hostname - Base hostname for all URLs
 * @property {string} [seoProperty='seo'] - Frontmatter property containing SEO data
 * @property {string} [metadataPath='site'] - Path to site metadata in metalsmith.metadata() (e.g., 'site' or 'data.site')
 * @property {Object} [defaults] - Default values for missing metadata
 * @property {Object} [fallbacks] - Fallback property mappings
 * @property {Object} [social] - Social media configuration
 * @property {Object} [jsonLd] - JSON-LD structured data configuration
 * @property {Object} [sitemap] - Sitemap generation options
 * @property {Object} [robots] - Robots.txt generation options
 * @property {boolean} [enableSitemap=true] - Whether to generate sitemap.xml
 * @property {boolean} [enableRobots=true] - Whether to generate/update robots.txt
 * @property {number} [batchSize=10] - Number of files to process in parallel
 * @property {number} [wordsPerMinute=200] - Reading speed for calculating reading time
 */

/**
 * @typedef {Object} SocialConfig
 * @property {string} [siteName] - Site name for Open Graph
 * @property {string} [locale='en_US'] - Default locale
 * @property {string} [twitterSite] - Twitter site handle (@username)
 * @property {string} [twitterCreator] - Default Twitter creator handle
 * @property {string} [facebookAppId] - Facebook App ID
 * @property {string|Array<string>} [facebookAdmins] - Facebook admin IDs
 * @property {string} [viewport] - Viewport meta tag content
 * @property {number} [twitterDescriptionLength=200] - Max length for Twitter descriptions
 */

/**
 * @typedef {Object} JsonLdConfig
 * @property {Object} [organization] - Organization schema data
 * @property {string} [searchUrl] - Search URL for WebSite schema
 * @property {Array<string>} [enableSchemas] - Schemas to auto-generate
 */

/**
 * @typedef {Object} SitemapConfig
 * @property {string} [output='sitemap.xml'] - Sitemap output filename
 * @property {string} [pattern] - File pattern for sitemap inclusion (default: all HTML files)
 * @property {boolean} [omitIndex=false] - Omit index.html from URLs
 * @property {boolean} [auto=false] - Auto-calculate priority and changefreq
 */

/**
 * @typedef {Object} RobotsConfig
 * @property {boolean} [generateRobots=true] - Generate robots.txt if none exists
 * @property {boolean} [addSitemapReference=true] - Add sitemap reference to existing robots.txt
 * @property {Array<string>} [disallowPaths=[]] - Additional paths to disallow
 * @property {string} [userAgent='*'] - User agent for robots directives
 */

/**
 * @typedef {string|SeoOptions} Options
 * @description Plugin configuration - can be a hostname string or full options object
 */

/**
 * Creates a comprehensive Metalsmith SEO plugin.
 *
 * This plugin provides unified SEO optimization including:
 * - HTML head optimization with meta tags
 * - Open Graph tags for social media sharing
 * - Twitter Card tags
 * - JSON-LD structured data
 * - Sitemap.xml generation
 *
 * @param {Options} options - Plugin configuration options
 *   When passed as a string, it will be treated as the hostname.
 *   When passed as an object, it supports the full configuration.
 * @returns {Function} Configured plugin function
 * @throws {Error} When hostname is not provided or configuration is invalid
 *
 * @example
 * // Simple usage with hostname only
 * metalsmith.use(seo('https://example.com'))
 *
 * @example
 * // Using site metadata from data.site instead of site
 * metalsmith.use(seo({
 *   metadataPath: 'data.site'  // Will look for metadata().data.site.url, etc.
 * }))
 *
 * @example
 * // Advanced configuration
 * metalsmith.use(seo({
 *   hostname: 'https://example.com',
 *   defaults: {
 *     title: 'My Site',
 *     description: 'Default description',
 *     socialImage: '/images/default-og.jpg'
 *   },
 *   social: {
 *     siteName: 'My Site',
 *     twitterSite: '@mysite'
 *   },
 *   jsonLd: {
 *     organization: {
 *       name: 'My Company',
 *       logo: 'https://example.com/logo.png'
 *     }
 *   }
 * }))
 */
function plugin(options = {}) {
  let pluginOptions = options;

  // Accept string option to specify the hostname
  if (typeof pluginOptions === "string") {
    try {
      new URL(pluginOptions);
      pluginOptions = {
        hostname: pluginOptions
      };
    } catch {
      throw new Error(`Invalid URL provided as hostname: ${pluginOptions}`);
    }
  }

  /**
   * Main plugin function that processes files for comprehensive SEO optimization.
   * @param {Object} files - Object containing all files in the build
   * @param {Object} metalsmith - Metalsmith instance for accessing utilities
   * @param {Function} done - Callback to signal completion
   * @returns {void}
   */
  return function (files, metalsmith, done) {
    // Get the metadata path (default to 'site')
    const metadataPath = pluginOptions.metadataPath || "site";

    // Get site metadata from the configured path
    const siteMetadata = getNestedProperty(metalsmith.metadata(), metadataPath) || {};

    // Get the configurable SEO property name (defaults to "seo")
    const seoProperty = pluginOptions.seoProperty || "seo";

    /**
     * Build the complete configuration
     * Configuration will be merged with site metadata, file frontmatter,
     * deduced values and defaults in priority order:
     * pluginOptions > siteMetadata > deduced values > defaults
     */
    const config = buildConfig(pluginOptions, siteMetadata, files, seoProperty);

    // Validate configuration
    validateConfig(config, metadataPath);

    /**
     * Optimize <head> section
     * Includes title, description, and social tags
     */
    const headOptimization = batchOptimizeHeads(files, config);

    // Sitemap generation
    let sitemapGeneration = Promise.resolve();
    if (config.enableSitemap) {
      // Add hostname to sitemap config (required by processor)
      config.sitemap.hostname = config.hostname;
      sitemapGeneration = processSitemap(files, metalsmith, config.sitemap);
    }

    // Execute head optimization and sitemap in parallel
    Promise.all([headOptimization, sitemapGeneration]).then(() => {
      // Robots.txt generation/update - after sitemap is done
      if (config.enableRobots) {
        // Add required runtime values to robots config
        config.robots.hostname = config.hostname;
        config.robots.sitemapFile = config.sitemap.output;
        return processRobots(files, metalsmith, config.robots);
      }
    }).then(() => done()).catch(done);
  };
}

// Set function name for better debugging
Object.defineProperty(plugin, "name", {
  value: "metalsmith-seo"
});

module.exports = plugin;
